{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8dace2b5-d022-4def-8853-459492e35a59",
   "metadata": {},
   "source": [
    "## Parser for SIGNOR\n",
    "\n",
    "1. It is part of the pharmacogenomics KG\n",
    "\n",
    "2. Notes:\n",
    "\n",
    "3. with predicate biolink:physically_interacts_with and biolink:gene_associated_with_condition\n",
    "\n",
    "4. the corresponding config.json file is: config_bigGIM_interacts_with_associated_with\n",
    "\n",
    "5. all of them have same cols \"subject, predicate, object, agent_type, knowledge_level, knowledge_source, object_category, publications, subject_category\"\n",
    "\n",
    "6. Full list of tsv files handled in this group is:\n",
    "    - signor_genes.csv\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "82d4d6e4-8693-4570-8e49-3fad091f0b39",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load necessary packages\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "# import matplotlib.pyplot as plt\n",
    "\n",
    "## load the needed tqdm package\n",
    "from tqdm import tqdm\n",
    "\n",
    "# Enable tqdm for pandas apply\n",
    "tqdm.pandas()\n",
    "\n",
    "## Define the version number\n",
    "version_number = \"08_18_2025\"\n",
    "deployment_date = \"2025-08-18\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91b21b86-1dd8-4308-b2a5-04cda6e05b93",
   "metadata": {},
   "source": [
    "### Get the category and predicate dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "0a375067-408d-4f05-8ce4-6bf2d6bdd6c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Date of last update:  2025-09-01\n",
      "Order is to always process Node/category map first, since the Edeg/predicate map depends on biolink-complainat node values\n",
      "-----------------------------------------------------------------------------------------------------------------------------\n",
      "Dictionary: category_map, Key template: Subject_category or Object_category\n",
      "------------------------------------------------------------------------------------------\n",
      "Dictionary: predicate_map, Key template: (Subject_category, Object_category, Predicate)\n"
     ]
    }
   ],
   "source": [
    "## Load the Biolink category and predicate dictionary for mapping subject, object, and predicate types\n",
    "%run ./Biolink_category_and_predication_dictionary.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "442c47a5-3dfe-44b5-b89e-091c88c122ea",
   "metadata": {},
   "source": [
    "### Get all helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "c799e5d1-b60c-49e1-872a-77023f2938a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load all helper functions\n",
    "%run /Users/Weiqi0/ISB_working/Hadlock_lab/QI_ISB_Git_repo/TranslatorPharcogenomicsKG/Parser_helper_functions.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "b55974d6-08c2-4b23-a305-a7c36bd1922e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## load the open source LLM model from openAI for batch quality check\n",
    "# ## example case to confirm working\n",
    "\n",
    "# import ollama\n",
    "# response = ollama.chat(\n",
    "#     model=\"gpt-oss:20b\",\n",
    "#     messages=[\n",
    "#         {\"role\": \"user\", \"content\": \"How does SLCO1B1 gene affects a patient's response to Atorvastatin and Lovastatin? only 2 sentences summary is enough\"},\n",
    "#     ],\n",
    "# )\n",
    "# print(response[\"message\"][\"content\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f01aaab9-dc31-40c4-a9da-eacd7aec16f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "# Define the query function with the new model\n",
    "def query_ollama(prompt, model='gpt-oss:20b'):\n",
    "    response = requests.post(\n",
    "        'http://localhost:11434/api/generate',\n",
    "        json={\n",
    "            \"model\": model,\n",
    "            \"prompt\": prompt,\n",
    "            \"stream\": False\n",
    "        }\n",
    "    )\n",
    "    response.raise_for_status()\n",
    "    return response.json()['response'].strip()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5fb9a27-d555-4e21-a34b-9ebb0ebd9404",
   "metadata": {},
   "source": [
    "## Load files and concatenate them into one merge tsv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "87f4fcb7-a821-4327-b3f9-1a5ea02fa795",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Notice!! Please change the file path of following codes into your own\n",
    "raw_files_path = '/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/SIGNOR/'\n",
    "\n",
    "## Define the output path for node & edge files after formatting\n",
    "download_path_node_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/SIGNOR_parsed_node_{version_number}.tsv'\n",
    "download_path_edge_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/SIGNOR_parsed_edge_{version_number}.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a982eeed-4ef1-4c24-9ccb-28c92df19882",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "signor_genes.csv\n"
     ]
    }
   ],
   "source": [
    "## Check all node files being read\n",
    "## Read all BigGIM node csv file in group 1\n",
    "\n",
    "for f in os.listdir(raw_files_path):\n",
    "    if f.endswith('.csv'):\n",
    "        print(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e6d72089-f035-418f-83b1-a77b188ee42b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current row number is:  38847\n",
      "current row number is:  38847\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_identifier</th>\n",
       "      <th>subject_name</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_identifier</th>\n",
       "      <th>object_name</th>\n",
       "      <th>object_category</th>\n",
       "      <th>predicate</th>\n",
       "      <th>original_predicate</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>Primary_Knowledge_Source</th>\n",
       "      <th>publications</th>\n",
       "      <th>knowledge_level</th>\n",
       "      <th>anatomical_context_qualifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:1457</td>\n",
       "      <td>CSNK2A1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene:11036</td>\n",
       "      <td>GTF2A1L</td>\n",
       "      <td>Gene</td>\n",
       "      <td>upregulates</td>\n",
       "      <td>up-regulates activity</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-250870</td>\n",
       "      <td>PMID:36243968|PMID:12107178</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBIGene:2932</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene:6688</td>\n",
       "      <td>SPI1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>downregulates</td>\n",
       "      <td>down-regulates quantity by destabilization</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-277542</td>\n",
       "      <td>PMID:36243968|PMID:33188146</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBIGene:5566</td>\n",
       "      <td>PRKACA</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene:2011</td>\n",
       "      <td>MARK2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>downregulates</td>\n",
       "      <td>down-regulates activity</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-276870</td>\n",
       "      <td>PMID:36243968|PMID:25512381</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBIGene:5793</td>\n",
       "      <td>PTPRG</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene:6714</td>\n",
       "      <td>SRC</td>\n",
       "      <td>Gene</td>\n",
       "      <td>downregulates</td>\n",
       "      <td>down-regulates activity</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-254725</td>\n",
       "      <td>PMID:36243968|PMID:25624455</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBIGene:5592</td>\n",
       "      <td>PRKG1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene:29109</td>\n",
       "      <td>FHOD1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>upregulates</td>\n",
       "      <td>up-regulates</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-170094</td>\n",
       "      <td>PMID:36243968|PMID:21106951</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>BTO:0000887;BTO:0001260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NCBIGene:5595</td>\n",
       "      <td>MAPK3</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene:8036</td>\n",
       "      <td>SHOC2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>downregulates</td>\n",
       "      <td>down-regulates quantity by destabilization</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-277443</td>\n",
       "      <td>PMID:36243968|PMID:30865892</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NCBIGene:5914</td>\n",
       "      <td>RARA</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene:1499</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>downregulates</td>\n",
       "      <td>down-regulates</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-73274</td>\n",
       "      <td>PMID:36243968|PMID:10607566</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NCBIGene:3172</td>\n",
       "      <td>HNF4A</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene:114902</td>\n",
       "      <td>C1QTNF5</td>\n",
       "      <td>Gene</td>\n",
       "      <td>upregulates</td>\n",
       "      <td>up-regulates quantity by expression</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-254448</td>\n",
       "      <td>PMID:36243968|PMID:20621834</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NCBIGene:51497</td>\n",
       "      <td>NELFCD</td>\n",
       "      <td>Gene</td>\n",
       "      <td>SIGNOR:SIGNOR-C521</td>\n",
       "      <td>NELF</td>\n",
       "      <td>Complex</td>\n",
       "      <td>in_complex</td>\n",
       "      <td>form complex</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-271399</td>\n",
       "      <td>PMID:36243968|PMID:18628398</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SIGNOR:SIGNOR-C409</td>\n",
       "      <td>Sin3B_complex</td>\n",
       "      <td>Complex</td>\n",
       "      <td>NCBIGene:653604</td>\n",
       "      <td>H3C13</td>\n",
       "      <td>Gene</td>\n",
       "      <td>downregulates</td>\n",
       "      <td>down-regulates activity</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-266974</td>\n",
       "      <td>PMID:36243968|PMID:21041486</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   subject_identifier   subject_name subject_category   object_identifier  \\\n",
       "0       NCBIGene:1457        CSNK2A1             Gene      NCBIGene:11036   \n",
       "1       NCBIGene:2932          GSK3B             Gene       NCBIGene:6688   \n",
       "2       NCBIGene:5566         PRKACA             Gene       NCBIGene:2011   \n",
       "3       NCBIGene:5793          PTPRG             Gene       NCBIGene:6714   \n",
       "4       NCBIGene:5592          PRKG1             Gene      NCBIGene:29109   \n",
       "5       NCBIGene:5595          MAPK3             Gene       NCBIGene:8036   \n",
       "6       NCBIGene:5914           RARA             Gene       NCBIGene:1499   \n",
       "7       NCBIGene:3172          HNF4A             Gene     NCBIGene:114902   \n",
       "8      NCBIGene:51497         NELFCD             Gene  SIGNOR:SIGNOR-C521   \n",
       "9  SIGNOR:SIGNOR-C409  Sin3B_complex          Complex     NCBIGene:653604   \n",
       "\n",
       "  object_name object_category      predicate  \\\n",
       "0     GTF2A1L            Gene    upregulates   \n",
       "1        SPI1            Gene  downregulates   \n",
       "2       MARK2            Gene  downregulates   \n",
       "3         SRC            Gene  downregulates   \n",
       "4       FHOD1            Gene    upregulates   \n",
       "5       SHOC2            Gene  downregulates   \n",
       "6      CTNNB1            Gene  downregulates   \n",
       "7     C1QTNF5            Gene    upregulates   \n",
       "8        NELF         Complex     in_complex   \n",
       "9       H3C13            Gene  downregulates   \n",
       "\n",
       "                           original_predicate provided_by  \\\n",
       "0                       up-regulates activity      SIGNOR   \n",
       "1  down-regulates quantity by destabilization      SIGNOR   \n",
       "2                     down-regulates activity      SIGNOR   \n",
       "3                     down-regulates activity      SIGNOR   \n",
       "4                                up-regulates      SIGNOR   \n",
       "5  down-regulates quantity by destabilization      SIGNOR   \n",
       "6                              down-regulates      SIGNOR   \n",
       "7         up-regulates quantity by expression      SIGNOR   \n",
       "8                                form complex      SIGNOR   \n",
       "9                     down-regulates activity      SIGNOR   \n",
       "\n",
       "  Primary_Knowledge_Source                 publications      knowledge_level  \\\n",
       "0            SIGNOR-250870  PMID:36243968|PMID:12107178  knowledge_assertion   \n",
       "1            SIGNOR-277542  PMID:36243968|PMID:33188146  knowledge_assertion   \n",
       "2            SIGNOR-276870  PMID:36243968|PMID:25512381  knowledge_assertion   \n",
       "3            SIGNOR-254725  PMID:36243968|PMID:25624455  knowledge_assertion   \n",
       "4            SIGNOR-170094  PMID:36243968|PMID:21106951  knowledge_assertion   \n",
       "5            SIGNOR-277443  PMID:36243968|PMID:30865892  knowledge_assertion   \n",
       "6             SIGNOR-73274  PMID:36243968|PMID:10607566  knowledge_assertion   \n",
       "7            SIGNOR-254448  PMID:36243968|PMID:20621834  knowledge_assertion   \n",
       "8            SIGNOR-271399  PMID:36243968|PMID:18628398  knowledge_assertion   \n",
       "9            SIGNOR-266974  PMID:36243968|PMID:21041486  knowledge_assertion   \n",
       "\n",
       "  anatomical_context_qualifier  \n",
       "0                          NaN  \n",
       "1                          NaN  \n",
       "2                          NaN  \n",
       "3                          NaN  \n",
       "4      BTO:0000887;BTO:0001260  \n",
       "5                          NaN  \n",
       "6                          NaN  \n",
       "7                          NaN  \n",
       "8                          NaN  \n",
       "9                          NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Read each individual csv files\n",
    "source_df = pd.read_csv(raw_files_path + 'signor_genes.csv')\n",
    "\n",
    "## check the length\n",
    "print(\"current row number is: \", len(source_df))\n",
    "\n",
    "## Drop nan values\n",
    "source_df = source_df.dropna(subset=['subject_identifier', 'object_identifier'])\n",
    "\n",
    "## check the length after filtering\n",
    "print(\"current row number is: \", len(source_df))\n",
    "\n",
    "source_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "031ed526-0519-4f69-82d8-8bcdd045ca8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['upregulates' 'downregulates' 'in_complex' 'unknown']\n"
     ]
    }
   ],
   "source": [
    "## check all unique values in the predicate column\n",
    "unique_values = source_df['predicate'].unique()\n",
    "print(unique_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d1d6caef-34ea-4277-8126-3f96e38e8a06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current row number is:  37764\n"
     ]
    }
   ],
   "source": [
    "## remove those rows with col(predicate) == 'unknown' from source_df\n",
    "source_df = source_df[source_df['predicate'] != 'unknown']\n",
    "\n",
    "print(\"current row number is: \", len(source_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "15be7183-2a94-4b40-97e0-3bc6ca3f25e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Gene' 'Complex' 'Chemical' 'Phenotype' 'Protein' 'Smallmolecule'\n",
      " 'Proteinfamily' 'Stimulus' 'Fusion Protein' 'Mirna' 'Antibody' 'Ncrna']\n"
     ]
    }
   ],
   "source": [
    "## check all unique values in the subject_category column\n",
    "unique_values = source_df['subject_category'].unique()\n",
    "print(unique_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ac35bbd9-4b8b-4a39-91a3-d2780d2b74e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Gene' 'Complex' 'Phenotype' 'Smallmolecule' 'Proteinfamily' 'Chemical'\n",
      " 'Protein' 'Fusion Protein' 'Mirna' 'Stimulus']\n"
     ]
    }
   ],
   "source": [
    "## check all unique values in the subject_category column\n",
    "unique_values = source_df['object_category'].unique()\n",
    "print(unique_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c1867168-bae4-4141-9521-617f39d0b209",
   "metadata": {},
   "outputs": [],
   "source": [
    "## check all unique combinations between subject & object\n",
    "from itertools import combinations\n",
    "\n",
    "# Get all combinations of 3 column names\n",
    "col_combos = ['subject_category', 'object_category', 'predicate']\n",
    "\n",
    "unique_combinations = source_df[col_combos].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "835ca26e-1511-4d50-8b69-9e576e7f0d3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      subject_category object_category      predicate\n",
      "0                 Gene            Gene    upregulates\n",
      "1                 Gene            Gene  downregulates\n",
      "8                 Gene         Complex     in_complex\n",
      "9              Complex            Gene  downregulates\n",
      "11             Complex       Phenotype    upregulates\n",
      "...                ...             ...            ...\n",
      "22783            Mirna       Phenotype  downregulates\n",
      "24774    Smallmolecule         Protein  downregulates\n",
      "32052    Smallmolecule         Complex  downregulates\n",
      "32220        Phenotype   Proteinfamily    upregulates\n",
      "34290             Gene            Gene     in_complex\n",
      "\n",
      "[98 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "print(unique_combinations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "becfcc29-3765-4310-a62e-ed20bd56d0d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the output path for node & edge files after formatting\n",
    "download_path_temp_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/temp/unique_combos_subject_object_predicate.tsv'\n",
    "\n",
    "\n",
    "## download both node and edge files\n",
    "## Download the result df\n",
    "## disable download for testing\n",
    "unique_combinations.to_csv(download_path_temp_file, sep ='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca7773d2-97c3-41ce-9c27-e6697faf784b",
   "metadata": {},
   "source": [
    "## Subject & object mapping\n",
    "* using the current Biolink_category_and_predication_dictionary\n",
    "* existing biolink category mapping: \n",
    "    * 'Gene': 'biolink:Gene',\n",
    "    * 'Chemical': 'biolink:ChemicalEntity',\n",
    "    * 'Smallmolecule': 'biolink:SmallMolecule',\n",
    "    * 'Phenotype': 'biolink:PhenotypicFeature',\n",
    "    * 'Protein': 'biolink:Protein',\n",
    "* went through valid check cause there is potential issue:\n",
    "    * 'Antibody': 'biolink:Drug', ## check if all are indeed drug\n",
    "    * 'Complex': 'biolink:MacromolecularComplex',\n",
    "    * 'Mirna': 'biolink:MicroRNA',\n",
    "    * 'Ncrna': 'biolink:Noncoding_RNAProduct',\n",
    "* proposed in the translator ingest group ticket for furture category:\n",
    "* not included in current version of parsed file\n",
    "    * 'Fusion Protein': parent node, 'biolink:Protein', ## Potential new entity to raise for data ingest meeting\n",
    "    * 'Stimulus': partially overlap with 'biolink:EnvironmentalProcess', ## Potential new entity to raise for data ingest meeting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "106912eb-3959-4e95-9cee-60d9339cb62d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_17658/2806490594.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  node_subject_df.rename(columns={'subject_identifier': 'id', 'subject_name': 'name', 'subject_category': 'category'}, inplace=True)\n",
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_17658/2806490594.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  node_object_df.rename(columns={'object_identifier': 'id', 'object_name': 'name', 'object_category': 'category'}, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:1457</td>\n",
       "      <td>CSNK2A1</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBIGene:2932</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBIGene:5566</td>\n",
       "      <td>PRKACA</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBIGene:5793</td>\n",
       "      <td>PTPRG</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBIGene:5592</td>\n",
       "      <td>PRKG1</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              id     name category\n",
       "0  NCBIGene:1457  CSNK2A1     Gene\n",
       "1  NCBIGene:2932    GSK3B     Gene\n",
       "2  NCBIGene:5566   PRKACA     Gene\n",
       "3  NCBIGene:5793    PTPRG     Gene\n",
       "4  NCBIGene:5592    PRKG1     Gene"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## create a temp node file\n",
    "\n",
    "node_subject_df = source_df[['subject_identifier', 'subject_name', 'subject_category']]\n",
    "node_object_df = source_df[['object_identifier',\t'object_name',\t'object_category']]\n",
    "\n",
    "## rename those columns into desired format\n",
    "node_subject_df.rename(columns={'subject_identifier': 'id', 'subject_name': 'name', 'subject_category': 'category'}, inplace=True)\n",
    "node_object_df.rename(columns={'object_identifier': 'id', 'object_name': 'name', 'object_category': 'category'}, inplace=True)\n",
    "\n",
    "concat_node_df = pd.concat([node_subject_df, node_object_df]).drop_duplicates(keep='first')\n",
    "\n",
    "concat_node_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d507ad9e-88e1-4040-a6bd-ba0c96b2b023",
   "metadata": {},
   "source": [
    "### Now check all \"gene\" nodes\n",
    "* use LLM model to confirm they have the correct ids\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9fdc47ff-2f5a-402a-a770-1ab3836bc0d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene_node_df = concat_node_df[concat_node_df['category'] == 'Gene']\n",
    "# print(len(gene_node_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d80a90cb-884c-49ef-b5df-d312eda047e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Define the row labeling function\n",
    "# def gene_row_checker(row):\n",
    "#     combined_text = f\"{row['name']} {row['id']}\"\n",
    "#     prompt = f\"\"\"Analyze if the NCBI gene identifier and names are matching and if it is an existing Gene, answer only in \"yes\" if both satisfied, or \"no\" if any condition is incorrect:\n",
    "# \\\"\\\"\\\"{combined_text}\\\"\\\"\\\"\n",
    "# Sentiment:\"\"\"\n",
    "#     return query_ollama(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "2bad1a76-92c5-49d1-b992-437f103fa7ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## apply to each row of the dataframe\n",
    "# gene_node_df['label'] = gene_node_df.progress_apply(gene_row_checker, axis=1)\n",
    "# print(gene_node_df[gene_node_df['label'] == 'no'])\n",
    "# print(len(antibody_node_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "599978ed-8026-4dbf-804b-253ec10e9086",
   "metadata": {},
   "source": [
    "### Now check all \"antibody\" nodes\n",
    "* use LLM model as a simple example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e170747f-6e53-42df-86fa-e9791d87eff6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25\n"
     ]
    }
   ],
   "source": [
    "antibody_node_df = concat_node_df[concat_node_df['category'] == 'Antibody']\n",
    "print(len(antibody_node_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "22d8c399-324a-4e7f-81ed-749107da7306",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     id                                      name  category\n",
      "3156   DRUGBANK:DB06650                                ofatumumab  Antibody\n",
      "3449   DRUGBANK:DB09037                             pembrolizumab  Antibody\n",
      "5396   DRUGBANK:DB00087                               alemtuzumab  Antibody\n",
      "5782   DRUGBANK:DB06273                               Tocilizumab  Antibody\n",
      "7210   DRUGBANK:DB08904                     Certolizumab (Cimzia)  Antibody\n",
      "7223   DRUGBANK:DB06366                                pertuzumab  Antibody\n",
      "7558   DRUGBANK:DB00112                               bevacizumab  Antibody\n",
      "8145   DRUGBANK:DB08935                              obinutuzumab  Antibody\n",
      "10318  DRUGBANK:DB05773                 ado-trastuzumab emtansine  Antibody\n",
      "10909  DRUGBANK:DB09052                              blinatumomab  Antibody\n",
      "14210  DRUGBANK:DB00056                     gemtuzumab ozogamicin  Antibody\n",
      "14477  DRUGBANK:DB06674                       Golimumab (Simponi)  Antibody\n",
      "15942  DRUGBANK:DB09035                                 nivolumab  Antibody\n",
      "17459  DRUGBANK:DB05578                               ramucirumab  Antibody\n",
      "17698  DRUGBANK:DB06643                                 denosumab  Antibody\n",
      "18040  DRUGBANK:DB01269                               panitumumab  Antibody\n",
      "24386  DRUGBANK:DB00078                      ibritumomab tiuxetan  Antibody\n",
      "24556  DRUGBANK:DB06186                                ipilimumab  Antibody\n",
      "24632  DRUGBANK:DB00072                               trastuzumab  Antibody\n",
      "24877  DRUGBANK:DB00002                                 cetuximab  Antibody\n",
      "26663  DRUGBANK:DB00051                      Adalimumab (Imraldi)  Antibody\n",
      "27311  DRUGBANK:DB00081  tositumomab and iodine i 131 tositumomab  Antibody\n",
      "28318  DRUGBANK:DB09029                    Secukinumab (Cosentyx)  Antibody\n",
      "32761  DRUGBANK:DB00065                     Infliximab (Remicade)  Antibody\n",
      "33419  DRUGBANK:DB00073                                 rituximab  Antibody\n"
     ]
    }
   ],
   "source": [
    "print(antibody_node_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "673c1b6d-8271-4d10-a812-e3cabe07d889",
   "metadata": {},
   "outputs": [],
   "source": [
    "## run the LLM model using a custom query\n",
    "\n",
    "# Define the row labeling function\n",
    "def antibody_row_checker(row):\n",
    "    combined_text = f\"{row['name']} {row['id']}\"\n",
    "    prompt = f\"\"\"Analyze if the drugbank identifier and names are matching and if it is an existing FDA-approved drug, answer only in \"yes\" if both satisfied, or \"no\" if any condition is incorrect:\n",
    "\\\"\\\"\\\"{combined_text}\\\"\\\"\\\"\n",
    "Sentiment:\"\"\"\n",
    "    return query_ollama(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "5a7d6466-e034-4d82-95c6-59270e5b17f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     id        name  category label\n",
      "24556  DRUGBANK:DB06186  ipilimumab  Antibody    no\n",
      "24877  DRUGBANK:DB00002   cetuximab  Antibody    no\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_79386/919110168.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  antibody_node_df['label'] = antibody_node_df.apply(antibody_row_checker, axis=1)\n"
     ]
    }
   ],
   "source": [
    "## apply to each row of the dataframe\n",
    "antibody_node_df['label'] = antibody_node_df.apply(antibody_row_checker, axis=1)\n",
    "print(antibody_node_df[antibody_node_df['label'] == 'no'])\n",
    "# print(antibody_node_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95092151-eace-447a-af59-fb7595457c54",
   "metadata": {},
   "source": [
    "### LLM flagged edges -> manual check\n",
    "* DRUGBANK:DB06186  ipilimumab is correct: https://go.drugbank.com/drugs/DB06186\n",
    "* DRUGBANK:DB00002   cetuximab is correct: https://go.drugbank.com/drugs/DB00002"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "0785837a-bb4d-4395-b4ed-0f6d66ed1dd0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    id        name  category  \\\n",
      "3156  DRUGBANK:DB06650  ofatumumab  Antibody   \n",
      "\n",
      "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 label  \n",
      "3156  **Verification**\\n\\n| Field | Value | Source/Notes |\\n|-------|-------|--------------|\\n| **DrugBank ID** | `DB06650` | Listed in DrugBank as ofatumumab |\\n| **Drug name** | `ofatumumab` | Matches the ID in the database |\\n| **FDA approval status** | **Approved** | • U.S. FDA approved (2014) for chronic lymphocytic leukemia (brand name *Arzerra*).  <br>• FDA approved (2019) for relapsing forms of multiple sclerosis (brand name *Kesimpta*). |\\n| **Mechanism of action** | Human monoclonal antibody targeting CD20 on B‑cells | Induces complement‑mediated cytotoxicity and apoptosis |\\n\\n**Conclusion**\\n\\n- The DrugBank identifier **DB06650** correctly corresponds to **ofatumumab**.  \\n- Ofatumumab is an existing FDA‑approved therapeutic agent, with approvals for chronic lymphocytic leukemia and relapsing multiple sclerosis.\\n\\n**Sentiment**: Neutral (factual confirmation).  \n"
     ]
    }
   ],
   "source": [
    "# # Temporarily show full column width (no truncation)\n",
    "# pd.set_option('display.max_colwidth', None)\n",
    "\n",
    "# print(antibody_node_df.head(1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09cba8a8-ac20-42f7-866e-18ed8b6b052e",
   "metadata": {},
   "source": [
    "### Now check all \"Complex\" nodes\n",
    "* "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a57b6527-edc5-4a55-9164-4bfffd73caad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "519\n"
     ]
    }
   ],
   "source": [
    "complex_node_df = concat_node_df[concat_node_df['category'] == 'Complex']\n",
    "print(len(complex_node_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "43ae0be6-287a-4e18-b6bd-493475dae905",
   "metadata": {},
   "outputs": [],
   "source": [
    "## run the LLM model using a custom query\n",
    "\n",
    "# Define the row labeling function\n",
    "def complex_row_checker(row):\n",
    "    combined_text = f\"{row['name']} {row['id']}\"\n",
    "    prompt = f\"\"\"Analyze if it can be considered as a MacromolecularComplex, answer only in \"yes\" if satisfied, or \"no\" if condition is incorrect:\n",
    "\\\"\\\"\\\"{combined_text}\\\"\\\"\\\"\n",
    "Sentiment:\"\"\"\n",
    "    return query_ollama(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e58a374e-2bc3-426c-86f6-576fed5cdfa2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 38%|██████████████████████████▉                                            | 197/519 [16:42<28:59,  5.40s/it]"
     ]
    }
   ],
   "source": [
    "## apply to each row of the dataframe\n",
    "complex_node_df['label'] = complex_node_df.progress_apply(complex_row_checker, axis=1)\n",
    "print(complex_node_df[complex_node_df['label'] == 'no'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "32df2618-a175-4778-8722-f20d46058c0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       id                                          name  \\\n",
      "427    SIGNOR:SIGNOR-C532                                         Erlin   \n",
      "521    SIGNOR:SIGNOR-C217                                           DGC   \n",
      "757    SIGNOR:SIGNOR-C101                                           TSC   \n",
      "831    SIGNOR:SIGNOR-C398                     Succinyl-CoA  ATP variant   \n",
      "971    SIGNOR:SIGNOR-C131                                           NAE   \n",
      "1441   SIGNOR:SIGNOR-C517                             R2SP co-chaperone   \n",
      "1660   SIGNOR:SIGNOR-C113                                         RAGAC   \n",
      "2073   SIGNOR:SIGNOR-C411                  Inner_mitochondrial_membrane   \n",
      "2388   SIGNOR:SIGNOR-C481   Muscle cell-specific SWI/SNF ARID1A variant   \n",
      "4017   SIGNOR:SIGNOR-C372                                          POMT   \n",
      "5022   SIGNOR:SIGNOR-C396                                           IDH   \n",
      "6244   SIGNOR:SIGNOR-C449                                          WICH   \n",
      "6963   SIGNOR:SIGNOR-C555                                           CBM   \n",
      "7197   SIGNOR:SIGNOR-C399                     Succinyl-CoA  GTP variant   \n",
      "7799   SIGNOR:SIGNOR-C218                                  MAPK12/CARM1   \n",
      "8996   SIGNOR:SIGNOR-C501                             MCU_MICU3_variant   \n",
      "10097  SIGNOR:SIGNOR-C349                              NMDA receptor_2C   \n",
      "11559  SIGNOR:SIGNOR-C486        Brain-specific SWI/SNF SMARCA4 variant   \n",
      "12008   SIGNOR:SIGNOR-C54                                   KAT6A/KAT6B   \n",
      "12649  SIGNOR:SIGNOR-C463                                           ETF   \n",
      "12772  SIGNOR:SIGNOR-C219                                        LPR5/6   \n",
      "12961  SIGNOR:SIGNOR-C474                                      MPO-ANCA   \n",
      "13520  SIGNOR:SIGNOR-C448                                         HNuRF   \n",
      "20534  SIGNOR:SIGNOR-C475                                      Pr3-ANCA   \n",
      "24523  SIGNOR:SIGNOR-C485        Brain-specific SWI/SNF SMARCA2 variant   \n",
      "26534  SIGNOR:SIGNOR-C109                                  ATG12/5/16L1   \n",
      "29866  SIGNOR:SIGNOR-C523                                           KPC   \n",
      "30387  SIGNOR:SIGNOR-C530                                          EDVP   \n",
      "30578  SIGNOR:SIGNOR-C430                       α-D-glucosyl-glycogenin   \n",
      "31182  SIGNOR:SIGNOR-C412                                           TSH   \n",
      "31200  SIGNOR:SIGNOR-C199                                 FCER1G/FCER1G   \n",
      "31280  SIGNOR:SIGNOR-C522                             BAF250b E3 ligase   \n",
      "31887  SIGNOR:SIGNOR-C469                                       RalGAP2   \n",
      "31948  SIGNOR:SIGNOR-C436                                        BCR-Dl   \n",
      "32741  SIGNOR:SIGNOR-C400                                           SDH   \n",
      "33571  SIGNOR:SIGNOR-C483  Muscle cell-specific SWI/SNF SMARCA4 variant   \n",
      "33869  SIGNOR:SIGNOR-C554                                           CPC   \n",
      "35458  SIGNOR:SIGNOR-C340                                          CERF   \n",
      "36224  SIGNOR:SIGNOR-C468                                       RalGAP1   \n",
      "37186  SIGNOR:SIGNOR-C467                                          GBAF   \n",
      "1724    SIGNOR:SIGNOR-C19                                           HMT   \n",
      "1827   SIGNOR:SIGNOR-C149                                         hb:hp   \n",
      "5023   SIGNOR:SIGNOR-C129                                     MYOD/E2-2   \n",
      "7082   SIGNOR:SIGNOR-C128                                      MYOD/HEB   \n",
      "7320    SIGNOR:SIGNOR-C52                                FGFR1OP/CEP350   \n",
      "10361  SIGNOR:SIGNOR-C116                                         RAGBD   \n",
      "11324  SIGNOR:SIGNOR-C410                  Outer_mitochondrial_membrane   \n",
      "13740  SIGNOR:SIGNOR-C114                                         RAGAD   \n",
      "18307   SIGNOR:SIGNOR-C25                                    NKX3-1/SRF   \n",
      "22927  SIGNOR:SIGNOR-C115                                         RAGBC   \n",
      "23390   SIGNOR:SIGNOR-C75                                   SS18/MLLT10   \n",
      "\n",
      "      category label  \n",
      "427    Complex    no  \n",
      "521    Complex    no  \n",
      "757    Complex    no  \n",
      "831    Complex    no  \n",
      "971    Complex    no  \n",
      "1441   Complex    no  \n",
      "1660   Complex    no  \n",
      "2073   Complex    no  \n",
      "2388   Complex    no  \n",
      "4017   Complex    no  \n",
      "5022   Complex    no  \n",
      "6244   Complex    no  \n",
      "6963   Complex    no  \n",
      "7197   Complex    no  \n",
      "7799   Complex    no  \n",
      "8996   Complex    no  \n",
      "10097  Complex    no  \n",
      "11559  Complex    no  \n",
      "12008  Complex    no  \n",
      "12649  Complex    no  \n",
      "12772  Complex    no  \n",
      "12961  Complex    no  \n",
      "13520  Complex    no  \n",
      "20534  Complex    no  \n",
      "24523  Complex    no  \n",
      "26534  Complex    no  \n",
      "29866  Complex    no  \n",
      "30387  Complex    no  \n",
      "30578  Complex    no  \n",
      "31182  Complex    no  \n",
      "31200  Complex    no  \n",
      "31280  Complex    no  \n",
      "31887  Complex    no  \n",
      "31948  Complex    no  \n",
      "32741  Complex    no  \n",
      "33571  Complex    no  \n",
      "33869  Complex    no  \n",
      "35458  Complex    no  \n",
      "36224  Complex    no  \n",
      "37186  Complex    no  \n",
      "1724   Complex    no  \n",
      "1827   Complex    no  \n",
      "5023   Complex    no  \n",
      "7082   Complex    no  \n",
      "7320   Complex    no  \n",
      "10361  Complex    no  \n",
      "11324  Complex    no  \n",
      "13740  Complex    no  \n",
      "18307  Complex    no  \n",
      "22927  Complex    no  \n",
      "23390  Complex    no  \n"
     ]
    }
   ],
   "source": [
    "print(complex_node_df[complex_node_df['label'] == 'no'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "921c7df8-59e5-4d6b-bc63-950312d8bfe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the output path for node & edge files after formatting\n",
    "download_path_temp_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/temp/Signor_LLM_complex_no_rows.tsv'\n",
    "\n",
    "\n",
    "## download both node and edge files\n",
    "## Download the result df\n",
    "## disable download for testing\n",
    "complex_node_df[complex_node_df['label'] == 'no'].to_csv(download_path_temp_file, sep ='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c238df66-592a-45f9-a3b0-6755a6132910",
   "metadata": {},
   "source": [
    "### LLM flagged edges -> manual check\n",
    "* Those entities exist on SIGNOR, need to verify them on a thrid party source\n",
    "* tried following DB: \n",
    "    * https://www.ebi.ac.uk/complexportal/\n",
    "    * https://mmcif.wwpdb.org/cgi-bin/swish/swish.cgi?query=&si=0&sort=swishrank\n",
    "* But I need domain expert help to decipher those results\n",
    "* **Now remove all those rows being flagged by LLM**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b71f2d59-e145-40bd-a6b0-647d7bea71a5",
   "metadata": {},
   "source": [
    "### Now check all \"Mirna\" nodes\n",
    "* "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "2910dc08-cb85-4d32-863e-68e4fc6d6493",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25\n"
     ]
    }
   ],
   "source": [
    "Mirna_node_df = concat_node_df[concat_node_df['category'] == 'Mirna']\n",
    "print(len(Mirna_node_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "d8849566-8402-4e98-a9df-bec04281dc7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## run the LLM model using a custom query\n",
    "# Define the row labeling function\n",
    "def Mirna_row_checker(row):\n",
    "    combined_text = f\"{row['name']} {row['id']}\"\n",
    "    prompt = f\"\"\"Analyze if the identifier and names are matching and if it is an existing MicroRNA, answer only in \"yes\" if both satisfied, or \"no\" if any condition is incorrect:\n",
    "\\\"\\\"\\\"{combined_text}\\\"\\\"\\\"\n",
    "Sentiment:\"\"\"\n",
    "    return query_ollama(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "0ec28e0d-d324-4dca-809e-96cc053b520b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                  id         name category label\n",
      "1642   RNAcentral:URS000039E12D_9606     miR-130a    Mirna   yes\n",
      "3330   RNAcentral:URS000062749E_9606      miR-155    Mirna   yes\n",
      "4156   RNAcentral:URS00004CAC40_9606      mir-10b    Mirna   yes\n",
      "4892   RNAcentral:URS0000150A7D_9606      miR-29b    Mirna   yes\n",
      "6840   RNAcentral:URS0000759977_9606     miR-199a    Mirna   yes\n",
      "6927   RNAcentral:URS00005EB596_9606    mir-133a1    Mirna   yes\n",
      "8331   RNAcentral:URS0000233054_9606      miR-27a    Mirna   yes\n",
      "11333  RNAcentral:URS000075CF56_9606       MIR1-1    Mirna    no\n",
      "12565  RNAcentral:URS000037EC34_9606  hsa-mir-223    Mirna   yes\n",
      "13634  RNAcentral:URS00001CC864_9606      miR-23a    Mirna   yes\n",
      "14078  RNAcentral:URS0000245997_9606       miR221    Mirna   yes\n",
      "16566  RNAcentral:URS00001F4E81_9606      miR-132    Mirna   yes\n",
      "18198  RNAcentral:URS000075C517_9606      miR-495    Mirna   yes\n",
      "26261  RNAcentral:URS000075B799_9606      miR-29c    Mirna   yes\n",
      "27280  RNAcentral:URS000075D8A0_9606     miR-146a    Mirna    no\n",
      "30558  RNAcentral:URS0000389B41_9606      mir-206    Mirna   yes\n",
      "30969  RNAcentral:URS00004E9304_9606      miR-29a    Mirna   yes\n",
      "32214  RNAcentral:URS000075B0A5_9606      miR-27b    Mirna   yes\n",
      "32889  RNAcentral:URS0000283D0A_9606      miR-23b    Mirna   yes\n",
      "35695  RNAcentral:URS000075AE07_9606   miR-142-3p    Mirna   yes\n",
      "36273  RNAcentral:URS0000008A99_9606      mir-143    Mirna   yes\n",
      "38190  RNAcentral:URS000021E7E5_9606     miR-4784    Mirna    no\n",
      "826    RNAcentral:URS000075E706_9606      miR-338    Mirna   yes\n",
      "2631   RNAcentral:URS000033F823_9606       miR-34    Mirna    no\n",
      "22957  RNAcentral:URS00005675D3_9606    mir-133a2    Mirna   yes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_79386/849294236.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Mirna_node_df['label'] = Mirna_node_df.apply(Mirna_row_checker, axis=1)\n"
     ]
    }
   ],
   "source": [
    "## apply to each row of the dataframe\n",
    "Mirna_node_df['label'] = Mirna_node_df.apply(Mirna_row_checker, axis=1)\n",
    "print(Mirna_node_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "75758f70-bfcf-467f-af91-dd08427bbd6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                  id      name category label\n",
      "11333  RNAcentral:URS000075CF56_9606    MIR1-1    Mirna    no\n",
      "27280  RNAcentral:URS000075D8A0_9606  miR-146a    Mirna    no\n",
      "38190  RNAcentral:URS000021E7E5_9606  miR-4784    Mirna    no\n",
      "2631   RNAcentral:URS000033F823_9606    miR-34    Mirna    no\n"
     ]
    }
   ],
   "source": [
    "print(Mirna_node_df[Mirna_node_df['label'] == 'no'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a45d533-a1c8-4df5-8706-498809aa6700",
   "metadata": {},
   "source": [
    "### LLM flagged -> manual check\n",
    "* RNAcentral:URS000075CF56_9606       MIR1-1    Mirna is correct, https://rnacentral.org/rna/URS000075CF56/9606\n",
    "* RNAcentral:URS000075D8A0_9606     miR-146a    Mirna is correct, https://rnacentral.org/rna/URS000075D8A0/9606\n",
    "* RNAcentral:URS000021E7E5_9606     miR-4784    Mirna is correct, https://rnacentral.org/rna/URS000021E7E5/9606\n",
    "* RNAcentral:URS000033F823_9606       miR-34    Mirna is indeed incorrect, \n",
    "* **https://rnacentral.org/rna/URS000033F823/9606, should be mir-34a**\n",
    "* https://rnacentral.org/rna/URS0002914588/9606, miR-34 is actually the precursor family\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bbefc33-abca-44ef-a585-e281d666a11f",
   "metadata": {},
   "source": [
    "### Now check all \"Ncrna\" nodes\n",
    "* "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "cff718d5-f92f-4a37-a9cb-a0a648041b5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "Ncrna_node_df = concat_node_df[concat_node_df['category'] == 'Ncrna']\n",
    "print(len(Ncrna_node_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "50ad6822-6aab-43da-9dc8-e8474e796981",
   "metadata": {},
   "outputs": [],
   "source": [
    "## run the LLM model using a custom query\n",
    "# Define the row labeling function\n",
    "def Ncrna_row_checker(row):\n",
    "    combined_text = f\"{row['name']} {row['id']}\"\n",
    "    prompt = f\"\"\"Analyze if the identifier and names are matching and if it is an existing Noncoding_RNAProduct, answer only in \"yes\" if both satisfied, or \"no\" if any condition is incorrect:\n",
    "\\\"\\\"\\\"{combined_text}\\\"\\\"\\\"\n",
    "Sentiment:\"\"\"\n",
    "    return query_ollama(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "e505bb34-aa88-4b5f-8896-70bf06f80eaa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                 id    name category label\n",
      "4968  RNAcentral:URS000075C808_9606  HOTAIR    Ncrna   yes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_79386/2995235901.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Ncrna_node_df['label'] = Ncrna_node_df.apply(Ncrna_row_checker, axis=1)\n"
     ]
    }
   ],
   "source": [
    "## apply to each row of the dataframe\n",
    "Ncrna_node_df['label'] = Ncrna_node_df.apply(Ncrna_row_checker, axis=1)\n",
    "print(Ncrna_node_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0326321c-eea1-43b2-baef-7f14e756916c",
   "metadata": {},
   "source": [
    "### LLM flagged -> manual check\n",
    "* all seems correct now"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb343444-682b-46cf-b47d-3baf52da839d",
   "metadata": {},
   "source": [
    "## Now check stimulus\n",
    "* "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "77a1ce1e-3825-46db-85f9-d4239865d4d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26\n"
     ]
    }
   ],
   "source": [
    "## now try to check Stimulus\n",
    "Stimulus_node_df = concat_node_df[concat_node_df['category'] == 'Stimulus']\n",
    "print(len(Stimulus_node_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "3dc8a521-7b56-4272-8e7e-5993201fc2c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## run the LLM model using a custom query\n",
    "# Define the row labeling function\n",
    "def Stimulus_row_checker(row):\n",
    "    combined_text = f\"{row['name']} {row['id']}\"\n",
    "    prompt = f\"\"\"Analyze if the identifier and names are matching and if it is an existing EnvironmentalProcess, answer only in \"yes\" if both satisfied, or \"no\" if any condition is incorrect:\n",
    "\\\"\\\"\\\"{combined_text}\\\"\\\"\\\"\n",
    "Sentiment:\"\"\"\n",
    "    return query_ollama(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "eaac3e28-ae3c-493d-b1bf-992d0ae02768",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       id                 name  category label\n",
      "123    SIGNOR:SIGNOR-ST22    Unfolded_Proteins  Stimulus    no\n",
      "180     SIGNOR:SIGNOR-ST1           DNA_damage  Stimulus    no\n",
      "183    SIGNOR:SIGNOR-ST11                PAMPs  Stimulus    no\n",
      "282    SIGNOR:SIGNOR-ST28       Osmotic_stress  Stimulus    no\n",
      "559    SIGNOR:SIGNOR-ST20                  ECM  Stimulus    no\n",
      "721     SIGNOR:SIGNOR-ST2                  ROS  Stimulus    no\n",
      "1461    SIGNOR:SIGNOR-ST7            UV stress  Stimulus    no\n",
      "2710   SIGNOR:SIGNOR-ST26  Blood vessel damage  Stimulus    no\n",
      "3057    SIGNOR:SIGNOR-ST5           AminoAcids  Stimulus    no\n",
      "3110   SIGNOR:SIGNOR-ST17        UVB radiation  Stimulus    no\n",
      "3165    SIGNOR:SIGNOR-ST9            ER stress  Stimulus   yes\n",
      "3915   SIGNOR:SIGNOR-ST13    Cell-Cell_contact  Stimulus    no\n",
      "4478   SIGNOR:SIGNOR-ST21          Viral_dsRNA  Stimulus    no\n",
      "4799   SIGNOR:SIGNOR-ST12                  GFs  Stimulus    no\n",
      "6148   SIGNOR:SIGNOR-ST18                DAMPS  Stimulus    no\n",
      "7232   SIGNOR:SIGNOR-ST15     Immune complexes  Stimulus    no\n",
      "9799   SIGNOR:SIGNOR-ST16   Ionizing radiation  Stimulus   yes\n",
      "10712  SIGNOR:SIGNOR-ST29                Z-RNA  Stimulus    no\n",
      "10732  SIGNOR:SIGNOR-ST25              Hypoxia  Stimulus   yes\n",
      "13382   SIGNOR:SIGNOR-ST8     Survival Factors  Stimulus    no\n",
      "17541  SIGNOR:SIGNOR-ST19           Fatty acid  Stimulus    no\n",
      "19365   SIGNOR:SIGNOR-ST4           Starvation  Stimulus    no\n",
      "21010  SIGNOR:SIGNOR-ST30        Cytosolic DNA  Stimulus    no\n",
      "26974  SIGNOR:SIGNOR-ST23    Viral_replication  Stimulus   yes\n",
      "28248   SIGNOR:SIGNOR-ST3                Serum  Stimulus    no\n",
      "31072  SIGNOR:SIGNOR-ST27      Sound_vibration  Stimulus    no\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_79386/4283732433.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Stimulus_node_df['label'] = Stimulus_node_df.apply(Stimulus_row_checker, axis=1)\n"
     ]
    }
   ],
   "source": [
    "## apply to each row of the dataframe\n",
    "Stimulus_node_df['label'] = Stimulus_node_df.apply(Stimulus_row_checker, axis=1)\n",
    "print(Stimulus_node_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8eef8978-551e-4d18-9f67-d58344c9a229",
   "metadata": {},
   "source": [
    "### LLM flagged -> manual check\n",
    "* Ionizing radiation: high-energy radiation capable of removing electrons from atoms, which can cause DNA damage, protein oxidation, and cellular stress.\n",
    "* Hypoxia means a deficiency in oxygen levels, it activates a variety of cellular pathways\n",
    "* in the long term we want a separate stimulus category in Biolink\n",
    "* so for now, **removing all stimulus related edges**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c03004a3-b17c-47b3-985d-15f2e1d364d2",
   "metadata": {},
   "source": [
    "## Now check Fusion protein\n",
    "* "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "08d245f5-0f78-46cd-beff-03d3794b5518",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15\n"
     ]
    }
   ],
   "source": [
    "## now try to check fusion_protein\n",
    "fusion_protein_node_df = concat_node_df[concat_node_df['category'] == 'Fusion Protein']\n",
    "print(len(fusion_protein_node_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "3a99ee0e-36d7-4100-97b7-621ffdab27d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      id           name        category\n",
      "1282   SIGNOR:SIGNOR-FP1       AML1-ETO  Fusion Protein\n",
      "1283   SIGNOR:SIGNOR-FP2   PML-RARalpha  Fusion Protein\n",
      "1331  SIGNOR:SIGNOR-FP16   NUP98 Fusion  Fusion Protein\n",
      "1674   SIGNOR:SIGNOR-FP6        BCR-ABL  Fusion Protein\n",
      "1891   SIGNOR:SIGNOR-FP3  CBFbeta-MYH11  Fusion Protein\n"
     ]
    }
   ],
   "source": [
    "print(fusion_protein_node_df.head(5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "4cd59a03-f56d-41e0-9e2a-68e491109da8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## run the LLM model using a custom query\n",
    "# Define the row labeling function\n",
    "def fusion_protein_row_checker(row):\n",
    "    combined_text = f\"{row['name']} {row['id']}\"\n",
    "    prompt = f\"\"\"Analyze if the name is an existing fusion protein, answer only in \"yes\" if satisfied, or \"no\" if incorrect:\n",
    "\\\"\\\"\\\"{combined_text}\\\"\\\"\\\"\n",
    "Sentiment:\"\"\"\n",
    "    return query_ollama(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "00024fd0-419a-4675-9623-765b6fe16bee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       id          name        category label\n",
      "1331   SIGNOR:SIGNOR-FP16  NUP98 Fusion  Fusion Protein    no\n",
      "1674    SIGNOR:SIGNOR-FP6       BCR-ABL  Fusion Protein    no\n",
      "2593   SIGNOR:SIGNOR-FP14    MLL Fusion  Fusion Protein    no\n",
      "8434    SIGNOR:SIGNOR-FP5       MLL-AF9  Fusion Protein    no\n",
      "19983  SIGNOR:SIGNOR-FP10      ELE1-RET  Fusion Protein    no\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_79386/2359603948.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  fusion_protein_node_df['label'] = fusion_protein_node_df.apply(fusion_protein_row_checker, axis=1)\n"
     ]
    }
   ],
   "source": [
    "## apply to each row of the dataframe\n",
    "fusion_protein_node_df['label'] = fusion_protein_node_df.apply(fusion_protein_row_checker, axis=1)\n",
    "print(fusion_protein_node_df[fusion_protein_node_df['label'] == 'no'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2ce8acc-5e71-48c8-8fd8-7167600d3cc6",
   "metadata": {},
   "source": [
    "### LLM flagged -> manual check\n",
    "* used FusionPub to manually verify: https://compbio.uth.edu/FusionPub/\n",
    "    * NUP98 fusion found 58 pairs\n",
    "    * only BCR-ABL1 fusion found\n",
    "    * MLL fusion found 34 pairs\n",
    "    * MLL-AF9 **not found**\n",
    "    * ELE1-RET **not found**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf27fd48-019d-4150-b289-e841bfa5cb42",
   "metadata": {},
   "source": [
    "# Following codes are quality control steps and used to keep track of which changes made to source file\n",
    "* Remove all rows with 'Fusion Protein' and 'Stimulus' for now\n",
    "* Antibody all passed LLM test\n",
    "* Remove those complex flagged by LLM\n",
    "* replace miR-34 to miR-34a, which is the correct name for the identifier\n",
    "* ncrna all passed LLM test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b11c2664-6a23-44c6-8356-104cbbcab8e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total row number before quality control is:  37051\n"
     ]
    }
   ],
   "source": [
    "print(\"total row number before quality control is: \", len(source_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5c0e7ae3-4358-4535-be94-0c334067b706",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "remaining row number is:  37051\n"
     ]
    }
   ],
   "source": [
    "## remove those rows with col(predicate) == 'unknown' from source_df\n",
    "source_df = source_df[ (source_df['subject_category'] != 'Fusion Protein') & (source_df['object_category'] != 'Fusion Protein')]\n",
    "source_df = source_df[ (source_df['subject_category'] != 'Stimulus') & (source_df['object_category'] != 'Stimulus')]\n",
    "\n",
    "print(\"remaining row number is: \", len(source_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "af1c7abc-42a3-441e-8597-800dd0c0abfc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "remaining row number is:  37051\n"
     ]
    }
   ],
   "source": [
    "## obtain all complex names need to be removed \n",
    "flagged_comple_df = complex_node_df[complex_node_df['label'] == 'no']\n",
    "flagged_complex_names = flagged_comple_df['name'].tolist()\n",
    "# print(flagged_complex_names)\n",
    "\n",
    "## remove those rows with either subject_name or object_name in those flagged complex names\n",
    "source_df = source_df[\n",
    "    (~source_df['subject_name'].isin(flagged_complex_names)) &\n",
    "    (~source_df['object_name'].isin(flagged_complex_names))\n",
    "]\n",
    "\n",
    "print(\"remaining row number is: \", len(source_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d84b23ce-77a5-4309-8a17-9b625306df33",
   "metadata": {},
   "outputs": [],
   "source": [
    "## replace all 'miR-34' to 'miR-34a' in two columns subject_category and object_category in the pandas dataframe\n",
    "source_df['subject_name'] = source_df['subject_name'].replace('miR-34', 'miR-34a')\n",
    "source_df['object_name'] = source_df['object_name'].replace('miR-34', 'miR-34a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "1f86078b-d716-4476-9cb0-368252bd9f9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  subject_name      predicate object_name\n",
      "0      CSNK2A1    upregulates     GTF2A1L\n",
      "1        GSK3B  downregulates        SPI1\n",
      "2       PRKACA  downregulates       MARK2\n",
      "3        PTPRG  downregulates         SRC\n",
      "4        PRKG1    upregulates       FHOD1\n"
     ]
    }
   ],
   "source": [
    "## now check the source_df again with selected columns\n",
    "source_edge_pair_df = source_df[['subject_name', 'predicate', 'object_name']]\n",
    "\n",
    "print(source_edge_pair_df.head(5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a56f37f1-88c6-4a8d-a2e4-cb9435c9f311",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20\n"
     ]
    }
   ],
   "source": [
    "temp_df = source_edge_pair_df.head(20)\n",
    "print(len(temp_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d474d7e7-6ce8-43f2-9eaf-c8b576a38fe8",
   "metadata": {},
   "source": [
    "## Now run a similar LLM to check and see if the pair/edge makes sense\n",
    "* flag anything doesnt seem to be a vaild edge during to **biological meaning not satisfied**\n",
    "*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "19e6ce73-5636-4bfd-b761-16c51f55ead5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## run the LLM model using a custom query\n",
    "# Define the row labeling function\n",
    "def edge_pair_row_checker(row):\n",
    "    combined_text = f\"{row['subject_name']} {row['predicate']} {row['object_name']}\"\n",
    "    prompt = f\"\"\"Analyze whether the following represents a biologically valid subject → predicate → object relationship, \n",
    "                no known literature or no evidence indicates or no studies report or no literature support alone are not sufficient reasons for 'no',\n",
    "                more specifically you need to find if there are evidence or literature showing opposite or different mechanismes. \n",
    "Reply strictly in this format:\n",
    "\n",
    "Label: yes or no\n",
    "Reason: within 30 words\n",
    "\n",
    "\\\"\\\"\\\"{combined_text}\\\"\\\"\\\"\n",
    "\"\"\"\n",
    "    response = query_ollama(prompt)\n",
    "\n",
    "    # Parse the response\n",
    "    label = None\n",
    "    reason = None\n",
    "    for line in response.splitlines():\n",
    "        if line.lower().startswith(\"label:\"):\n",
    "            label = line.split(\":\", 1)[1].strip().lower()\n",
    "        elif line.lower().startswith(\"reason:\"):\n",
    "            reason = line.split(\":\", 1)[1].strip()\n",
    "\n",
    "    return pd.Series({\n",
    "        \"label\": label,\n",
    "        \"reason\": reason\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "687a252b-cbaa-43a9-b098-e36ba7fcbffa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 10%|████████▌                                                                             | 2/20 [00:29<04:27, 14.85s/it]"
     ]
    }
   ],
   "source": [
    "## apply to each row of the dataframe\n",
    "temp_df[['label', 'reason']] = temp_df.progress_apply(edge_pair_row_checker, axis=1)\n",
    "print(temp_df[temp_df['label'] == 'no'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "1bd37968-3dde-4f91-bc05-3c3e369c8c1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     subject_name      predicate      object_name label  \\\n",
      "0         CSNK2A1    upregulates          GTF2A1L    no   \n",
      "1           GSK3B  downregulates             SPI1    no   \n",
      "2          PRKACA  downregulates            MARK2    no   \n",
      "4           PRKG1    upregulates            FHOD1    no   \n",
      "5           MAPK3  downregulates            SHOC2    no   \n",
      "7           HNF4A    upregulates          C1QTNF5    no   \n",
      "9   Sin3B_complex  downregulates            H3C13    no   \n",
      "11      ESCRT-III    upregulates  Membrane_fusion    no   \n",
      "12         MAP2K4    upregulates           MAP2K4    no   \n",
      "14          F2RL3    upregulates             GNAZ    no   \n",
      "17         PCDH15    upregulates             TMC2    no   \n",
      "19           DOK1  downregulates            ITGB4    no   \n",
      "\n",
      "                                               reason  \n",
      "0   No literature reports CSNK2A1 activating GTF2A...  \n",
      "1   No published evidence shows GSK3B downregulate...  \n",
      "2   No literature reports PRKACA downregulating MA...  \n",
      "4   No literature reports PRKG1 upregulating FHOD1...  \n",
      "5   No literature shows MAPK3 reduces SHOC2; SHOC2...  \n",
      "7   No studies report HNF4A activating C1QTNF5; ev...  \n",
      "9   No evidence indicates Sin3B directly downregul...  \n",
      "11  No literature demonstrates ESCRT‑III promotes ...  \n",
      "12  MAP2K4 is activated by upstream MAP3Ks, not kn...  \n",
      "14  No peer‑reviewed study reports that F2RL3 expr...  \n",
      "17  No studies report PCDH15 upregulating TMC2; li...  \n",
      "19  No published studies show DOK1 directly downre...  \n"
     ]
    }
   ],
   "source": [
    "print(temp_df[temp_df['label'] == 'no'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "9e8d028d-df75-4359-acfb-328870be1c93",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Sample 2000 rows randomly\n",
    "sampled_df = source_edge_pair_df.sample(n=3000, random_state=1984)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "5046c9b2-d465-48cc-9d0a-7659fc849c16",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████| 2000/2000 [9:13:22<00:00, 16.60s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1119\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "## apply to sampled rows\n",
    "\n",
    "## apply to each row of the dataframe\n",
    "# source_edge_pair_df[['label', 'reason']] = source_edge_pair_df.apply(edge_pair_row_checker, axis=1)\n",
    "\n",
    "sampled_df[['label', 'reason']] = sampled_df.progress_apply(edge_pair_row_checker, axis=1)\n",
    "print(len(sampled_df[sampled_df['label'] == 'no']))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "397959e4-954a-4820-9ef9-408a8050d2bf",
   "metadata": {},
   "source": [
    "## Manual check to spot any obvious pairs with issues\n",
    "* Here starts the list of spotted issues:\n",
    "* Structure: Subject, object, predicate, publications\n",
    "* 1. self regulates vs indirect positive feedback loop\n",
    "* MAP2K4\tMAP2K4\t\tupregulates SIGNOR-251420\tPMID:36243968|PMID:9162092\n",
    "* Coexpression of MEKK2 or MEKK3 with MKK4 in COS-7 cells resulted in activation of MKK4\n",
    "* source: https://www.sciencedirect.com/science/article/pii/S0021925819625767?via%3Dihub\n",
    "*\n",
    "* TEK TEK upregulates PMID:36243968|PMID:11513602\n",
    "* FGFR1 FGFR1 upregulates PMID:36243968|PMID:8622701\n",
    "*\n",
    "* 2\n",
    "* Duplicates rows of PRKCA -> CSNK1D\n",
    "* 3\n",
    "* Pairs that are unlikely\n",
    "* Phenotype Gene upregulates\n",
    "* Phenotypes don't regulate genes; they are outcomes.\n",
    "* Gene Phenotype downregulates A gene doesn't downregulate a phenotype. It can cause it\n",
    "* Phenotype\n",
    "Proteinfamily\n",
    "downregulates\n",
    "Phenotype is not a regulatory entity.\n",
    "* Phenotype\n",
    "Complex\n",
    "downregulates / upregulates\n",
    "Same reasoning — phenotype isn't active agent.\n",
    "* Phenotype\n",
    "Phenotype\n",
    "downregulates\n",
    "Conceptually invalid; phenotypes don’t regulate each other.\n",
    "* 4\n",
    "* 5\n",
    "* 6\n",
    "* 7\n",
    "* 8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d0f14468-8289-48c7-b32e-2dccaa682c8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_identifier</th>\n",
       "      <th>subject_name</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_identifier</th>\n",
       "      <th>object_name</th>\n",
       "      <th>object_category</th>\n",
       "      <th>predicate</th>\n",
       "      <th>original_predicate</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>Primary_Knowledge_Source</th>\n",
       "      <th>publications</th>\n",
       "      <th>knowledge_level</th>\n",
       "      <th>anatomical_context_qualifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:1457</td>\n",
       "      <td>CSNK2A1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene:11036</td>\n",
       "      <td>GTF2A1L</td>\n",
       "      <td>Gene</td>\n",
       "      <td>upregulates</td>\n",
       "      <td>up-regulates activity</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-250870</td>\n",
       "      <td>PMID:36243968|PMID:12107178</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBIGene:2932</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NCBIGene:6688</td>\n",
       "      <td>SPI1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>downregulates</td>\n",
       "      <td>down-regulates quantity by destabilization</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-277542</td>\n",
       "      <td>PMID:36243968|PMID:33188146</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_identifier subject_name subject_category object_identifier  \\\n",
       "0      NCBIGene:1457      CSNK2A1             Gene    NCBIGene:11036   \n",
       "1      NCBIGene:2932        GSK3B             Gene     NCBIGene:6688   \n",
       "\n",
       "  object_name object_category      predicate  \\\n",
       "0     GTF2A1L            Gene    upregulates   \n",
       "1        SPI1            Gene  downregulates   \n",
       "\n",
       "                           original_predicate provided_by  \\\n",
       "0                       up-regulates activity      SIGNOR   \n",
       "1  down-regulates quantity by destabilization      SIGNOR   \n",
       "\n",
       "  Primary_Knowledge_Source                 publications      knowledge_level  \\\n",
       "0            SIGNOR-250870  PMID:36243968|PMID:12107178  knowledge_assertion   \n",
       "1            SIGNOR-277542  PMID:36243968|PMID:33188146  knowledge_assertion   \n",
       "\n",
       "  anatomical_context_qualifier  \n",
       "0                          NaN  \n",
       "1                          NaN  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "bcce4495-dea9-4102-bb06-9dba158bb8ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Row counts after LLM checking\n",
      "37051\n",
      "Row counts after removing obvious issues\n",
      "32160\n"
     ]
    }
   ],
   "source": [
    "print(\"Row counts after LLM checking\")\n",
    "print(len(source_df))\n",
    "\n",
    "## Now remove those obvious errors\n",
    "# Drop rows where subject_name == object_name\n",
    "source_df_filtered = source_df[source_df['subject_name'] != source_df['object_name']]\n",
    "\n",
    "## rename a column from predicate to relationship, notice!\n",
    "source_df_filtered = source_df_filtered.rename(columns={'predicate': 'relationship'})\n",
    "\n",
    "## rename columns subject_category and object_category to subject and object\n",
    "source_df_filtered = source_df_filtered.rename(columns={'subject_category': 'subject'})\n",
    "source_df_filtered = source_df_filtered.rename(columns={'object_category': 'object'})\n",
    "\n",
    "## mapping the subject_category and object_category \n",
    "source_df_filtered['subject_category'] = (\n",
    "    source_df_filtered['subject'].map(category_map)\n",
    ")\n",
    "\n",
    "## mapping the subject_category and object_category \n",
    "source_df_filtered['object_category'] = (\n",
    "    source_df_filtered['object'].map(category_map)\n",
    ")\n",
    "\n",
    "## match only combinations allowed in the pair dictionary\n",
    "## Apply the mapping and return a Series with 3 columns\n",
    "source_df_filtered[['predicate', 'subject_direction_qualifier', 'object_direction_qualifier']] = source_df_filtered.apply(\n",
    "    lambda row: pd.Series(\n",
    "        predicate_map.get(\n",
    "            (row['subject_category'], row['object_category'], row['relationship']),\n",
    "            [None, None, None]  # Default if not found\n",
    "        )\n",
    "    ),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "## drop columns no longer needed\n",
    "source_df_filtered = source_df_filtered.drop(columns=['subject', 'object', 'relationship', 'original_predicate'])\n",
    "\n",
    "## Drop those unmatched rows\n",
    "## Drop rows where 'name' is NaN, None, or empty string\n",
    "source_df_filtered = source_df_filtered[~source_df_filtered['predicate'].isna() & (source_df_filtered['predicate'].str.strip() != '')]\n",
    "\n",
    "\n",
    "print(\"Row counts after removing obvious issues\")\n",
    "print(len(source_df_filtered))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "dd04f4cf-18d0-49fc-9876-d8fc53799d0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    subject_category                object_category  \\\n",
      "0                       biolink:Gene                   biolink:Gene   \n",
      "9      biolink:MacromolecularComplex                   biolink:Gene   \n",
      "11     biolink:MacromolecularComplex      biolink:PhenotypicFeature   \n",
      "15            biolink:ChemicalEntity                   biolink:Gene   \n",
      "37                   biolink:Protein                   biolink:Gene   \n",
      "47                      biolink:Gene  biolink:MacromolecularComplex   \n",
      "60             biolink:SmallMolecule                   biolink:Gene   \n",
      "64             biolink:SmallMolecule          biolink:SmallMolecule   \n",
      "86             biolink:SmallMolecule          biolink:ProteinFamily   \n",
      "109            biolink:ProteinFamily                   biolink:Gene   \n",
      "165                     biolink:Gene      biolink:PhenotypicFeature   \n",
      "298    biolink:MacromolecularComplex  biolink:MacromolecularComplex   \n",
      "308                     biolink:Gene                biolink:Protein   \n",
      "492            biolink:ProteinFamily          biolink:ProteinFamily   \n",
      "552            biolink:ProteinFamily  biolink:MacromolecularComplex   \n",
      "557           biolink:ChemicalEntity  biolink:MacromolecularComplex   \n",
      "615            biolink:SmallMolecule  biolink:MacromolecularComplex   \n",
      "680    biolink:MacromolecularComplex                biolink:Protein   \n",
      "768    biolink:MacromolecularComplex          biolink:ProteinFamily   \n",
      "840            biolink:ProteinFamily  biolink:MacromolecularComplex   \n",
      "1166   biolink:MacromolecularComplex  biolink:MacromolecularComplex   \n",
      "1332           biolink:SmallMolecule  biolink:MacromolecularComplex   \n",
      "1348          biolink:ChemicalEntity          biolink:ProteinFamily   \n",
      "1642                biolink:MicroRNA                   biolink:Gene   \n",
      "1932           biolink:ProteinFamily      biolink:PhenotypicFeature   \n",
      "2358           biolink:SmallMolecule      biolink:PhenotypicFeature   \n",
      "3156                    biolink:Drug                   biolink:Gene   \n",
      "3450                 biolink:Protein  biolink:MacromolecularComplex   \n",
      "4502                 biolink:Protein  biolink:MacromolecularComplex   \n",
      "4968    biolink:Noncoding_RNAProduct                   biolink:Gene   \n",
      "6553                 biolink:Protein          biolink:ProteinFamily   \n",
      "7529           biolink:ProteinFamily                biolink:Protein   \n",
      "7863          biolink:ChemicalEntity  biolink:MacromolecularComplex   \n",
      "11883          biolink:SmallMolecule                biolink:Protein   \n",
      "12565               biolink:MicroRNA      biolink:PhenotypicFeature   \n",
      "15569          biolink:ProteinFamily                biolink:Protein   \n",
      "16209                biolink:Protein      biolink:PhenotypicFeature   \n",
      "22676                biolink:Protein          biolink:ProteinFamily   \n",
      "\n",
      "                     predicate  \n",
      "0            biolink:regulates  \n",
      "9              biolink:affects  \n",
      "11             biolink:affects  \n",
      "15             biolink:affects  \n",
      "37           biolink:regulates  \n",
      "47             biolink:affects  \n",
      "60             biolink:affects  \n",
      "64             biolink:affects  \n",
      "86             biolink:affects  \n",
      "109          biolink:regulates  \n",
      "165            biolink:affects  \n",
      "298            biolink:affects  \n",
      "308          biolink:regulates  \n",
      "492          biolink:regulates  \n",
      "552            biolink:affects  \n",
      "557            biolink:affects  \n",
      "615            biolink:affects  \n",
      "680            biolink:affects  \n",
      "768            biolink:affects  \n",
      "840    biolink:in_complex_with  \n",
      "1166   biolink:in_complex_with  \n",
      "1332   biolink:in_complex_with  \n",
      "1348           biolink:affects  \n",
      "1642           biolink:affects  \n",
      "1932           biolink:affects  \n",
      "2358           biolink:affects  \n",
      "3156           biolink:affects  \n",
      "3450           biolink:affects  \n",
      "4502   biolink:in_complex_with  \n",
      "4968           biolink:affects  \n",
      "6553           biolink:affects  \n",
      "7529         biolink:regulates  \n",
      "7863   biolink:in_complex_with  \n",
      "11883          biolink:affects  \n",
      "12565          biolink:affects  \n",
      "15569          biolink:affects  \n",
      "16209          biolink:affects  \n",
      "22676        biolink:regulates  \n"
     ]
    }
   ],
   "source": [
    "## check unique combination again\n",
    "## check all unique combinations between subject & object\n",
    "from itertools import combinations\n",
    "\n",
    "# Get all combinations of 3 column names\n",
    "col_combos = ['subject_category', 'object_category', 'predicate']\n",
    "\n",
    "unique_combinations = source_df_filtered[col_combos].drop_duplicates()\n",
    "\n",
    "print(unique_combinations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "1c745012-08e0-4d9d-98d6-3add13804a9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_identifier</th>\n",
       "      <th>subject_name</th>\n",
       "      <th>object_identifier</th>\n",
       "      <th>object_name</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>Primary_Knowledge_Source</th>\n",
       "      <th>publications</th>\n",
       "      <th>knowledge_level</th>\n",
       "      <th>anatomical_context_qualifier</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_category</th>\n",
       "      <th>predicate</th>\n",
       "      <th>subject_direction_qualifier</th>\n",
       "      <th>object_direction_qualifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:1457</td>\n",
       "      <td>CSNK2A1</td>\n",
       "      <td>NCBIGene:11036</td>\n",
       "      <td>GTF2A1L</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-250870</td>\n",
       "      <td>PMID:36243968|PMID:12107178</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:regulates</td>\n",
       "      <td>None</td>\n",
       "      <td>upregulated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBIGene:2932</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>NCBIGene:6688</td>\n",
       "      <td>SPI1</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-277542</td>\n",
       "      <td>PMID:36243968|PMID:33188146</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:regulates</td>\n",
       "      <td>None</td>\n",
       "      <td>downregulated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBIGene:5566</td>\n",
       "      <td>PRKACA</td>\n",
       "      <td>NCBIGene:2011</td>\n",
       "      <td>MARK2</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-276870</td>\n",
       "      <td>PMID:36243968|PMID:25512381</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:regulates</td>\n",
       "      <td>None</td>\n",
       "      <td>downregulated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBIGene:5793</td>\n",
       "      <td>PTPRG</td>\n",
       "      <td>NCBIGene:6714</td>\n",
       "      <td>SRC</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-254725</td>\n",
       "      <td>PMID:36243968|PMID:25624455</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:regulates</td>\n",
       "      <td>None</td>\n",
       "      <td>downregulated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBIGene:5592</td>\n",
       "      <td>PRKG1</td>\n",
       "      <td>NCBIGene:29109</td>\n",
       "      <td>FHOD1</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-170094</td>\n",
       "      <td>PMID:36243968|PMID:21106951</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>BTO:0000887;BTO:0001260</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:regulates</td>\n",
       "      <td>None</td>\n",
       "      <td>upregulated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_identifier subject_name object_identifier object_name provided_by  \\\n",
       "0      NCBIGene:1457      CSNK2A1    NCBIGene:11036     GTF2A1L      SIGNOR   \n",
       "1      NCBIGene:2932        GSK3B     NCBIGene:6688        SPI1      SIGNOR   \n",
       "2      NCBIGene:5566       PRKACA     NCBIGene:2011       MARK2      SIGNOR   \n",
       "3      NCBIGene:5793        PTPRG     NCBIGene:6714         SRC      SIGNOR   \n",
       "4      NCBIGene:5592        PRKG1    NCBIGene:29109       FHOD1      SIGNOR   \n",
       "\n",
       "  Primary_Knowledge_Source                 publications      knowledge_level  \\\n",
       "0            SIGNOR-250870  PMID:36243968|PMID:12107178  knowledge_assertion   \n",
       "1            SIGNOR-277542  PMID:36243968|PMID:33188146  knowledge_assertion   \n",
       "2            SIGNOR-276870  PMID:36243968|PMID:25512381  knowledge_assertion   \n",
       "3            SIGNOR-254725  PMID:36243968|PMID:25624455  knowledge_assertion   \n",
       "4            SIGNOR-170094  PMID:36243968|PMID:21106951  knowledge_assertion   \n",
       "\n",
       "  anatomical_context_qualifier subject_category object_category  \\\n",
       "0                          NaN     biolink:Gene    biolink:Gene   \n",
       "1                          NaN     biolink:Gene    biolink:Gene   \n",
       "2                          NaN     biolink:Gene    biolink:Gene   \n",
       "3                          NaN     biolink:Gene    biolink:Gene   \n",
       "4      BTO:0000887;BTO:0001260     biolink:Gene    biolink:Gene   \n",
       "\n",
       "           predicate subject_direction_qualifier object_direction_qualifier  \n",
       "0  biolink:regulates                        None                upregulated  \n",
       "1  biolink:regulates                        None              downregulated  \n",
       "2  biolink:regulates                        None              downregulated  \n",
       "3  biolink:regulates                        None              downregulated  \n",
       "4  biolink:regulates                        None                upregulated  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source_df_filtered.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "16842711-c827-4979-b046-a6b9e8c1b4fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['subject_identifier', 'subject_name', 'object_identifier', 'object_name', 'provided_by', 'Primary_Knowledge_Source', 'publications', 'knowledge_level', 'anatomical_context_qualifier', 'subject_category', 'object_category', 'predicate', 'subject_direction_qualifier', 'object_direction_qualifier']\n"
     ]
    }
   ],
   "source": [
    "# Print all column names\n",
    "print(source_df_filtered.columns.tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7661d587-116d-4171-811b-6e9dc3472593",
   "metadata": {},
   "source": [
    "### finalize the final edge file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "ca5dc6a2-0088-4c32-a8d7-2b3a22c1a63f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['subject_identifier', 'subject_name', 'object_identifier', 'object_name', 'provided_by', 'Primary_Knowledge_Source', 'publications', 'knowledge_level', 'anatomical_context_qualifier', 'subject_category', 'object_category', 'predicate', 'subject_direction_qualifier', 'object_direction_qualifier']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>subject_name</th>\n",
       "      <th>object</th>\n",
       "      <th>object_name</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>knowledge_source</th>\n",
       "      <th>publications</th>\n",
       "      <th>knowledge_level</th>\n",
       "      <th>context_qualifier</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_category</th>\n",
       "      <th>predicate</th>\n",
       "      <th>subject_direction_qualifier</th>\n",
       "      <th>object_direction_qualifier</th>\n",
       "      <th>agent_type</th>\n",
       "      <th>deploy_date</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:1457</td>\n",
       "      <td>CSNK2A1</td>\n",
       "      <td>NCBIGene:11036</td>\n",
       "      <td>GTF2A1L</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-250870</td>\n",
       "      <td>PMID:36243968|PMID:12107178</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:regulates</td>\n",
       "      <td>None</td>\n",
       "      <td>upregulated</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-08-18</td>\n",
       "      <td>d8c53d1b-7e3d-53ee-965c-e60523bb3f4a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBIGene:2932</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>NCBIGene:6688</td>\n",
       "      <td>SPI1</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-277542</td>\n",
       "      <td>PMID:36243968|PMID:33188146</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:regulates</td>\n",
       "      <td>None</td>\n",
       "      <td>downregulated</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-08-18</td>\n",
       "      <td>32a7bd45-71ac-5b11-9cdb-f40329429aa6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBIGene:5566</td>\n",
       "      <td>PRKACA</td>\n",
       "      <td>NCBIGene:2011</td>\n",
       "      <td>MARK2</td>\n",
       "      <td>SIGNOR</td>\n",
       "      <td>SIGNOR-276870</td>\n",
       "      <td>PMID:36243968|PMID:25512381</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:regulates</td>\n",
       "      <td>None</td>\n",
       "      <td>downregulated</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-08-18</td>\n",
       "      <td>3cd03f76-b82b-57a2-b773-62f8538a9d28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         subject subject_name          object object_name provided_by  \\\n",
       "0  NCBIGene:1457      CSNK2A1  NCBIGene:11036     GTF2A1L      SIGNOR   \n",
       "1  NCBIGene:2932        GSK3B   NCBIGene:6688        SPI1      SIGNOR   \n",
       "2  NCBIGene:5566       PRKACA   NCBIGene:2011       MARK2      SIGNOR   \n",
       "\n",
       "  knowledge_source                 publications      knowledge_level  \\\n",
       "0    SIGNOR-250870  PMID:36243968|PMID:12107178  knowledge_assertion   \n",
       "1    SIGNOR-277542  PMID:36243968|PMID:33188146  knowledge_assertion   \n",
       "2    SIGNOR-276870  PMID:36243968|PMID:25512381  knowledge_assertion   \n",
       "\n",
       "  context_qualifier subject_category object_category          predicate  \\\n",
       "0               NaN     biolink:Gene    biolink:Gene  biolink:regulates   \n",
       "1               NaN     biolink:Gene    biolink:Gene  biolink:regulates   \n",
       "2               NaN     biolink:Gene    biolink:Gene  biolink:regulates   \n",
       "\n",
       "  subject_direction_qualifier object_direction_qualifier       agent_type  \\\n",
       "0                        None                upregulated  automated_agent   \n",
       "1                        None              downregulated  automated_agent   \n",
       "2                        None              downregulated  automated_agent   \n",
       "\n",
       "  deploy_date                                    id  \n",
       "0  2025-08-18  d8c53d1b-7e3d-53ee-965c-e60523bb3f4a  \n",
       "1  2025-08-18  32a7bd45-71ac-5b11-9cdb-f40329429aa6  \n",
       "2  2025-08-18  3cd03f76-b82b-57a2-b773-62f8538a9d28  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## copy to a final edge df\n",
    "edge_df = source_df_filtered.copy(deep = True)\n",
    "\n",
    "# Print all column names\n",
    "# print(edge_df.columns.tolist())\n",
    "\n",
    "## rename columns to desired parsed format\n",
    "edge_df = edge_df.rename(columns={'Primary_Knowledge_Source': 'knowledge_source'})\n",
    "edge_df = edge_df.rename(columns={'anatomical_context_qualifier': 'context_qualifier'})\n",
    "edge_df = edge_df.rename(columns={'subject_identifier': 'subject'})\n",
    "edge_df = edge_df.rename(columns={'object_identifier': 'object'})\n",
    "\n",
    "## add a new agent_type column and set value to be 'manual_agent'\n",
    "edge_df['agent_type'] = 'automated_agent'\n",
    "\n",
    "## add a deploy date column\n",
    "edge_df['deploy_date'] = deployment_date\n",
    "\n",
    "### Add resources_id column, checking whether edge is already\n",
    "column_list = ['subject', 'predicate', 'object', 'context_qualifier', 'deploy_date']\n",
    "# Apply the function to each row to generate UUIDs\n",
    "edge_df['id'] = edge_df[column_list].apply(generate_uuid, axis=1)\n",
    "\n",
    "## print head to check\n",
    "edge_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "2354c035-fae4-4926-8ac1-4da0dc6a34e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['subject', 'subject_name', 'subject_category', 'object', 'object_name', 'object_category', 'predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'provided_by', 'knowledge_source', 'publications', 'knowledge_level', 'context_qualifier', 'agent_type', 'deploy_date', 'id']\n"
     ]
    }
   ],
   "source": [
    "## reorder the dataframe\n",
    "desired_order = [\n",
    "    'subject', 'subject_name', 'subject_category', 'object', 'object_name', 'object_category', \n",
    "    'predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'provided_by', \n",
    "    'knowledge_source', 'publications', 'knowledge_level', 'context_qualifier', 'agent_type', \n",
    "    'deploy_date', 'id'\n",
    "]\n",
    "\n",
    "edge_df = edge_df[desired_order]\n",
    "\n",
    "print(edge_df.columns.tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd7d2172-0620-488b-aadd-f2eac9cf52f6",
   "metadata": {},
   "source": [
    "### Obtain the corresponding node file from the final edge file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "472893d7-23b3-425c-89a9-8c5ffeb7f7ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64320\n",
      "8666\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_17658/2150963183.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  signor_genes_node_subject_df.rename(columns={'subject': 'id', 'subject_name': 'name',\n",
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_17658/2150963183.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  signor_genes_node_object_df.rename(columns={'object': 'id', 'object_name': 'name',\n"
     ]
    }
   ],
   "source": [
    "signor_genes_node_subject_df = edge_df[['subject', 'subject_name', 'subject_category']]\n",
    "signor_genes_node_object_df = edge_df[['object', 'object_name', 'object_category']]\n",
    "\n",
    "## rename those columns into desired format\n",
    "signor_genes_node_subject_df.rename(columns={'subject': 'id', 'subject_name': 'name',\n",
    "                                                 'subject_category': 'category'}, inplace=True)\n",
    "signor_genes_node_object_df.rename(columns={'object': 'id', 'object_name': 'name',\n",
    "                                                 'object_category': 'category'}, inplace=True)\n",
    "\n",
    "## vertical concatenation\n",
    "signor_genes_node_df = pd.concat([signor_genes_node_subject_df, signor_genes_node_object_df])\n",
    "\n",
    "## check # of total nodes\n",
    "print(len(signor_genes_node_df))\n",
    "\n",
    "# Drop duplicate rows\n",
    "signor_genes_node_df = signor_genes_node_df.drop_duplicates()\n",
    "\n",
    "## check # of total nodes\n",
    "print(len(signor_genes_node_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4407ba38-ed43-4cdd-8ff9-dfefd7eca032",
   "metadata": {},
   "source": [
    "## following codes are used for quality control and sanity check\n",
    "* check and confirm all subject & object types are correctly formatted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "12966b0f-b7b0-4de5-9c42-bd9c7e35f469",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "category\n",
      "biolink:Gene                     6375\n",
      "biolink:ChemicalEntity            992\n",
      "biolink:SmallMolecule             418\n",
      "biolink:MacromolecularComplex     416\n",
      "biolink:PhenotypicFeature         206\n",
      "biolink:Protein                   126\n",
      "biolink:ProteinFamily              86\n",
      "biolink:Drug                       25\n",
      "biolink:MicroRNA                   21\n",
      "biolink:Noncoding_RNAProduct        1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = signor_genes_node_df['category'].value_counts()\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "91d059b1-b8e5-48cb-8ec1-3de0c5dc68a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "subject_category\n",
      "biolink:Gene                     24401\n",
      "biolink:MacromolecularComplex     2540\n",
      "biolink:ChemicalEntity            2501\n",
      "biolink:ProteinFamily             1334\n",
      "biolink:SmallMolecule             1079\n",
      "biolink:Protein                    231\n",
      "biolink:MicroRNA                    42\n",
      "biolink:Drug                        29\n",
      "biolink:Noncoding_RNAProduct         3\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = edge_df['subject_category'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "85939723-1fa8-4f82-8ed2-ba9abd9ba711",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "object_category\n",
      "biolink:Gene                     28651\n",
      "biolink:PhenotypicFeature         1550\n",
      "biolink:MacromolecularComplex     1117\n",
      "biolink:SmallMolecule              424\n",
      "biolink:ProteinFamily              222\n",
      "biolink:Protein                    196\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = edge_df['object_category'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "f72d3c9d-db7c-47ef-90b9-6244dc1d54cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "predicate                          0\n",
      "subject_direction_qualifier    32160\n",
      "object_direction_qualifier       109\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check missingness\n",
    "print(edge_df[['predicate', 'subject_direction_qualifier', 'object_direction_qualifier']].isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "fba1a6c5-9eee-4162-9824-0a7fad283b9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 predicate subject_direction_qualifier  \\\n",
      "0        biolink:regulates                        None   \n",
      "1        biolink:regulates                        None   \n",
      "2          biolink:affects                        None   \n",
      "3          biolink:affects                        None   \n",
      "4  biolink:in_complex_with                        None   \n",
      "\n",
      "  object_direction_qualifier  count  \n",
      "0                upregulated  15032  \n",
      "1              downregulated   8832  \n",
      "2                  increased   4886  \n",
      "3                  decreased   3301  \n",
      "4                       None    109  \n"
     ]
    }
   ],
   "source": [
    "## fill na temporily to get the combo number\n",
    "filled = edge_df[['predicate', 'subject_direction_qualifier', 'object_direction_qualifier']].fillna('None')\n",
    "counts = filled.value_counts().reset_index(name='count')\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "1e4f68ae-1978-4710-8d55-26af1c8afc7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes 8666\n",
      "Number of edges 24131\n",
      "Average degree 5.56912070159243\n"
     ]
    }
   ],
   "source": [
    "## Create a graph from the DataFrame\n",
    "graph = nx.from_pandas_edgelist(edge_df, 'subject', 'object', edge_attr='predicate')\n",
    "\n",
    "## Print graph information\n",
    "print('Number of nodes', len(set(graph.nodes)))\n",
    "print('Number of edges', len(set(graph.edges)))\n",
    "print('Average degree', sum(dict(graph.degree).values()) / len(graph.nodes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc1e0907-735d-4aa2-84b6-05fd888395e1",
   "metadata": {},
   "source": [
    "## Now output those parsed files\n",
    "*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "a6c90d72-ce1c-450e-9463-2bf8f65bf118",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the output path for node & edge files after formatting\n",
    "download_path_node_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/wQualityControl/SINGNOR_parsed_node_{version_number}.tsv'\n",
    "download_path_edge_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/wQualityControl/SINGNOR_parsed_edge_{version_number}.tsv'\n",
    "\n",
    "## download both node and edge files\n",
    "## Download the result df\n",
    "## disable download for testing\n",
    "signor_genes_node_df.to_csv(download_path_node_file, sep ='\\t', index=False)\n",
    "edge_df.to_csv(download_path_edge_file, sep ='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6e2fd90-55db-476a-84a7-6938ce4f3de9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "dd32f529-119b-47ec-98f7-f83d3caa79e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Temporarily show full column width (no truncation)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "\n",
    "# Set pandas to display all rows\n",
    "pd.set_option('display.max_rows', None)\n",
    "# print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "630f923e-8b0c-41cb-b285-a02ad328ca33",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (pytorch_env)",
   "language": "python",
   "name": "pytorch_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
