{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "416c7c81-2d1b-4c7b-870c-817905f5d9f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load necessary packages\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "## load the TCT related packages\n",
    "from TCT import node_normalizer\n",
    "from TCT import name_resolver\n",
    "from TCT import translator_metakg\n",
    "from TCT import translator_kpinfo\n",
    "from TCT import translator_query\n",
    "from TCT import TCT\n",
    "\n",
    "## Define the version number\n",
    "version_number = \"08_10_2025\"\n",
    "deployment_date = \"2025-08-10\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "edaa061c-bd8a-459d-9111-28176043ee04",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Date of last update:  2025-07-08\n",
      "Order is to always process Node/category map first, since the Edeg/predicate map depends on biolink-complainat node values\n",
      "-----------------------------------------------------------------------------------------------------------------------------\n",
      "Dictionary: category_map, Key template: Subject_category or Object_category\n",
      "------------------------------------------------------------------------------------------\n",
      "Dictionary: predicate_map, Key template: (Subject_category, Object_category, Predicate)\n"
     ]
    }
   ],
   "source": [
    "## Load the Biolink category and predicate dictionary for mapping subject, object, and predicate types\n",
    "%run ./Biolink_category_and_predication_dictionary.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "0b6e719f-eae3-4d0c-9064-a66042ec3c9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(category_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40fb6377-901f-43a0-ae87-9d813968cad5",
   "metadata": {},
   "source": [
    "## Load files and convert them into separate node & edge files\n",
    "* check all imported file structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "d784b211-517b-4799-b608-095a5d8e118b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Notice!! Please change the file path of following codes into your own\n",
    "raw_files_path = '/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/Pharmacogenetic_Associations_table/'\n",
    "\n",
    "## Define the output path for node & edge files after formatting\n",
    "download_path_node_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/Pharmacogenetic_Associations_table_parsed_node_{version_number}.tsv'\n",
    "download_path_edge_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/Pharmacogenetic_Associations_table_parsed_edge_{version_number}.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "c423f96d-4b9b-4a9d-923a-706877afe31e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pharmacogenetic_associations_table.csv\n"
     ]
    }
   ],
   "source": [
    "## Check all node files being read\n",
    "## Read all BigGIM node csv file in group 1\n",
    "\n",
    "for f in os.listdir(raw_files_path):\n",
    "    if f.endswith('.csv'):\n",
    "        print(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "36c1cbb2-2a7a-48ea-839f-4c4ed3880589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Affected_subgroups</th>\n",
       "      <th>Description_of_Gene_Drug_Interaction</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abacavir</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>*57:01 allele positive</td>\n",
       "      <td>Results in higher adverse reaction risk (hyper...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Abrocitinib</td>\n",
       "      <td>CYP2C19</td>\n",
       "      <td>poor metabolizers</td>\n",
       "      <td>Results in higher systemic concentrations and ...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Amifampridine</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>poor metabolizers</td>\n",
       "      <td>Results in higher systemic concentrations and ...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Amifampridine Phosphate</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>poor metabolizers</td>\n",
       "      <td>Results in higher systemic concentrations. Use...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Amphetamine</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>poor metabolizers</td>\n",
       "      <td>May affect systemic concentrations and adverse...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Aripiprazole</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>poor metabolizers</td>\n",
       "      <td>Results in higher systemic concentrations and ...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Aripiprazole Lauroxil</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>poor metabolizers</td>\n",
       "      <td>Results in higher systemic concentrations. Dos...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Atomoxetine</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>poor metabolizers</td>\n",
       "      <td>Results in higher systemic concentrations and ...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Azathioprine</td>\n",
       "      <td>TPMT and/or NUDT15</td>\n",
       "      <td>intermediate or poor metabolizers</td>\n",
       "      <td>Alters systemic active metabolite concentratio...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Belinostat</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>*28/*28 (poor metabolizers)</td>\n",
       "      <td>May result in higher systemic concentrations a...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Drug                Gene  \\\n",
       "0                 Abacavir               HLA-B   \n",
       "1              Abrocitinib             CYP2C19   \n",
       "2            Amifampridine                NAT2   \n",
       "3  Amifampridine Phosphate                NAT2   \n",
       "4              Amphetamine              CYP2D6   \n",
       "5             Aripiprazole              CYP2D6   \n",
       "6    Aripiprazole Lauroxil              CYP2D6   \n",
       "7              Atomoxetine              CYP2D6   \n",
       "8             Azathioprine  TPMT and/or NUDT15   \n",
       "9               Belinostat              UGT1A1   \n",
       "\n",
       "                  Affected_subgroups  \\\n",
       "0             *57:01 allele positive   \n",
       "1                  poor metabolizers   \n",
       "2                  poor metabolizers   \n",
       "3                  poor metabolizers   \n",
       "4                  poor metabolizers   \n",
       "5                  poor metabolizers   \n",
       "6                  poor metabolizers   \n",
       "7                  poor metabolizers   \n",
       "8  intermediate or poor metabolizers   \n",
       "9        *28/*28 (poor metabolizers)   \n",
       "\n",
       "                Description_of_Gene_Drug_Interaction  \\\n",
       "0  Results in higher adverse reaction risk (hyper...   \n",
       "1  Results in higher systemic concentrations and ...   \n",
       "2  Results in higher systemic concentrations and ...   \n",
       "3  Results in higher systemic concentrations. Use...   \n",
       "4  May affect systemic concentrations and adverse...   \n",
       "5  Results in higher systemic concentrations and ...   \n",
       "6  Results in higher systemic concentrations. Dos...   \n",
       "7  Results in higher systemic concentrations and ...   \n",
       "8  Alters systemic active metabolite concentratio...   \n",
       "9  May result in higher systemic concentrations a...   \n",
       "\n",
       "                                                Type  \n",
       "0  Data Support Therapeutic Management Recommenda...  \n",
       "1  Data Support Therapeutic Management Recommenda...  \n",
       "2  Data Support Therapeutic Management Recommenda...  \n",
       "3  Data Support Therapeutic Management Recommenda...  \n",
       "4  Data Support Therapeutic Management Recommenda...  \n",
       "5  Data Support Therapeutic Management Recommenda...  \n",
       "6  Data Support Therapeutic Management Recommenda...  \n",
       "7  Data Support Therapeutic Management Recommenda...  \n",
       "8  Data Support Therapeutic Management Recommenda...  \n",
       "9  Data Support Therapeutic Management Recommenda...  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Read each individual csv files\n",
    "source_df = pd.read_csv(raw_files_path + 'Pharmacogenetic_associations_table.csv')\n",
    "\n",
    "## Drop nan values\n",
    "source_df = source_df.dropna(subset=['Drug', 'Gene'])\n",
    "\n",
    "source_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "5ab54492-f8ca-4b59-8225-7b8a624ae138",
   "metadata": {},
   "outputs": [],
   "source": [
    "## check all unique values in the Description_of_Gene_Drug_Interaction column\n",
    "unique_description_values = source_df['Description_of_Gene_Drug_Interaction'].unique()\n",
    "# print(unique_description_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "c1452dde-8bae-467b-9186-593433ecb8e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Abacavir' 'Abrocitinib' 'Amifampridine' 'Amifampridine Phosphate'\n",
      " 'Amphetamine' 'Aripiprazole' 'Aripiprazole Lauroxil' 'Atomoxetine'\n",
      " 'Azathioprine' 'Belinostat' 'Belzutifan' 'Brexpiprazole' 'Brivaracetam'\n",
      " 'Capecitabine' 'Carbamazepine' 'Celecoxib' 'Citalopram' 'Clobazam'\n",
      " 'Clopidogrel' 'Clozapine' 'Codeine' 'Deutetrabenazine' 'Dronabinol'\n",
      " 'Eliglustat' 'Erdafitinib' 'Flibanserin' 'Flurbiprofen' 'Fluorouracil'\n",
      " 'Fosphenytoin' 'Gefitinib' 'Iloperidone' 'Irinotecan' 'Lofexidine'\n",
      " 'Meclizine' 'Meloxicam' 'Metoclopramide' 'Mercaptopurine' 'Mivacurium'\n",
      " 'Nateglinide' 'Oliceridine' 'Pantoprazole' 'Phenytoin' 'Pimozide'\n",
      " 'Piroxicam' 'Pitolisant' 'Propafenone' 'Sacituzumab Govitecan-hziy'\n",
      " 'Siponimod' 'Succinylcholine' 'Tacrolimus' 'Tetrabenazine' 'Thioguanine'\n",
      " 'Thioridazine' 'Tramadol' 'Valbenazine' 'Venlafaxine' 'Vortioxetine'\n",
      " 'Warfarin' 'Allopurinol' 'Carvedilol' 'Cevimeline' 'Efavirenz'\n",
      " 'Isoniazid' 'Lapatinib' 'Mavacamten' 'Nilotinib' 'Oxcarbazepine'\n",
      " 'Pazopanib' 'Perphenazine' 'Procainamide' 'Simvastatin'\n",
      " 'Sulfamethoxazole and Trimethoprim' 'Sulfasalazine' 'Tolterodine'\n",
      " 'Voriconazole' 'Amitriptyline' 'Amoxapine' 'Atorvastatin' 'Avatrombopag'\n",
      " 'Carisoprodol' 'Clomipramine' 'Darifenacin' 'Desipramine'\n",
      " 'Dexlansoprazole' 'Diazepam' 'Dolutegravir' 'Donepezil' 'Doxepin'\n",
      " 'Elagolix' 'Escitalopram' 'Esomeprazole' 'Fesoterodine' 'Fluvoxamine'\n",
      " 'Galantamine' 'Hydralazine' 'Ibuprofen' 'Imipramine' 'Lansoprazole'\n",
      " 'Metoprolol' 'Mirabegron' 'Nebivolol' 'Nortriptyline' 'Omeprazole'\n",
      " 'Paroxetine' 'Propranolol' 'Protriptyline' 'Rabeprazole' 'Raltegravir'\n",
      " 'Risperidone' 'Rosuvastatin' 'Tamoxifen' 'Tamsulosin' 'Trimipramine'\n",
      " 'Viloxazine']\n"
     ]
    }
   ],
   "source": [
    "## check all unique values in the Description_of_Gene_Drug_Interaction column\n",
    "unique_drug_values = source_df['Drug'].unique()\n",
    "print(unique_drug_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "73780462-1179-4207-b2f1-bbb5d583e631",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HLA-B' 'CYP2C19' 'NAT2' 'CYP2D6' 'TPMT and/or NUDT15' 'UGT1A1'\n",
      " 'CYP2C19 and/or UGT2B17' 'DPYD' 'CYP2C9' 'BCHE' 'CYP3A5' 'CYP4F2'\n",
      " 'VKORC1' 'HLA-A' 'CYP2B6' 'Nonspecific (NAT)' 'HLA-DRB1' 'HLA-DQA1'\n",
      " 'SLCO1B1']\n"
     ]
    }
   ],
   "source": [
    "## check all unique values in the Description_of_Gene_Drug_Interaction column\n",
    "unique_gene_values = source_df['Gene'].unique()\n",
    "print(unique_gene_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "c8ffd326-ca19-46af-9b39-43510e10d54f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add new columns subject_category and object_category\n",
    "source_df['subject_category'] = 'biolink:Drug'\n",
    "source_df['object_category'] = 'biolink:Gene'\n",
    "source_df['predicate'] = 'biolink:DrugToGeneAssociation'\n",
    "\n",
    "## change the name of columns from col(Drug) to col(subject_name) and col(Gene) to col(object_name) for pandas dataframe full_df\n",
    "source_df = source_df.rename(columns={\n",
    "    'Drug': 'subject_name',\n",
    "    'Gene': 'object_name'\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "aa60bc03-c4e3-415d-994b-11ae10df46dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_name</th>\n",
       "      <th>object_name</th>\n",
       "      <th>Affected_subgroups</th>\n",
       "      <th>Description_of_Gene_Drug_Interaction</th>\n",
       "      <th>Type</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_category</th>\n",
       "      <th>predicate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abacavir</td>\n",
       "      <td>HLA-B</td>\n",
       "      <td>*57:01 allele positive</td>\n",
       "      <td>Results in higher adverse reaction risk (hyper...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:DrugToGeneAssociation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Abrocitinib</td>\n",
       "      <td>CYP2C19</td>\n",
       "      <td>poor metabolizers</td>\n",
       "      <td>Results in higher systemic concentrations and ...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:DrugToGeneAssociation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Amifampridine</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>poor metabolizers</td>\n",
       "      <td>Results in higher systemic concentrations and ...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:DrugToGeneAssociation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Amifampridine Phosphate</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>poor metabolizers</td>\n",
       "      <td>Results in higher systemic concentrations. Use...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:DrugToGeneAssociation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Amphetamine</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>poor metabolizers</td>\n",
       "      <td>May affect systemic concentrations and adverse...</td>\n",
       "      <td>Data Support Therapeutic Management Recommenda...</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:DrugToGeneAssociation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              subject_name object_name      Affected_subgroups  \\\n",
       "0                 Abacavir       HLA-B  *57:01 allele positive   \n",
       "1              Abrocitinib     CYP2C19       poor metabolizers   \n",
       "2            Amifampridine        NAT2       poor metabolizers   \n",
       "3  Amifampridine Phosphate        NAT2       poor metabolizers   \n",
       "4              Amphetamine      CYP2D6       poor metabolizers   \n",
       "\n",
       "                Description_of_Gene_Drug_Interaction  \\\n",
       "0  Results in higher adverse reaction risk (hyper...   \n",
       "1  Results in higher systemic concentrations and ...   \n",
       "2  Results in higher systemic concentrations and ...   \n",
       "3  Results in higher systemic concentrations. Use...   \n",
       "4  May affect systemic concentrations and adverse...   \n",
       "\n",
       "                                                Type subject_category  \\\n",
       "0  Data Support Therapeutic Management Recommenda...     biolink:Drug   \n",
       "1  Data Support Therapeutic Management Recommenda...     biolink:Drug   \n",
       "2  Data Support Therapeutic Management Recommenda...     biolink:Drug   \n",
       "3  Data Support Therapeutic Management Recommenda...     biolink:Drug   \n",
       "4  Data Support Therapeutic Management Recommenda...     biolink:Drug   \n",
       "\n",
       "  object_category                      predicate  \n",
       "0    biolink:Gene  biolink:DrugToGeneAssociation  \n",
       "1    biolink:Gene  biolink:DrugToGeneAssociation  \n",
       "2    biolink:Gene  biolink:DrugToGeneAssociation  \n",
       "3    biolink:Gene  biolink:DrugToGeneAssociation  \n",
       "4    biolink:Gene  biolink:DrugToGeneAssociation  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fdcf190-0f54-43d2-9a12-ac4456d2ccfd",
   "metadata": {},
   "source": [
    "## Execute name resolver to try to find all corresponding identifiers in Translator\n",
    "* use name_resolver.loopup() function\n",
    "* use name_resolver.batch_lookup() function for batch mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "667ac114-3f09-4d9d-88d1-0682367f728e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TranslatorNode(curie='CHEBI:93457', label='5MCA-NAT', types=['biolink:SmallMolecule', 'biolink:MolecularEntity', 'biolink:ChemicalEntity', 'biolink:PhysicalEssence', 'biolink:ChemicalOrDrugOrTreatment', 'biolink:ChemicalEntityOrGeneOrGeneProduct', 'biolink:ChemicalEntityOrProteinOrPolypeptide', 'biolink:NamedThing', 'biolink:Entity', 'biolink:PhysicalEssenceOrOccurrent'], synonyms=None, curie_synonyms=None)\n"
     ]
    }
   ],
   "source": [
    "## test case - Name resolver red flag...\n",
    "name = 'Nonspecific (NAT)'\n",
    "input_node_info = name_resolver.lookup(name)\n",
    "print(input_node_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "5b79d7ee-d95d-4791-ad53-f2392faa77cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TranslatorNode(curie='MONDO:0015912', label='macrothrombocytopenia and granulocyte inclusions with or without nephritis or sensorineural hearing loss', types=['biolink:Disease', 'biolink:DiseaseOrPhenotypicFeature', 'biolink:BiologicalEntity', 'biolink:ThingWithTaxon', 'biolink:NamedThing', 'biolink:Entity'], synonyms=None, curie_synonyms=None)\n"
     ]
    }
   ],
   "source": [
    "## test case - another red flag?\n",
    "name = 'CYP2C19 and/or UGT2B17'\n",
    "input_node_info = name_resolver.lookup(name)\n",
    "print(input_node_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "755d538d-7f70-44f8-bdcd-9bda83d3bd25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TranslatorNode(curie='NCBIGene:590', label='BCHE', types=['biolink:Gene', 'biolink:GeneOrGeneProduct', 'biolink:GenomicEntity', 'biolink:ChemicalEntityOrGeneOrGeneProduct', 'biolink:PhysicalEssence', 'biolink:OntologyClass', 'biolink:BiologicalEntity', 'biolink:ThingWithTaxon', 'biolink:NamedThing', 'biolink:Entity', 'biolink:PhysicalEssenceOrOccurrent', 'biolink:MacromolecularMachineMixin'], synonyms=None, curie_synonyms=None)\n"
     ]
    }
   ],
   "source": [
    "## test case - correct\n",
    "name = 'BCHE'\n",
    "input_node_info = name_resolver.lookup(name)\n",
    "print(input_node_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "2cef1dd4-58e0-4173-aadd-ec05b62159f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHEBI:93457\n"
     ]
    }
   ],
   "source": [
    "print(input_node_info.curie)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "bca3ee68-092c-42c4-a3b6-a17ba025cdbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Apply name_resolver.lookup and extract .curie for subject name\n",
    "# source_df['subject'] = source_df['subject_name'].apply(lambda name: name_resolver.lookup(name).curie if name_resolver.lookup(name) else None)\n",
    "\n",
    "## if True, then the function will only give the first matched row/record\n",
    "## return_top_response = False\n",
    "\n",
    "## switch to use batch_lookup?\n",
    "import pandas as pd\n",
    "\n",
    "# Get all names\n",
    "names = source_df['subject_name'].tolist()\n",
    "\n",
    "# Break into batches of 25\n",
    "batch_size = 25\n",
    "batches = [names[i:i + batch_size] for i in range(0, len(names), batch_size)]\n",
    "\n",
    "# Run batch lookups and collect results\n",
    "results = {}\n",
    "for batch in batches:\n",
    "    lookup_results = name_resolver.batch_lookup(batch)  # Expected to return a dict: {name: result or None}\n",
    "    for name, result in lookup_results.items():\n",
    "        results[name] = result.curie if result else None\n",
    "\n",
    "# Map the resolved CURIEs back to the DataFrame\n",
    "source_df['subject'] = source_df['subject_name'].map(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dadb7492-0d8a-431d-994c-b6ab53aa47ab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed1ea07f-1e86-4c03-9806-8cf4a82a88e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5c8a753-9cff-4440-91b6-acd1a9c90992",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (pytorch_env)",
   "language": "python",
   "name": "pytorch_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
