{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "a3307e15-3432-4044-b012-df30c3efb0eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load necessary packages\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "## Define the version number\n",
    "version_number = \"07_08_2025\"\n",
    "deployment_date = \"2025-07-08\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06be4729-d496-44b5-b3b6-6fa238708649",
   "metadata": {},
   "source": [
    "## Load files and convert them into separate node & edge files\n",
    "* check all imported file structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "cd91d68a-9689-468c-83aa-7130d25c1942",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Notice!! Please change the file path of following codes into your own\n",
    "raw_files_path = '/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/pharmGKB/files/'\n",
    "\n",
    "## Define the output path for node & edge files after formatting\n",
    "download_path_node_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/PharmGKB_parsed_node_{version_number}.tsv'\n",
    "download_path_edge_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/PharmGKB_parsed_edge_{version_number}.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "15e237a3-7f26-4695-ac6c-bf4aee8e3029",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "relationships.tsv\n",
      "phenotypes.tsv\n",
      "genes.tsv\n",
      "chemicals.tsv\n",
      "drugs.tsv\n",
      "variants.tsv\n"
     ]
    }
   ],
   "source": [
    "## Check all node files being read\n",
    "## Read all BigGIM node csv file in group 1\n",
    "\n",
    "for f in os.listdir(raw_files_path):\n",
    "    if f.endswith('.tsv'):\n",
    "        print(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "e03660da-9382-4deb-97c3-8a05a65a87dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "62016\n"
     ]
    }
   ],
   "source": [
    "## Read each individual csv files\n",
    "edges_df = pd.read_csv(raw_files_path + 'relationships.tsv', sep = '\\t')\n",
    "\n",
    "## count rows\n",
    "print(len(edges_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "205301ab-1add-40c1-92d0-a0a20b1b6082",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Entity1_id', 'Entity1_name', 'Entity1_type', 'Entity2_id',\n",
       "       'Entity2_name', 'Entity2_type', 'Evidence', 'Association', 'PK', 'PD',\n",
       "       'PMIDs'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "5176f570-4d6f-4e68-8ad3-e13f8b05be93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All possible relation are here:  ['associated' 'ambiguous' 'not associated']\n"
     ]
    }
   ],
   "source": [
    "## check unique relation\n",
    "unique_relation_values = edges_df['Association'].unique()\n",
    "print(\"All possible relation are here: \" , unique_relation_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "e6886b15-986a-4217-b537-283366b754b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52050\n"
     ]
    }
   ],
   "source": [
    "## filter to only include the associated relation\n",
    "edges_filtered_df = edges_df[edges_df['Association'] == 'associated']\n",
    "## count rows\n",
    "print(len(edges_filtered_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "2db850e2-aa71-47a5-9736-b3b0ddceafa4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entity1_id</th>\n",
       "      <th>Entity1_name</th>\n",
       "      <th>Entity1_type</th>\n",
       "      <th>Entity2_id</th>\n",
       "      <th>Entity2_name</th>\n",
       "      <th>Entity2_type</th>\n",
       "      <th>Evidence</th>\n",
       "      <th>Association</th>\n",
       "      <th>PMIDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA142672624</td>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA447288</td>\n",
       "      <td>Essential hypertension</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ClinicalAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>25695618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA142672624</td>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA449899</td>\n",
       "      <td>hydrochlorothiazide</td>\n",
       "      <td>Chemical</td>\n",
       "      <td>ClinicalAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>25695618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA31744</td>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA128406956</td>\n",
       "      <td>fluorouracil</td>\n",
       "      <td>Chemical</td>\n",
       "      <td>ClinicalAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>18511948;25545243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA31744</td>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA131285527</td>\n",
       "      <td>oxaliplatin</td>\n",
       "      <td>Chemical</td>\n",
       "      <td>ClinicalAnnotation,Literature,MultilinkAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>24924344;25545243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PA31744</td>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA151958383</td>\n",
       "      <td>Gastrointestinal Stromal Tumors</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ClinicalAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>30237583</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Entity1_id Entity1_name Entity1_type   Entity2_id  \\\n",
       "0  PA142672624       ANKFN1         Gene     PA447288   \n",
       "1  PA142672624       ANKFN1         Gene     PA449899   \n",
       "3      PA31744         NQO1         Gene  PA128406956   \n",
       "4      PA31744         NQO1         Gene  PA131285527   \n",
       "5      PA31744         NQO1         Gene  PA151958383   \n",
       "\n",
       "                      Entity2_name Entity2_type  \\\n",
       "0           Essential hypertension      Disease   \n",
       "1              hydrochlorothiazide     Chemical   \n",
       "3                     fluorouracil     Chemical   \n",
       "4                      oxaliplatin     Chemical   \n",
       "5  Gastrointestinal Stromal Tumors      Disease   \n",
       "\n",
       "                                            Evidence Association  \\\n",
       "0                                 ClinicalAnnotation  associated   \n",
       "1                                 ClinicalAnnotation  associated   \n",
       "3                                 ClinicalAnnotation  associated   \n",
       "4  ClinicalAnnotation,Literature,MultilinkAnnotation  associated   \n",
       "5                                 ClinicalAnnotation  associated   \n",
       "\n",
       "               PMIDs  \n",
       "0           25695618  \n",
       "1           25695618  \n",
       "3  18511948;25545243  \n",
       "4  24924344;25545243  \n",
       "5           30237583  "
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Then only keep the needed columns\n",
    "edges_filtered_df2 = edges_filtered_df[['Entity1_id', 'Entity1_name', 'Entity1_type', 'Entity2_id',\n",
    "       'Entity2_name', 'Entity2_type', 'Evidence', 'Association','PMIDs']]\n",
    "\n",
    "edges_filtered_df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "37153b8a-011a-4abb-a90d-5f130e852c48",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entity1_type\n",
      "Chemical     14804\n",
      "Gene         12296\n",
      "Variant      10873\n",
      "Disease       9471\n",
      "Haplotype     4606\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## Check the knowledge_source column again\n",
    "## Count occurrences of each unique value in 'knowledge_source'\n",
    "counts = edges_filtered_df2['Entity1_type'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "b0137824-98c7-4a96-8e90-069919df35c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entity2_type\n",
      "Chemical     14804\n",
      "Gene         12296\n",
      "Variant      10873\n",
      "Disease       9471\n",
      "Haplotype     4606\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## Check the knowledge_source column again\n",
    "## Count occurrences of each unique value in 'knowledge_source'\n",
    "counts = edges_filtered_df2['Entity2_type'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "69528626-53c7-4609-9a73-011e8768c05b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/3967290731.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  chemicals_node_subject_df['PubChem Compound Identifiers'] = pd.to_numeric(chemicals_node_subject_df['PubChem Compound Identifiers'], errors='coerce').astype('Int64')\n",
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/3967290731.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  drugs_node_subject_df['PubChem Compound Identifiers'] = pd.to_numeric(drugs_node_subject_df['PubChem Compound Identifiers'], errors='coerce').astype('Int64')\n"
     ]
    }
   ],
   "source": [
    "## Read in corresponding node files\n",
    "## Read each individual tsv files\n",
    "genes_df = pd.read_csv(raw_files_path + 'genes.tsv', sep = '\\t')\n",
    "\n",
    "## select only needed columns and separate into node & edge files\n",
    "genes_node_subject_df = genes_df[['PharmGKB Accession Id', 'NCBI Gene ID']]\n",
    "\n",
    "chemicals_df = pd.read_csv(raw_files_path + 'chemicals.tsv', sep = '\\t')\n",
    "\n",
    "## select only needed columns and separate into node & edge files\n",
    "chemicals_node_subject_df = chemicals_df[['PharmGKB Accession Id', 'PubChem Compound Identifiers']]\n",
    "chemicals_node_subject_df['PubChem Compound Identifiers'] = pd.to_numeric(chemicals_node_subject_df['PubChem Compound Identifiers'], errors='coerce').astype('Int64')\n",
    "\n",
    "drugs_df = pd.read_csv(raw_files_path + 'drugs.tsv', sep = '\\t')\n",
    "\n",
    "## select only needed columns and separate into node & edge files\n",
    "drugs_node_subject_df = drugs_df[['PharmGKB Accession Id', 'PubChem Compound Identifiers']]\n",
    "drugs_node_subject_df['PubChem Compound Identifiers'] = pd.to_numeric(drugs_node_subject_df['PubChem Compound Identifiers'], errors='coerce').astype('Int64')\n",
    "\n",
    "phenotypes_df = pd.read_csv(raw_files_path + 'phenotypes.tsv', sep = '\\t')\n",
    "\n",
    "## select only needed columns and separate into node & edge files\n",
    "phenotypes_node_subject_df = phenotypes_df[['PharmGKB Accession Id', 'External Vocabulary']]\n",
    "\n",
    "variants_df = pd.read_csv(raw_files_path + 'variants.tsv', sep = '\\t')\n",
    "\n",
    "## select only needed columns and separate into node & edge files\n",
    "variants_node_subject_df = variants_df[['Variant ID', 'Gene IDs']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "ef6c0ded-e048-47f3-b7db-8a78a8e7b56c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PharmGKB Accession Id</th>\n",
       "      <th>NCBI Gene ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA24356</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA165392995</td>\n",
       "      <td>503538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA162375098</td>\n",
       "      <td>29974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA24357</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA142670460</td>\n",
       "      <td>144568</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PharmGKB Accession Id NCBI Gene ID\n",
       "0               PA24356            1\n",
       "1           PA165392995       503538\n",
       "2           PA162375098        29974\n",
       "3               PA24357            2\n",
       "4           PA142670460       144568"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_node_subject_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81d1d075-d95e-480f-b907-a3c226c2ea07",
   "metadata": {},
   "source": [
    "## Dealing with gene df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "88543148-de6c-4d2f-ae64-2b7465e5dbe5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/3572205126.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  genes_node_subject_df['biolink_id'] = 'NCBIGene:' + genes_node_subject_df['NCBI Gene ID'].astype(str)\n",
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/3572205126.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  genes_node_subject_df.rename(columns={'PharmGKB Accession Id': 'PharmGKB_id'}, inplace=True)\n"
     ]
    }
   ],
   "source": [
    "## Create a biolink_id column for genes df\n",
    "## format NCBIGene:XX\n",
    "genes_node_subject_df['biolink_id'] = 'NCBIGene:' + genes_node_subject_df['NCBI Gene ID'].astype(str)\n",
    "\n",
    "\n",
    "## rename those columns into desired format\n",
    "genes_node_subject_df.rename(columns={'PharmGKB Accession Id': 'PharmGKB_id'}, inplace=True)\n",
    "\n",
    "genes_final_df = genes_node_subject_df[['PharmGKB_id', 'biolink_id']]\n",
    "# genes_final_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "1e6ec0df-cdcf-4211-af69-cb3ff82129c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PharmGKB Accession Id</th>\n",
       "      <th>PubChem Compound Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA166250381</td>\n",
       "      <td>59272813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA166250405</td>\n",
       "      <td>117693284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA166250403</td>\n",
       "      <td>59272796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA166178620</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA166131395</td>\n",
       "      <td>101987375</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PharmGKB Accession Id  PubChem Compound Identifiers\n",
       "0           PA166250381                      59272813\n",
       "1           PA166250405                     117693284\n",
       "2           PA166250403                      59272796\n",
       "3           PA166178620                            10\n",
       "4           PA166131395                     101987375"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chemicals_node_subject_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb6e0b74-520e-4e4b-9c37-58ff6252b274",
   "metadata": {},
   "source": [
    "## Dealing with chemicals df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "e412cf0d-b2ee-4307-934c-b52b240a9026",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/3101691258.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  chemicals_node_subject_df['biolink_id'] = 'PUBCHEM.COMPOUND:' + chemicals_node_subject_df['PubChem Compound Identifiers'].astype(str)\n",
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/3101691258.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  chemicals_node_subject_df.rename(columns={'PharmGKB Accession Id': 'PharmGKB_id'}, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PharmGKB_id</th>\n",
       "      <th>biolink_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA166250381</td>\n",
       "      <td>PUBCHEM.COMPOUND:59272813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA166250405</td>\n",
       "      <td>PUBCHEM.COMPOUND:117693284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA166250403</td>\n",
       "      <td>PUBCHEM.COMPOUND:59272796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA166178620</td>\n",
       "      <td>PUBCHEM.COMPOUND:10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA166131395</td>\n",
       "      <td>PUBCHEM.COMPOUND:101987375</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PharmGKB_id                  biolink_id\n",
       "0  PA166250381   PUBCHEM.COMPOUND:59272813\n",
       "1  PA166250405  PUBCHEM.COMPOUND:117693284\n",
       "2  PA166250403   PUBCHEM.COMPOUND:59272796\n",
       "3  PA166178620         PUBCHEM.COMPOUND:10\n",
       "4  PA166131395  PUBCHEM.COMPOUND:101987375"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Create a biolink_id column for genes df\n",
    "## format NCBIGene:XX\n",
    "chemicals_node_subject_df['biolink_id'] = 'PUBCHEM.COMPOUND:' + chemicals_node_subject_df['PubChem Compound Identifiers'].astype(str)\n",
    "\n",
    "\n",
    "## rename those columns into desired format\n",
    "chemicals_node_subject_df.rename(columns={'PharmGKB Accession Id': 'PharmGKB_id'}, inplace=True)\n",
    "\n",
    "chemicals_final_df = chemicals_node_subject_df[['PharmGKB_id', 'biolink_id']]\n",
    "chemicals_final_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "9599c26c-0c86-4ca1-b0d8-4ad88a626492",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PharmGKB Accession Id</th>\n",
       "      <th>External Vocabulary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA446220</td>\n",
       "      <td>MeSH:D015746(Abdominal Pain), SnoMedCT:2072050...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA443224</td>\n",
       "      <td>\"MeSH:D000014(Abnormalities, Drug-Induced)\", \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA444117</td>\n",
       "      <td>MeSH:D005124(Eye Abnormalities), SnoMedCT:1941...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA443227</td>\n",
       "      <td>\"MeSH:D000022(Abortion, Spontaneous)\", \"SnoMed...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA443235</td>\n",
       "      <td>MeSH:D000038(Abscess), SnoMedCT:128477000(Absc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PharmGKB Accession Id                                External Vocabulary\n",
       "0              PA446220  MeSH:D015746(Abdominal Pain), SnoMedCT:2072050...\n",
       "1              PA443224  \"MeSH:D000014(Abnormalities, Drug-Induced)\", \"...\n",
       "2              PA444117  MeSH:D005124(Eye Abnormalities), SnoMedCT:1941...\n",
       "3              PA443227  \"MeSH:D000022(Abortion, Spontaneous)\", \"SnoMed...\n",
       "4              PA443235  MeSH:D000038(Abscess), SnoMedCT:128477000(Absc..."
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotypes_node_subject_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "990b0f8d-44df-4611-8e97-ba68b1777fc7",
   "metadata": {},
   "source": [
    "## Dealing with drugs df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "f753b904-6d1b-48e5-a2cc-e8b85a8e75ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/1052140425.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  drugs_node_subject_df['biolink_id'] = 'PUBCHEM.COMPOUND:' + drugs_node_subject_df['PubChem Compound Identifiers'].astype(str)\n",
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/1052140425.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  drugs_node_subject_df.rename(columns={'PharmGKB Accession Id': 'PharmGKB_id'}, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PharmGKB_id</th>\n",
       "      <th>biolink_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA166238901</td>\n",
       "      <td>PUBCHEM.COMPOUND:23671798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA166238881</td>\n",
       "      <td>PUBCHEM.COMPOUND:9547222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA166238883</td>\n",
       "      <td>PUBCHEM.COMPOUND:23671797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA166238882</td>\n",
       "      <td>PUBCHEM.COMPOUND:23671799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA166238921</td>\n",
       "      <td>PUBCHEM.COMPOUND:71316159</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PharmGKB_id                 biolink_id\n",
       "0  PA166238901  PUBCHEM.COMPOUND:23671798\n",
       "1  PA166238881   PUBCHEM.COMPOUND:9547222\n",
       "2  PA166238883  PUBCHEM.COMPOUND:23671797\n",
       "3  PA166238882  PUBCHEM.COMPOUND:23671799\n",
       "4  PA166238921  PUBCHEM.COMPOUND:71316159"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Create a biolink_id column for genes df\n",
    "## format NCBIGene:XX\n",
    "drugs_node_subject_df['biolink_id'] = 'PUBCHEM.COMPOUND:' + drugs_node_subject_df['PubChem Compound Identifiers'].astype(str)\n",
    "\n",
    "\n",
    "## rename those columns into desired format\n",
    "drugs_node_subject_df.rename(columns={'PharmGKB Accession Id': 'PharmGKB_id'}, inplace=True)\n",
    "\n",
    "drugs_final_df = drugs_node_subject_df[['PharmGKB_id', 'biolink_id']]\n",
    "drugs_final_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e59281a9-0179-4cbd-9636-fcc943b067b5",
   "metadata": {},
   "source": [
    "## Dealing with phenotypes df\n",
    "* https://github.com/TranslatorSRI/NodeNormalization/blob/master/documentation/NodeNormalization.ipynb\n",
    "* all following concepts are acceptable\n",
    "<!-- * \"UMLS\": \"225822\",\n",
    "      \"SNOMEDCT\": \"152614\",\n",
    "      \"MEDDRA\": \"23228\",\n",
    "      \"NCIT\": \"39158\",\n",
    "      \"MONDO\": \"44526\",\n",
    "      \"ORPHANET\": \"18282\",\n",
    "      \"MESH\": \"21066\",\n",
    "      \"HP\": \"3478\",\n",
    "      \"DOID\": \"19624\",\n",
    "      \"OMIM\": \"28954\",\n",
    "      \"EFO\": \"3820\",\n",
    "      \"ICD10\": \"24\",\n",
    "      \"ICD9\": \"12\",\n",
    "      \"MP\": \"4\",\n",
    "      \"medgen\": \"4\" -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "b2cac969-b51a-41a1-9258-32ad84c29409",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/209805454.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  phenotypes_node_subject_df['matched_codes'] = phenotypes_node_subject_df['External Vocabulary'].str.findall(pattern)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PharmGKB Accession Id</th>\n",
       "      <th>External Vocabulary</th>\n",
       "      <th>matched_codes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA446220</td>\n",
       "      <td>MeSH:D015746(Abdominal Pain), SnoMedCT:2072050...</td>\n",
       "      <td>[MeSH:D015746, SnoMedCT:207205003, SnoMedCT:21...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA443224</td>\n",
       "      <td>\"MeSH:D000014(Abnormalities, Drug-Induced)\", \"...</td>\n",
       "      <td>[MeSH:D000014, UMLS:C0000771, MONDO:MONDO:0016...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA444117</td>\n",
       "      <td>MeSH:D005124(Eye Abnormalities), SnoMedCT:1941...</td>\n",
       "      <td>[MeSH:D005124, SnoMedCT:19416009, SnoMedCT:204...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA443227</td>\n",
       "      <td>\"MeSH:D000022(Abortion, Spontaneous)\", \"SnoMed...</td>\n",
       "      <td>[MeSH:D000022, SnoMedCT:17369002, SnoMedCT:198...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA443235</td>\n",
       "      <td>MeSH:D000038(Abscess), SnoMedCT:128477000(Absc...</td>\n",
       "      <td>[MeSH:D000038, SnoMedCT:128477000, SnoMedCT:20...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PharmGKB Accession Id                                External Vocabulary  \\\n",
       "0              PA446220  MeSH:D015746(Abdominal Pain), SnoMedCT:2072050...   \n",
       "1              PA443224  \"MeSH:D000014(Abnormalities, Drug-Induced)\", \"...   \n",
       "2              PA444117  MeSH:D005124(Eye Abnormalities), SnoMedCT:1941...   \n",
       "3              PA443227  \"MeSH:D000022(Abortion, Spontaneous)\", \"SnoMed...   \n",
       "4              PA443235  MeSH:D000038(Abscess), SnoMedCT:128477000(Absc...   \n",
       "\n",
       "                                       matched_codes  \n",
       "0  [MeSH:D015746, SnoMedCT:207205003, SnoMedCT:21...  \n",
       "1  [MeSH:D000014, UMLS:C0000771, MONDO:MONDO:0016...  \n",
       "2  [MeSH:D005124, SnoMedCT:19416009, SnoMedCT:204...  \n",
       "3  [MeSH:D000022, SnoMedCT:17369002, SnoMedCT:198...  \n",
       "4  [MeSH:D000038, SnoMedCT:128477000, SnoMedCT:20...  "
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define regex pattern for the prefixes\n",
    "pattern = r'(MeSH:[^\\(]*|UMLS:[^\\(]*|SnoMedCT:[^\\(]*|MONDO:[^\\(]*|HP:[^\\(]*|MedDRA:[^\\(]*)(?=\\()'\n",
    "\n",
    "# Extract all matches (as lists)\n",
    "phenotypes_node_subject_df['matched_codes'] = phenotypes_node_subject_df['External Vocabulary'].str.findall(pattern)\n",
    "\n",
    "phenotypes_node_subject_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "471d9ee9-15d7-480b-b5c2-e1c2c82752b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/993676129.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  phenotypes_node_subject_df['first_code'] = phenotypes_node_subject_df['matched_codes'].apply(select_first_by_priority)\n"
     ]
    }
   ],
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "\n",
    "# Define ordered prefix list (priority)\n",
    "# Define priority list\n",
    "priority_order = ['MeSH', 'UMLS', 'MONDO', 'HP', 'MedDRA', 'SnoMedCT']\n",
    "\n",
    "# Function to select first matching code by prefix priority\n",
    "def select_first_by_priority(code_list):\n",
    "    if not isinstance(code_list, list):\n",
    "        return None\n",
    "    for prefix in priority_order:\n",
    "        for code in code_list:\n",
    "            if isinstance(code, str) and code.startswith(prefix + ':'):\n",
    "                return code\n",
    "    return None\n",
    "\n",
    "# Apply function to new column\n",
    "phenotypes_node_subject_df['first_code'] = phenotypes_node_subject_df['matched_codes'].apply(select_first_by_priority)\n",
    "\n",
    "# print(phenotypes_node_subject_df[['PharmGKB Accession Id', 'first_code']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "8f052c91-68e1-44be-b426-3d59d11f7838",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/2965652440.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  phenotypes_node_subject_df['first_code'] = phenotypes_node_subject_df['first_code'].apply(remove_duplicate_prefix)\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "# Function to remove repeated prefix\n",
    "def remove_duplicate_prefix(code):\n",
    "    if isinstance(code, str):\n",
    "        return re.sub(r'^([A-Z]+):\\1:', r'\\1:', code)\n",
    "    return code\n",
    "\n",
    "# Apply to 'first_code' column\n",
    "phenotypes_node_subject_df['first_code'] = phenotypes_node_subject_df['first_code'].apply(remove_duplicate_prefix)\n",
    "\n",
    "# print(phenotypes_node_subject_df[['PharmGKB Accession Id', 'first_code']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "2d247b2a-2fc5-4984-b626-8802ba95638a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/935992019.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  phenotypes_node_subject_df.rename(columns={'PharmGKB Accession Id': 'PharmGKB_id', 'first_code': 'biolink_id'}, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PharmGKB_id</th>\n",
       "      <th>biolink_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA446220</td>\n",
       "      <td>MeSH:D015746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA443224</td>\n",
       "      <td>MeSH:D000014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA444117</td>\n",
       "      <td>MeSH:D005124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA443227</td>\n",
       "      <td>MeSH:D000022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA443235</td>\n",
       "      <td>MeSH:D000038</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PharmGKB_id    biolink_id\n",
       "0    PA446220  MeSH:D015746\n",
       "1    PA443224  MeSH:D000014\n",
       "2    PA444117  MeSH:D005124\n",
       "3    PA443227  MeSH:D000022\n",
       "4    PA443235  MeSH:D000038"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## rename those columns into desired format\n",
    "phenotypes_node_subject_df.rename(columns={'PharmGKB Accession Id': 'PharmGKB_id', 'first_code': 'biolink_id'}, inplace=True)\n",
    "\n",
    "phenotypes_final_df = phenotypes_node_subject_df[['PharmGKB_id', 'biolink_id']]\n",
    "\n",
    "## exclude rows with Null values in the biolink_id column\n",
    "phenotypes_final_df = phenotypes_final_df.dropna(subset=['biolink_id'])\n",
    "\n",
    "phenotypes_final_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "8da15b11-85a5-491e-9c12-d39fea0257fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>Gene IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA166156302</td>\n",
       "      <td>PA395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA166156746</td>\n",
       "      <td>PA142671652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA166195421</td>\n",
       "      <td>PA361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA166177121</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA166156636</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Variant ID     Gene IDs\n",
       "0  PA166156302        PA395\n",
       "1  PA166156746  PA142671652\n",
       "2  PA166195421        PA361\n",
       "3  PA166177121          NaN\n",
       "4  PA166156636          NaN"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_node_subject_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7915af5c-dfc7-4c38-aeb7-a1949d854cb0",
   "metadata": {},
   "source": [
    "## Dealing with variants df \n",
    "* not implemented for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "f8bcdbaa-b052-4efc-9a52-1a9085b78cf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Left join on both columns\n",
    "# variants_merged_df = pd.merge(variants_node_subject_df, df2, how='left', on=['PharmGKB_id', 'Gene IDs'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ba9e2b8-cf35-47e7-9618-77d60d053681",
   "metadata": {},
   "source": [
    "## Concatenate those parsed dataframes\n",
    "* genes_final_df\n",
    "* drugs_final_df\n",
    "* chemicals_final_df\n",
    "* phenotypes_final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "1bec4811-f4bd-41bc-8f48-4cf7e96fe262",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35533\n",
      "31762\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PharmGKB_id</th>\n",
       "      <th>biolink_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA24356</td>\n",
       "      <td>NCBIGene:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA165392995</td>\n",
       "      <td>NCBIGene:503538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA162375098</td>\n",
       "      <td>NCBIGene:29974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA24357</td>\n",
       "      <td>NCBIGene:2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA142670460</td>\n",
       "      <td>NCBIGene:144568</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PharmGKB_id       biolink_id\n",
       "0      PA24356       NCBIGene:1\n",
       "1  PA165392995  NCBIGene:503538\n",
       "2  PA162375098   NCBIGene:29974\n",
       "3      PA24357       NCBIGene:2\n",
       "4  PA142670460  NCBIGene:144568"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## vertical concatenation\n",
    "all_node_df = pd.concat([genes_final_df, drugs_final_df, chemicals_final_df, phenotypes_final_df])\n",
    "print(len(all_node_df))\n",
    "\n",
    "all_node_df = all_node_df.drop_duplicates()\n",
    "print(len(all_node_df))\n",
    "\n",
    "all_node_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "9321a43d-0f46-4333-b8ff-4edfe966fe68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entity1_id</th>\n",
       "      <th>Entity1_name</th>\n",
       "      <th>Entity1_type</th>\n",
       "      <th>Entity2_id</th>\n",
       "      <th>Entity2_name</th>\n",
       "      <th>Entity2_type</th>\n",
       "      <th>Evidence</th>\n",
       "      <th>Association</th>\n",
       "      <th>PMIDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA142672624</td>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA447288</td>\n",
       "      <td>Essential hypertension</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ClinicalAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>25695618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA142672624</td>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA449899</td>\n",
       "      <td>hydrochlorothiazide</td>\n",
       "      <td>Chemical</td>\n",
       "      <td>ClinicalAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>25695618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA31744</td>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA128406956</td>\n",
       "      <td>fluorouracil</td>\n",
       "      <td>Chemical</td>\n",
       "      <td>ClinicalAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>18511948;25545243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA31744</td>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA131285527</td>\n",
       "      <td>oxaliplatin</td>\n",
       "      <td>Chemical</td>\n",
       "      <td>ClinicalAnnotation,Literature,MultilinkAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>24924344;25545243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PA31744</td>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA151958383</td>\n",
       "      <td>Gastrointestinal Stromal Tumors</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ClinicalAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>30237583</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Entity1_id Entity1_name Entity1_type   Entity2_id  \\\n",
       "0  PA142672624       ANKFN1         Gene     PA447288   \n",
       "1  PA142672624       ANKFN1         Gene     PA449899   \n",
       "3      PA31744         NQO1         Gene  PA128406956   \n",
       "4      PA31744         NQO1         Gene  PA131285527   \n",
       "5      PA31744         NQO1         Gene  PA151958383   \n",
       "\n",
       "                      Entity2_name Entity2_type  \\\n",
       "0           Essential hypertension      Disease   \n",
       "1              hydrochlorothiazide     Chemical   \n",
       "3                     fluorouracil     Chemical   \n",
       "4                      oxaliplatin     Chemical   \n",
       "5  Gastrointestinal Stromal Tumors      Disease   \n",
       "\n",
       "                                            Evidence Association  \\\n",
       "0                                 ClinicalAnnotation  associated   \n",
       "1                                 ClinicalAnnotation  associated   \n",
       "3                                 ClinicalAnnotation  associated   \n",
       "4  ClinicalAnnotation,Literature,MultilinkAnnotation  associated   \n",
       "5                                 ClinicalAnnotation  associated   \n",
       "\n",
       "               PMIDs  \n",
       "0           25695618  \n",
       "1           25695618  \n",
       "3  18511948;25545243  \n",
       "4  24924344;25545243  \n",
       "5           30237583  "
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_filtered_df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "ac322112-0223-401c-9dea-81dc5782b9d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Entity1_id', 'Entity1_name', 'Entity1_type', 'Entity2_id',\n",
       "       'Entity2_name', 'Entity2_type', 'Evidence', 'Association', 'PMIDs'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_filtered_df2.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "8f7226de-fda3-496a-ab80-b8968fbfe993",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52050\n",
      "36401\n"
     ]
    }
   ],
   "source": [
    "## Then left join with the edge df to obtain subject id and object ids\n",
    "\n",
    "## Left join on two columns with different names\n",
    "edges_subject_id_df = pd.merge(\n",
    "    edges_filtered_df2, all_node_df, how='left',\n",
    "    left_on=['Entity1_id'],\n",
    "    right_on = ['PharmGKB_id'],\n",
    ")\n",
    "\n",
    "print(len(edges_subject_id_df))\n",
    "\n",
    "## exclude rows with Null values in the PharmGKB_id column\n",
    "edges_subject_id_df = edges_subject_id_df.dropna(subset=['PharmGKB_id'])\n",
    "\n",
    "print(len(edges_subject_id_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "33417036-6b7f-4bb4-9bb6-e868c3845249",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_name</th>\n",
       "      <th>Entity1_type</th>\n",
       "      <th>Entity2_id</th>\n",
       "      <th>Entity2_name</th>\n",
       "      <th>Entity2_type</th>\n",
       "      <th>Evidence</th>\n",
       "      <th>Association</th>\n",
       "      <th>PMIDs</th>\n",
       "      <th>subject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA447288</td>\n",
       "      <td>Essential hypertension</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ClinicalAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>25695618</td>\n",
       "      <td>NCBIGene:162282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA449899</td>\n",
       "      <td>hydrochlorothiazide</td>\n",
       "      <td>Chemical</td>\n",
       "      <td>ClinicalAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>25695618</td>\n",
       "      <td>NCBIGene:162282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA128406956</td>\n",
       "      <td>fluorouracil</td>\n",
       "      <td>Chemical</td>\n",
       "      <td>ClinicalAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>18511948;25545243</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA131285527</td>\n",
       "      <td>oxaliplatin</td>\n",
       "      <td>Chemical</td>\n",
       "      <td>ClinicalAnnotation,Literature,MultilinkAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>24924344;25545243</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PA151958383</td>\n",
       "      <td>Gastrointestinal Stromal Tumors</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ClinicalAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td>30237583</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_name Entity1_type   Entity2_id                     Entity2_name  \\\n",
       "0       ANKFN1         Gene     PA447288           Essential hypertension   \n",
       "1       ANKFN1         Gene     PA449899              hydrochlorothiazide   \n",
       "2         NQO1         Gene  PA128406956                     fluorouracil   \n",
       "3         NQO1         Gene  PA131285527                      oxaliplatin   \n",
       "4         NQO1         Gene  PA151958383  Gastrointestinal Stromal Tumors   \n",
       "\n",
       "  Entity2_type                                           Evidence Association  \\\n",
       "0      Disease                                 ClinicalAnnotation  associated   \n",
       "1     Chemical                                 ClinicalAnnotation  associated   \n",
       "2     Chemical                                 ClinicalAnnotation  associated   \n",
       "3     Chemical  ClinicalAnnotation,Literature,MultilinkAnnotation  associated   \n",
       "4      Disease                                 ClinicalAnnotation  associated   \n",
       "\n",
       "               PMIDs          subject  \n",
       "0           25695618  NCBIGene:162282  \n",
       "1           25695618  NCBIGene:162282  \n",
       "2  18511948;25545243    NCBIGene:1728  \n",
       "3  24924344;25545243    NCBIGene:1728  \n",
       "4           30237583    NCBIGene:1728  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## rename the added biolink_id to be subject\n",
    "## rename those columns into desired format\n",
    "edges_subject_id_df.rename(columns={'biolink_id': 'subject', 'Entity1_name': \"subject_name\"}, inplace=True)\n",
    "\n",
    "## drop not needed columns\n",
    "edges_subject_id_df = edges_subject_id_df.drop(columns=['Entity1_id', 'PharmGKB_id'])\n",
    "\n",
    "edges_subject_id_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "825115d9-1e03-46a7-8025-5bb49554db2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36401\n",
      "20958\n"
     ]
    }
   ],
   "source": [
    "## Then left join with the edge df to obtain subject id and object ids\n",
    "\n",
    "## Left join on two columns with different names\n",
    "edges_object_id_df = pd.merge(\n",
    "    edges_subject_id_df, all_node_df, how='left',\n",
    "    left_on=['Entity2_id'],\n",
    "    right_on = ['PharmGKB_id'],\n",
    ")\n",
    "\n",
    "print(len(edges_object_id_df))\n",
    "\n",
    "## exclude rows with Null values in the PharmGKB_id column\n",
    "edges_object_id_df = edges_object_id_df.dropna(subset=['PharmGKB_id'])\n",
    "\n",
    "print(len(edges_object_id_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "517c84f8-6e94-402a-b38b-83a6f5bd441e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_name</th>\n",
       "      <th>Entity1_type</th>\n",
       "      <th>object_name</th>\n",
       "      <th>Entity2_type</th>\n",
       "      <th>Association</th>\n",
       "      <th>publications</th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Essential hypertension</td>\n",
       "      <td>Disease</td>\n",
       "      <td>associated</td>\n",
       "      <td>25695618</td>\n",
       "      <td>NCBIGene:162282</td>\n",
       "      <td>MeSH:D000075222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>hydrochlorothiazide</td>\n",
       "      <td>Chemical</td>\n",
       "      <td>associated</td>\n",
       "      <td>25695618</td>\n",
       "      <td>NCBIGene:162282</td>\n",
       "      <td>PUBCHEM.COMPOUND:3639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>fluorouracil</td>\n",
       "      <td>Chemical</td>\n",
       "      <td>associated</td>\n",
       "      <td>18511948;25545243</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>PUBCHEM.COMPOUND:3385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>oxaliplatin</td>\n",
       "      <td>Chemical</td>\n",
       "      <td>associated</td>\n",
       "      <td>24924344;25545243</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>PUBCHEM.COMPOUND:6857599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gastrointestinal Stromal Tumors</td>\n",
       "      <td>Disease</td>\n",
       "      <td>associated</td>\n",
       "      <td>30237583</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>MeSH:D046152</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_name Entity1_type                      object_name Entity2_type  \\\n",
       "0       ANKFN1         Gene           Essential hypertension      Disease   \n",
       "1       ANKFN1         Gene              hydrochlorothiazide     Chemical   \n",
       "2         NQO1         Gene                     fluorouracil     Chemical   \n",
       "3         NQO1         Gene                      oxaliplatin     Chemical   \n",
       "4         NQO1         Gene  Gastrointestinal Stromal Tumors      Disease   \n",
       "\n",
       "  Association       publications          subject                    object  \n",
       "0  associated           25695618  NCBIGene:162282           MeSH:D000075222  \n",
       "1  associated           25695618  NCBIGene:162282     PUBCHEM.COMPOUND:3639  \n",
       "2  associated  18511948;25545243    NCBIGene:1728     PUBCHEM.COMPOUND:3385  \n",
       "3  associated  24924344;25545243    NCBIGene:1728  PUBCHEM.COMPOUND:6857599  \n",
       "4  associated           30237583    NCBIGene:1728              MeSH:D046152  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## rename the added biolink_id to be subject\n",
    "## rename those columns into desired format\n",
    "edges_object_id_df.rename(columns={'biolink_id': 'object', 'Entity2_name': \"object_name\", 'PMIDs':'publications'}, inplace=True)\n",
    "\n",
    "## drop not needed columns\n",
    "edges_object_id_df = edges_object_id_df.drop(columns=['Entity2_id', 'PharmGKB_id', 'Evidence'])\n",
    "\n",
    "edges_object_id_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "48cf2d31-1d21-496d-b4a8-62f4335db4de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Date of last update:  2025-07-08\n",
      "Order is to always process Node/category map first, since the Edeg/predicate map depends on biolink-complainat node values\n",
      "-----------------------------------------------------------------------------------------------------------------------------\n",
      "Dictionary: category_map, Key template: Subject_category or Object_category\n",
      "------------------------------------------------------------------------------------------\n",
      "Dictionary: predicate_map, Key template: (Subject_category, Object_category, Predicate)\n"
     ]
    }
   ],
   "source": [
    "## Load the Biolink category and predicate dictionary for mapping subject, object, and predicate types\n",
    "%run ./Biolink_category_and_predication_dictionary.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "fb72eea7-1b47-4b88-9e05-e8a0cf41280d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## deep copy a df\n",
    "edges_deep_copy_df = edges_object_id_df.copy(deep=True)\n",
    "\n",
    "## deprecated\n",
    "# edges_deep_copy_df['subject_category'] = 'biolink:' + edges_deep_copy_df['Entity1_type'].astype(str)\n",
    "# edges_deep_copy_df['object_category'] = 'biolink:' + edges_deep_copy_df['Entity2_type'].astype(str)\n",
    "\n",
    "## assign biolink node category using dictionary directly\n",
    "edges_deep_copy_df['subject_category'] = (\n",
    "    edges_deep_copy_df['Entity1_type'].map(category_map)\n",
    ")\n",
    "\n",
    "edges_deep_copy_df['object_category'] = (\n",
    "    edges_deep_copy_df['Entity2_type'].map(category_map)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "15ae2d26-c682-449b-acc1-39bfc5287fdf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Predicate: associated\n",
      "  (biolink:Gene, biolink:Disease)\n",
      "  (biolink:Gene, biolink:ChemicalEntity)\n",
      "  (biolink:ChemicalEntity, biolink:Gene)\n",
      "  (biolink:ChemicalEntity, biolink:ChemicalEntity)\n",
      "  (biolink:Disease, biolink:Gene)\n",
      "  (biolink:Gene, biolink:Gene)\n",
      "  (biolink:Disease, biolink:Disease)\n"
     ]
    }
   ],
   "source": [
    "## Group by predicate\n",
    "grouped = edges_deep_copy_df.groupby('Association')\n",
    "\n",
    "## For each predicate, output unique (subject_category, object_category) pairs\n",
    "for predicate, group in grouped:\n",
    "    print(f\"\\nPredicate: {predicate}\")\n",
    "    pairs = group[['subject_category', 'object_category']].drop_duplicates()\n",
    "    for _, row in pairs.iterrows():\n",
    "        print(f\"  ({row['subject_category']}, {row['object_category']})\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "1ce581f9-e053-4f93-bf5e-4460b24aac7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_name</th>\n",
       "      <th>object_name</th>\n",
       "      <th>publications</th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_category</th>\n",
       "      <th>predicate</th>\n",
       "      <th>knowledge_source</th>\n",
       "      <th>knowledge_level</th>\n",
       "      <th>agent_type</th>\n",
       "      <th>deploy_date</th>\n",
       "      <th>context_qualifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>Essential hypertension</td>\n",
       "      <td>25695618</td>\n",
       "      <td>NCBIGene:162282</td>\n",
       "      <td>MeSH:D000075222</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Disease</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>hydrochlorothiazide</td>\n",
       "      <td>25695618</td>\n",
       "      <td>NCBIGene:162282</td>\n",
       "      <td>PUBCHEM.COMPOUND:3639</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NQO1</td>\n",
       "      <td>fluorouracil</td>\n",
       "      <td>18511948;25545243</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>PUBCHEM.COMPOUND:3385</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NQO1</td>\n",
       "      <td>oxaliplatin</td>\n",
       "      <td>24924344;25545243</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>PUBCHEM.COMPOUND:6857599</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gastrointestinal Stromal Tumors</td>\n",
       "      <td>30237583</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>MeSH:D046152</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Disease</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_name                      object_name       publications  \\\n",
       "0       ANKFN1           Essential hypertension           25695618   \n",
       "1       ANKFN1              hydrochlorothiazide           25695618   \n",
       "2         NQO1                     fluorouracil  18511948;25545243   \n",
       "3         NQO1                      oxaliplatin  24924344;25545243   \n",
       "4         NQO1  Gastrointestinal Stromal Tumors           30237583   \n",
       "\n",
       "           subject                    object subject_category  \\\n",
       "0  NCBIGene:162282           MeSH:D000075222     biolink:Gene   \n",
       "1  NCBIGene:162282     PUBCHEM.COMPOUND:3639     biolink:Gene   \n",
       "2    NCBIGene:1728     PUBCHEM.COMPOUND:3385     biolink:Gene   \n",
       "3    NCBIGene:1728  PUBCHEM.COMPOUND:6857599     biolink:Gene   \n",
       "4    NCBIGene:1728              MeSH:D046152     biolink:Gene   \n",
       "\n",
       "          object_category                predicate knowledge_source  \\\n",
       "0         biolink:Disease  biolink:associated_with         PharmGKB   \n",
       "1  biolink:ChemicalEntity  biolink:associated_with         PharmGKB   \n",
       "2  biolink:ChemicalEntity  biolink:associated_with         PharmGKB   \n",
       "3  biolink:ChemicalEntity  biolink:associated_with         PharmGKB   \n",
       "4         biolink:Disease  biolink:associated_with         PharmGKB   \n",
       "\n",
       "       knowledge_level       agent_type deploy_date  context_qualifier  \n",
       "0  knowledge_assertion  automated_agent  2025-07-08                NaN  \n",
       "1  knowledge_assertion  automated_agent  2025-07-08                NaN  \n",
       "2  knowledge_assertion  automated_agent  2025-07-08                NaN  \n",
       "3  knowledge_assertion  automated_agent  2025-07-08                NaN  \n",
       "4  knowledge_assertion  automated_agent  2025-07-08                NaN  "
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "## depracted\n",
    "# ## create predicates column\n",
    "# list_of_display_relations = ['associated']\n",
    "\n",
    "# relation_predicate_map = {\n",
    "#     'associated': 'biolink:associated_with',\n",
    "# }\n",
    "\n",
    "# ## Initialize new column with default values (original id as string)\n",
    "# edges_deep_copy_df['predicate'] = edges_deep_copy_df['Association'].astype(str)\n",
    "\n",
    "# ## Mask\n",
    "# mask = edges_deep_copy_df['Association'].isin(list_of_display_relations)\n",
    "# edges_deep_copy_df.loc[mask, 'predicate'] = (\n",
    "#     edges_deep_copy_df.loc[mask, 'Association'].map(relation_predicate_map)\n",
    "# )\n",
    "\n",
    "edges_deep_copy_df['predicate'] = edges_deep_copy_df.apply(\n",
    "    lambda row: predicate_map.get((row['subject_category'], row['object_category'], row['Association'])),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "## drop not needed columns\n",
    "edges_deep_copy_df = edges_deep_copy_df.drop(columns=['Entity1_type', 'Entity2_type', \"Association\"])\n",
    "\n",
    "## add a new knowledge_souce column and set value to be \"PrimeKG\"\n",
    "edges_deep_copy_df['knowledge_source'] = 'PharmGKB'\n",
    "## add a new knowledge_level column and set value to be 'knowledge_assertion'\n",
    "edges_deep_copy_df['knowledge_level'] = 'knowledge_assertion'\n",
    "## add a new agent_type column and set value to be 'manual_agent'\n",
    "edges_deep_copy_df['agent_type'] = 'automated_agent'\n",
    "\n",
    "edges_deep_copy_df['deploy_date'] = deployment_date\n",
    "\n",
    "## create a context_qualifier column and fill na\n",
    "## if all of them are empty then fill na\n",
    "edges_deep_copy_df['context_qualifier'] = np.nan\n",
    "\n",
    "edges_deep_copy_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "3f6a6e26-adf2-4206-b01f-dc19cf32bb64",
   "metadata": {},
   "outputs": [],
   "source": [
    "import uuid\n",
    "import pandas as pd\n",
    "\n",
    "## generate uuid from column combination\n",
    "def generate_uuid_from_columns(df, column_list, namespace=uuid.NAMESPACE_DNS):\n",
    "    \"\"\"\n",
    "    Generates UUIDs based on the values in a specified column of a Pandas DataFrame.\n",
    "\n",
    "    Args:\n",
    "        df (pd.DataFrame): The input DataFrame.\n",
    "        column_list (list): List of all names of columns to use for UUID generation.\n",
    "        namespace (uuid.UUID): A UUID namespace (default is uuid.NAMESPACE_DNS).\n",
    "\n",
    "    Returns:\n",
    "        pd.Series: A Pandas Series containing the generated UUIDs.\n",
    "    \"\"\"\n",
    "    return df[column_list].apply(lambda x: uuid.uuid5(namespace, str(x)).hex)\n",
    "\n",
    "def generate_uuid(row):\n",
    "    \"\"\"\n",
    "    Generates a UUID based on the combined values of multiple columns.\n",
    "    \"\"\"\n",
    "    combined_string = ''.join(row.astype(str))\n",
    "    return uuid.uuid5(uuid.NAMESPACE_DNS, combined_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "dcc21204-4124-4f70-865f-c06526846f2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_name</th>\n",
       "      <th>object_name</th>\n",
       "      <th>publications</th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_category</th>\n",
       "      <th>predicate</th>\n",
       "      <th>knowledge_source</th>\n",
       "      <th>knowledge_level</th>\n",
       "      <th>agent_type</th>\n",
       "      <th>deploy_date</th>\n",
       "      <th>context_qualifier</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>Essential hypertension</td>\n",
       "      <td>25695618</td>\n",
       "      <td>NCBIGene:162282</td>\n",
       "      <td>MeSH:D000075222</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Disease</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3f4622cd-cc81-5c60-9fce-4d25214b9b86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>hydrochlorothiazide</td>\n",
       "      <td>25695618</td>\n",
       "      <td>NCBIGene:162282</td>\n",
       "      <td>PUBCHEM.COMPOUND:3639</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2ebd74fd-1cb6-519c-9051-4e0e5880a57d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NQO1</td>\n",
       "      <td>fluorouracil</td>\n",
       "      <td>18511948;25545243</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>PUBCHEM.COMPOUND:3385</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3e9e9eee-e33e-5903-b6c0-fd6f960ab809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NQO1</td>\n",
       "      <td>oxaliplatin</td>\n",
       "      <td>24924344;25545243</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>PUBCHEM.COMPOUND:6857599</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5dfcff5b-dea4-53f7-b9ff-1db8d23309df</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NQO1</td>\n",
       "      <td>Gastrointestinal Stromal Tumors</td>\n",
       "      <td>30237583</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>MeSH:D046152</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Disease</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6149843b-5b1d-562b-b2e9-82391a552db3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_name                      object_name       publications  \\\n",
       "0       ANKFN1           Essential hypertension           25695618   \n",
       "1       ANKFN1              hydrochlorothiazide           25695618   \n",
       "2         NQO1                     fluorouracil  18511948;25545243   \n",
       "3         NQO1                      oxaliplatin  24924344;25545243   \n",
       "4         NQO1  Gastrointestinal Stromal Tumors           30237583   \n",
       "\n",
       "           subject                    object subject_category  \\\n",
       "0  NCBIGene:162282           MeSH:D000075222     biolink:Gene   \n",
       "1  NCBIGene:162282     PUBCHEM.COMPOUND:3639     biolink:Gene   \n",
       "2    NCBIGene:1728     PUBCHEM.COMPOUND:3385     biolink:Gene   \n",
       "3    NCBIGene:1728  PUBCHEM.COMPOUND:6857599     biolink:Gene   \n",
       "4    NCBIGene:1728              MeSH:D046152     biolink:Gene   \n",
       "\n",
       "          object_category                predicate knowledge_source  \\\n",
       "0         biolink:Disease  biolink:associated_with         PharmGKB   \n",
       "1  biolink:ChemicalEntity  biolink:associated_with         PharmGKB   \n",
       "2  biolink:ChemicalEntity  biolink:associated_with         PharmGKB   \n",
       "3  biolink:ChemicalEntity  biolink:associated_with         PharmGKB   \n",
       "4         biolink:Disease  biolink:associated_with         PharmGKB   \n",
       "\n",
       "       knowledge_level       agent_type deploy_date  context_qualifier  \\\n",
       "0  knowledge_assertion  automated_agent  2025-07-08                NaN   \n",
       "1  knowledge_assertion  automated_agent  2025-07-08                NaN   \n",
       "2  knowledge_assertion  automated_agent  2025-07-08                NaN   \n",
       "3  knowledge_assertion  automated_agent  2025-07-08                NaN   \n",
       "4  knowledge_assertion  automated_agent  2025-07-08                NaN   \n",
       "\n",
       "                                     id  \n",
       "0  3f4622cd-cc81-5c60-9fce-4d25214b9b86  \n",
       "1  2ebd74fd-1cb6-519c-9051-4e0e5880a57d  \n",
       "2  3e9e9eee-e33e-5903-b6c0-fd6f960ab809  \n",
       "3  5dfcff5b-dea4-53f7-b9ff-1db8d23309df  \n",
       "4  6149843b-5b1d-562b-b2e9-82391a552db3  "
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## copy to a final df\n",
    "edge_df = edges_deep_copy_df.copy(deep = True)\n",
    "\n",
    "### Add resources_id column, checking whether edge is already\n",
    "column_list = ['subject', 'predicate', 'object', 'context_qualifier', 'deploy_date']\n",
    "# Apply the function to each row to generate UUIDs\n",
    "edge_df['id'] = edge_df[column_list].apply(generate_uuid, axis=1)\n",
    "\n",
    "# edge_df['id'] = generate_uuid_from_columns(edge_df, column_list)\n",
    "edge_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "b844b699-af95-4bd2-b2c5-71ffdd0418a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Drop rows where 'name' is NaN, None, or empty string\n",
    "edge_df = edge_df[~edge_df['predicate'].isna() & (edge_df['predicate'].str.strip() != '')]\n",
    "\n",
    "## throw away those rows which either subject_category, object_category, or predicate is not start with \"biolink:\" prefix\n",
    "## since they cannot be biolink-compliant converted\n",
    "## Keep only rows where all three columns start with 'biolink:'\n",
    "edge_df = edge_df[\n",
    "    edge_df['subject_category'].str.startswith('biolink:') &\n",
    "    edge_df['object_category'].str.startswith('biolink:') &\n",
    "    edge_df['predicate'].str.startswith('biolink:')\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b429e30-d966-4b07-86e2-1c66b0ca3ee2",
   "metadata": {},
   "source": [
    "### Now create the corresponding node file\n",
    "* only need three columns: id, name, category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "fedb20dd-29dd-4873-af6c-440c68caf4c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/1611443582.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  node_subject_df.rename(columns={'subject': 'id', 'subject_name': 'name', 'subject_category': 'category'}, inplace=True)\n",
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_99818/1611443582.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  node_object_df.rename(columns={'object': 'id', 'object_name': 'name', 'object_category': 'category'}, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:162282</td>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>NQO1</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NCBIGene:6813</td>\n",
       "      <td>STXBP2</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PUBCHEM.COMPOUND:33</td>\n",
       "      <td>chloroacetaldehyde</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NCBIGene:8000</td>\n",
       "      <td>PSCA</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id                name                category\n",
       "0       NCBIGene:162282              ANKFN1            biolink:Gene\n",
       "2         NCBIGene:1728                NQO1            biolink:Gene\n",
       "21        NCBIGene:6813              STXBP2            biolink:Gene\n",
       "22  PUBCHEM.COMPOUND:33  chloroacetaldehyde  biolink:ChemicalEntity\n",
       "23        NCBIGene:8000                PSCA            biolink:Gene"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_subject_df = edge_df[['subject', 'subject_name', 'subject_category']]\n",
    "node_object_df = edge_df[['object', 'object_name', 'object_category']]\n",
    "\n",
    "## rename those columns into desired format\n",
    "node_subject_df.rename(columns={'subject': 'id', 'subject_name': 'name', 'subject_category': 'category'}, inplace=True)\n",
    "node_object_df.rename(columns={'object': 'id', 'object_name': 'name', 'object_category': 'category'}, inplace=True)\n",
    "\n",
    "concat_node_df = pd.concat([node_subject_df, node_object_df]).drop_duplicates(keep='first')\n",
    "\n",
    "concat_node_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "c0b630a1-dc7b-45d2-8d5e-3163da639bd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>publications</th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_category</th>\n",
       "      <th>predicate</th>\n",
       "      <th>knowledge_source</th>\n",
       "      <th>knowledge_level</th>\n",
       "      <th>agent_type</th>\n",
       "      <th>deploy_date</th>\n",
       "      <th>context_qualifier</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>25695618</td>\n",
       "      <td>NCBIGene:162282</td>\n",
       "      <td>MeSH:D000075222</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Disease</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3f4622cd-cc81-5c60-9fce-4d25214b9b86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25695618</td>\n",
       "      <td>NCBIGene:162282</td>\n",
       "      <td>PUBCHEM.COMPOUND:3639</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2ebd74fd-1cb6-519c-9051-4e0e5880a57d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18511948;25545243</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>PUBCHEM.COMPOUND:3385</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3e9e9eee-e33e-5903-b6c0-fd6f960ab809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>24924344;25545243</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>PUBCHEM.COMPOUND:6857599</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5dfcff5b-dea4-53f7-b9ff-1db8d23309df</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30237583</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>MeSH:D046152</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Disease</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6149843b-5b1d-562b-b2e9-82391a552db3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        publications          subject                    object  \\\n",
       "0           25695618  NCBIGene:162282           MeSH:D000075222   \n",
       "1           25695618  NCBIGene:162282     PUBCHEM.COMPOUND:3639   \n",
       "2  18511948;25545243    NCBIGene:1728     PUBCHEM.COMPOUND:3385   \n",
       "3  24924344;25545243    NCBIGene:1728  PUBCHEM.COMPOUND:6857599   \n",
       "4           30237583    NCBIGene:1728              MeSH:D046152   \n",
       "\n",
       "  subject_category         object_category                predicate  \\\n",
       "0     biolink:Gene         biolink:Disease  biolink:associated_with   \n",
       "1     biolink:Gene  biolink:ChemicalEntity  biolink:associated_with   \n",
       "2     biolink:Gene  biolink:ChemicalEntity  biolink:associated_with   \n",
       "3     biolink:Gene  biolink:ChemicalEntity  biolink:associated_with   \n",
       "4     biolink:Gene         biolink:Disease  biolink:associated_with   \n",
       "\n",
       "  knowledge_source      knowledge_level       agent_type deploy_date  \\\n",
       "0         PharmGKB  knowledge_assertion  automated_agent  2025-07-08   \n",
       "1         PharmGKB  knowledge_assertion  automated_agent  2025-07-08   \n",
       "2         PharmGKB  knowledge_assertion  automated_agent  2025-07-08   \n",
       "3         PharmGKB  knowledge_assertion  automated_agent  2025-07-08   \n",
       "4         PharmGKB  knowledge_assertion  automated_agent  2025-07-08   \n",
       "\n",
       "   context_qualifier                                    id  \n",
       "0                NaN  3f4622cd-cc81-5c60-9fce-4d25214b9b86  \n",
       "1                NaN  2ebd74fd-1cb6-519c-9051-4e0e5880a57d  \n",
       "2                NaN  3e9e9eee-e33e-5903-b6c0-fd6f960ab809  \n",
       "3                NaN  5dfcff5b-dea4-53f7-b9ff-1db8d23309df  \n",
       "4                NaN  6149843b-5b1d-562b-b2e9-82391a552db3  "
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## drop not needed columns\n",
    "edge_final_df = edge_df.drop(columns=['subject_name', 'object_name'])\n",
    "\n",
    "edge_final_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "9306a3b2-7e9e-41d2-8ec0-11cecff04d06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes 3217\n",
      "Number of edges 9919\n",
      "Average degree 6.16661485856388\n"
     ]
    }
   ],
   "source": [
    "## Create a graph from the DataFrame\n",
    "graph = nx.from_pandas_edgelist(edge_final_df, 'subject', 'object', edge_attr='predicate')\n",
    "\n",
    "## Print graph information\n",
    "print('Number of nodes', len(set(graph.nodes)))\n",
    "print('Number of edges', len(set(graph.edges)))\n",
    "print('Average degree', sum(dict(graph.degree).values()) / len(graph.nodes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62cfc11f-0baa-4e76-8b3e-04aa2b33c445",
   "metadata": {},
   "source": [
    "## Now download the concatenated node & edge files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "8b3bac0e-0ad5-4e2e-9f79-274fc3396992",
   "metadata": {},
   "outputs": [],
   "source": [
    "## download both node and edge files\n",
    "## Download the result df\n",
    "concat_node_df.to_csv(download_path_node_file, sep ='\\t', index=False)\n",
    "edge_final_df.to_csv(download_path_edge_file, sep ='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "1b0c559f-5e28-4bb9-8590-7102fa2dbcfa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The formatted node file will be saved in this path:  /Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/PharmGKB_parsed_node_07_08_2025.tsv\n",
      "The formatted edge file will be saved in this path:  /Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/PharmGKB_parsed_edge_07_08_2025.tsv\n"
     ]
    }
   ],
   "source": [
    "print(\"The formatted node file will be saved in this path: \", download_path_node_file)\n",
    "print(\"The formatted edge file will be saved in this path: \", download_path_edge_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4302f980-8773-43e1-b0cc-c94f29c6138c",
   "metadata": {},
   "source": [
    "## following codes are used for quality control and sanity check\n",
    "* check and confirm all subject & object types are correctly formatted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "d76e4c32-df1c-4d8c-b416-33e980102f78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "subject_category\n",
      "biolink:Gene              11702\n",
      "biolink:ChemicalEntity     5907\n",
      "biolink:Disease            3349\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = edge_final_df['subject_category'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "7e497d11-77e5-4aa4-a2e4-05b136ee30df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "object_category\n",
      "biolink:Gene              11702\n",
      "biolink:ChemicalEntity     5907\n",
      "biolink:Disease            3349\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = edge_final_df['object_category'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "07f05270-b4b4-429f-a51b-625c8b2d25cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "predicate\n",
      "biolink:associated_with    20958\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = edge_final_df['predicate'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dbc517d-5304-4b21-bf1e-e1390cf5273c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (pytorch_env)",
   "language": "python",
   "name": "pytorch_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
