{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1ec22cb6-0ca7-4dae-b98a-f6a1571bb3d0",
   "metadata": {},
   "source": [
    "## Group 1: merge and biolink format all tsv files\n",
    "1. with predicate biolink:physically_interacts_with and biolink:gene_associated_with_condition\n",
    "\n",
    "2. the corresponding config.json file is: config_bigGIM_interacts_with_associated_with\n",
    "\n",
    "3. all of them have same cols \"subject, predicate, object, agent_type, knowledge_level, knowledge_source, object_category, publications, subject_category\"\n",
    "\n",
    "4. Full list of tsv files handled in this group is:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "60bbd553-5589-4a47-b2a2-5b3a21a02363",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-08T01:03:06.764942Z",
     "start_time": "2025-10-08T01:03:06.762481Z"
    }
   },
   "outputs": [],
   "source": [
    "## To do list:\n",
    "## Add signor and cell_marker_genes.csv into pharmacogenomics KG\n",
    "    ## Done\n",
    "## Check overlap with BigGIM KG and the best way to remove redundancy edges\n",
    "    ## 0 duplicated edges when checking the same combination of subject, object and predicate\n",
    "## Check which part belongs to AML\n",
    "    ## 652 edges related to AML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "9f186896-231d-4705-853b-d1e6a39efed4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-08T01:03:09.562192Z",
     "start_time": "2025-10-08T01:03:06.780148Z"
    }
   },
   "outputs": [],
   "source": [
    "## Load necessary packages\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "## Define the version number\n",
    "version_number = \"01_06_2026\"\n",
    "deployment_date = \"2026-01-06\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "9465e3cf-9585-43e3-8c72-54b627baec09",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-08T01:03:15.113121Z",
     "start_time": "2025-10-08T01:03:14.883316Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Date of last update:  2025-09-04\n",
      "Order is to always process Node/category map first, since the Edeg/predicate map depends on biolink-complainat node values\n",
      "-----------------------------------------------------------------------------------------------------------------------------\n",
      "Dictionary: category_map, Key template: Subject_category or Object_category\n",
      "------------------------------------------------------------------------------------------\n",
      "Dictionary: predicate_map, Key template: (Subject_category, Object_category, Predicate)\n"
     ]
    }
   ],
   "source": [
    "## Load the Biolink category and predicate dictionary for mapping subject, object, and predicate types\n",
    "%run ./Biolink_category_and_predication_dictionary.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "1df45b8e-5c23-48f4-8d24-a3feb36b8b5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(category_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e11c9dd5-0bcd-4a4a-b2be-05cd945b701b",
   "metadata": {},
   "source": [
    "## Load files and convert them into separate node & edge files\n",
    "* check all imported file structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "85d8fd31-c247-4b72-b72e-7bd612d54584",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Notice!! Please change the file path of following codes into your own\n",
    "raw_files_path = '/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/primeKG/dataverse_files/'\n",
    "\n",
    "## Define the output path for node & edge files after formatting\n",
    "download_path_node_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/primeKG_parsed_node_{version_number}.tsv'\n",
    "download_path_edge_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/primeKG_parsed_edge_{version_number}.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "241665da-ce83-47e8-9e8f-ac6acce47bc8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "disease_features.csv\n",
      "nodes.csv\n",
      "kg_grouped.csv\n",
      "kg_grouped_diseases.csv\n",
      "kg_grouped_diseases_bert_map.csv\n",
      "kg_giant.csv\n",
      "kg.csv\n",
      "drug_features.csv\n",
      "kg_raw.csv\n",
      "edges.csv\n"
     ]
    }
   ],
   "source": [
    "## Check all node files being read\n",
    "## Read all BigGIM node csv file in group 1\n",
    "\n",
    "for f in os.listdir(raw_files_path):\n",
    "    if f.endswith('.csv'):\n",
    "        print(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "7461f92a-5730-4af3-99c4-e7da71fcbe79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -----------------------------------------------------------------------------------------------\n",
    "# Filename\t\t\t\tDescription\n",
    "# -----------------------------------------------------------------------------------------------\n",
    "# nodes.csv\t\t\t\tContains node level information\n",
    "# \t\t\t\t\tPrimary key: `node_index`\n",
    "# -----------------------------------------------------------------------------------------------\n",
    "# edges.csv\t\t\t\tContains undirected relationships between nodes \n",
    "# \t\t\t\t\tPrimary key: (`x_index`, `y_index`)\n",
    "# -----------------------------------------------------------------------------------------------\n",
    "# kg.csv\t\t\t\t\tThis is the Precision Medicine knowledge graph  \n",
    "# \t\t\t\t\tPrimary key: (`x_index`, `y_index`)\n",
    "# -----------------------------------------------------------------------------------------------\n",
    "# disease_features.csv\t\t\tContains textual descriptions of diseases \n",
    "# \t\t\t\t\tPrimary key: `node_index`\n",
    "# -----------------------------------------------------------------------------------------------\n",
    "# drug_features.csv\t\t\tContains textual descriptions of diseases \n",
    "# \t\t\t\t\tPrimary key: `node_index`\n",
    "# -----------------------------------------------------------------------------------------------\n",
    "# kg_raw.csv \t\t\t\tIntermediate PrimeKG made by joining nodes and edges\n",
    "# -----------------------------------------------------------------------------------------------\n",
    "# kg_giant.csv\t\t\t\tIntermediate PrimeKG made by taking LCC of kg_raw.csv \n",
    "# -----------------------------------------------------------------------------------------------\n",
    "# kg_grouped.csv\t\t\t\tIntermediate PrimeKG made by grouping diseases  \n",
    "# -----------------------------------------------------------------------------------------------\n",
    "# kg_grouped_diseases.csv\t\t\tList of all diseases and their assigned group name  \n",
    "# -----------------------------------------------------------------------------------------------\n",
    "# kg_grouped_diseases_bert_map.csv\tManual grouping created for diseases using BERT model\n",
    "# ---------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "ee9554a9-fc4d-4f50-8d30-c3c9f1fee9bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_index</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>node_name</th>\n",
       "      <th>node_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>9796</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PHYHIP</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>7918</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GPANK1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>8233</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>4899</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NRF1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>5297</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PI4KA</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>6564</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>SLC15A1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>8668</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>EIF3I</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>10826</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>FAXDC2</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>4489</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MT1A</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>6272</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>SORT1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   node_index node_id     node_type node_name node_source\n",
       "0           0    9796  gene/protein    PHYHIP        NCBI\n",
       "1           1    7918  gene/protein    GPANK1        NCBI\n",
       "2           2    8233  gene/protein     ZRSR2        NCBI\n",
       "3           3    4899  gene/protein      NRF1        NCBI\n",
       "4           4    5297  gene/protein     PI4KA        NCBI\n",
       "5           5    6564  gene/protein   SLC15A1        NCBI\n",
       "6           6    8668  gene/protein     EIF3I        NCBI\n",
       "7           7   10826  gene/protein    FAXDC2        NCBI\n",
       "8           8    4489  gene/protein      MT1A        NCBI\n",
       "9           9    6272  gene/protein     SORT1        NCBI"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Read each individual csv files\n",
    "nodes_df = pd.read_csv(raw_files_path + 'nodes.csv')\n",
    "\n",
    "nodes_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "ad0e000c-86e2-406c-8d1e-91ae54a09bef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All possible node_type are here:  ['gene/protein' 'drug' 'effect/phenotype' 'disease' 'biological_process'\n",
      " 'molecular_function' 'cellular_component' 'exposure' 'pathway' 'anatomy']\n",
      "All possible node_source are here:  ['NCBI' 'DrugBank' 'HPO' 'MONDO_grouped' 'MONDO' 'GO' 'CTD' 'REACTOME'\n",
      " 'UBERON']\n"
     ]
    }
   ],
   "source": [
    "## check unique node_type and their node_source\n",
    "unique_node_type_values = nodes_df['node_type'].unique()\n",
    "print(\"All possible node_type are here: \" ,unique_node_type_values)\n",
    "\n",
    "unique_node_source_values = nodes_df['node_source'].unique()\n",
    "print(\"All possible node_source are here: \" ,unique_node_source_values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "148ef307-e4c5-4762-9485-aa638b5ffbb3",
   "metadata": {},
   "source": [
    "### Implementation notes:\n",
    "* Essential list of nodes to add: gene/protein, drug, disease, effect/phenotype\n",
    "* Good to have list: pathway, exposure (more on the enviromental side), anatomy (only relevant if we need spatial information, e.g. cancer on specific organ)\n",
    "* Maybe not for now?: biological_process, molecular_function, cellular_component\n",
    "*\n",
    "* following example inputs here for node normalizer service:\n",
    "* https://github.com/TranslatorSRI/NodeNormalization/blob/master/documentation/NodeNormalization.ipynb\n",
    "* \n",
    "* for MONDO codes: need to change format to MONDO:XXXXXXX, most cases those numbers are not in correct 7 digits so need to padding zeros\n",
    "* Discussion: how to deal with MONDO_grouped?\n",
    "* see question 1 below\n",
    "* for NCBI code: add NCBIGene: prefix\n",
    "* for HPO code: also need to change format to HP:XXXXXXX, 7 digits\n",
    "* for DrugBank: need to check if Translator accepts DB: codes\n",
    "* for GO: also need to change format to HP:XXXXXXX, 7 digits\n",
    "* for CTD: change format to CID:, add prefix\n",
    "* for REACTOME:\n",
    "* for UBERON: also need to change format to HP:XXXXXXX, 7 digits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "b4fa8690-2dee-46ad-b4ad-808b2b8dd5f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_index</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>node_name</th>\n",
       "      <th>node_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22117</th>\n",
       "      <td>22117</td>\n",
       "      <td>1507</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Growth abnormality</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22118</th>\n",
       "      <td>22118</td>\n",
       "      <td>107</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Renal cyst</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22119</th>\n",
       "      <td>22119</td>\n",
       "      <td>1</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>All</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22120</th>\n",
       "      <td>22120</td>\n",
       "      <td>5</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Mode of inheritance</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22121</th>\n",
       "      <td>22121</td>\n",
       "      <td>10460</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Abnormality of the female genitalia</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22122</th>\n",
       "      <td>22122</td>\n",
       "      <td>812</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Abnormal internal genitalia</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22123</th>\n",
       "      <td>22123</td>\n",
       "      <td>14</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Abnormality of the bladder</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22124</th>\n",
       "      <td>22124</td>\n",
       "      <td>2719</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Recurrent infections</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22125</th>\n",
       "      <td>22125</td>\n",
       "      <td>11277</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Abnormality of the urinary system physiology</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22126</th>\n",
       "      <td>22126</td>\n",
       "      <td>8684</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Aplasia/hypoplasia of the uterus</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22127</th>\n",
       "      <td>22127</td>\n",
       "      <td>10936</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Abnormality of the lower urinary tract</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22128</th>\n",
       "      <td>22128</td>\n",
       "      <td>9</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Functional abnormality of the bladder</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22129</th>\n",
       "      <td>22129</td>\n",
       "      <td>31064</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Impaired continence</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22130</th>\n",
       "      <td>22130</td>\n",
       "      <td>10955</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Dilatation of the bladder</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22131</th>\n",
       "      <td>22131</td>\n",
       "      <td>10461</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Abnormality of the male genitalia</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       node_index node_id         node_type  \\\n",
       "22117       22117    1507  effect/phenotype   \n",
       "22118       22118     107  effect/phenotype   \n",
       "22119       22119       1  effect/phenotype   \n",
       "22120       22120       5  effect/phenotype   \n",
       "22121       22121   10460  effect/phenotype   \n",
       "22122       22122     812  effect/phenotype   \n",
       "22123       22123      14  effect/phenotype   \n",
       "22124       22124    2719  effect/phenotype   \n",
       "22125       22125   11277  effect/phenotype   \n",
       "22126       22126    8684  effect/phenotype   \n",
       "22127       22127   10936  effect/phenotype   \n",
       "22128       22128       9  effect/phenotype   \n",
       "22129       22129   31064  effect/phenotype   \n",
       "22130       22130   10955  effect/phenotype   \n",
       "22131       22131   10461  effect/phenotype   \n",
       "\n",
       "                                          node_name node_source  \n",
       "22117                            Growth abnormality         HPO  \n",
       "22118                                    Renal cyst         HPO  \n",
       "22119                                           All         HPO  \n",
       "22120                           Mode of inheritance         HPO  \n",
       "22121           Abnormality of the female genitalia         HPO  \n",
       "22122                   Abnormal internal genitalia         HPO  \n",
       "22123                    Abnormality of the bladder         HPO  \n",
       "22124                          Recurrent infections         HPO  \n",
       "22125  Abnormality of the urinary system physiology         HPO  \n",
       "22126              Aplasia/hypoplasia of the uterus         HPO  \n",
       "22127        Abnormality of the lower urinary tract         HPO  \n",
       "22128         Functional abnormality of the bladder         HPO  \n",
       "22129                           Impaired continence         HPO  \n",
       "22130                     Dilatation of the bladder         HPO  \n",
       "22131             Abnormality of the male genitalia         HPO  "
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sanity_check_df = nodes_df[nodes_df['node_type'] == 'effect/phenotype']\n",
    "\n",
    "sanity_check_df.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "3fe0f646-ea99-49bf-90e1-19b29db6a586",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        node_index        node_id           node_type  \\\n",
      "65751        65751           3476             anatomy   \n",
      "114674      114674        1902367  biological_process   \n",
      "124851      124851          70913  cellular_component   \n",
      "99417        99417          23137             disease   \n",
      "19830        19830        DB06980                drug   \n",
      "87918        87918           9598    effect/phenotype   \n",
      "61700        61700        C091375            exposure   \n",
      "34206        34206         348793        gene/protein   \n",
      "115725      115725          36200  molecular_function   \n",
      "128395      128395  R-HSA-1368082             pathway   \n",
      "\n",
      "                                                node_name node_source  \n",
      "65751              respiratory system venous blood vessel      UBERON  \n",
      "114674  negative regulation of Notch signaling pathway...          GO  \n",
      "124851                                 Ddb1-Wdr21 complex          GO  \n",
      "99417                        feigenbaum Bergeron syndrome       MONDO  \n",
      "19830                 (2S)-2-(1H-indol-3-yl)hexanoic acid    DrugBank  \n",
      "87918   Symphalangism of the proximal phalanx of the 2...         HPO  \n",
      "61700                                       9-phenanthrol         CTD  \n",
      "34206                                               WDR53        NCBI  \n",
      "115725       3-ketosteroid 9-alpha-monooxygenase activity          GO  \n",
      "128395                     RORA activates gene expression    REACTOME  \n"
     ]
    }
   ],
   "source": [
    "## Sanity check on the current format of the file\n",
    "\n",
    "# Randomly select one row for each unique node_type\n",
    "sampled_df = nodes_df.groupby('node_type').sample(n=1, random_state=151)  # Set random_state for reproducibility\n",
    "\n",
    "print(sampled_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef1e59e6-5f1b-4b46-9a89-f2b25a9e28af",
   "metadata": {},
   "source": [
    "### Following are questions for data ingestion of the PrimeKG into Pharmocogenomics KG\n",
    "* **Question 1:** How to handle those grouped Mondo codes?\n",
    "* Remove them for now!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "cd0ddc35-b209-4f10-bba3-d66c07e777be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_index</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>node_name</th>\n",
       "      <th>node_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27158</th>\n",
       "      <td>27158</td>\n",
       "      <td>13924_12592_14672_13460_12591_12536_30861_8146...</td>\n",
       "      <td>disease</td>\n",
       "      <td>osteogenesis imperfecta</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27159</th>\n",
       "      <td>27159</td>\n",
       "      <td>11160_13119_13978_12060_12327_12670_13210_1106...</td>\n",
       "      <td>disease</td>\n",
       "      <td>autosomal recessive nonsyndromic deafness</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27160</th>\n",
       "      <td>27160</td>\n",
       "      <td>8099_12497_12498</td>\n",
       "      <td>disease</td>\n",
       "      <td>congenital stationary night blindness autosoma...</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27161</th>\n",
       "      <td>27161</td>\n",
       "      <td>14854_14293_14470_12380_11832_14603_14853_1176...</td>\n",
       "      <td>disease</td>\n",
       "      <td>autosomal dominant nonsyndromic deafness</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27162</th>\n",
       "      <td>27162</td>\n",
       "      <td>33202_32776_30905_33670_33200_32740_32732_3320...</td>\n",
       "      <td>disease</td>\n",
       "      <td>deafness, autosomal recessive</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27163</th>\n",
       "      <td>27163</td>\n",
       "      <td>11396_7422</td>\n",
       "      <td>disease</td>\n",
       "      <td>keratoderma hereditarium mutilans</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27164</th>\n",
       "      <td>27164</td>\n",
       "      <td>14828_14829_9454_13553_133</td>\n",
       "      <td>disease</td>\n",
       "      <td>immunodeficiency-centromeric instability-facia...</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27167</th>\n",
       "      <td>27167</td>\n",
       "      <td>9260_9261_9262_18149</td>\n",
       "      <td>disease</td>\n",
       "      <td>GM1 gangliosidosis</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27170</th>\n",
       "      <td>27170</td>\n",
       "      <td>14083_13288_12987_13287_13289_20729_14840_1329...</td>\n",
       "      <td>disease</td>\n",
       "      <td>agammaglobulinemia</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27173</th>\n",
       "      <td>27173</td>\n",
       "      <td>14986_14987_10351_10953_13565_9213_12565_13248...</td>\n",
       "      <td>disease</td>\n",
       "      <td>Fanconi anemia complementation group</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       node_index                                            node_id  \\\n",
       "27158       27158  13924_12592_14672_13460_12591_12536_30861_8146...   \n",
       "27159       27159  11160_13119_13978_12060_12327_12670_13210_1106...   \n",
       "27160       27160                                   8099_12497_12498   \n",
       "27161       27161  14854_14293_14470_12380_11832_14603_14853_1176...   \n",
       "27162       27162  33202_32776_30905_33670_33200_32740_32732_3320...   \n",
       "27163       27163                                         11396_7422   \n",
       "27164       27164                         14828_14829_9454_13553_133   \n",
       "27167       27167                               9260_9261_9262_18149   \n",
       "27170       27170  14083_13288_12987_13287_13289_20729_14840_1329...   \n",
       "27173       27173  14986_14987_10351_10953_13565_9213_12565_13248...   \n",
       "\n",
       "      node_type                                          node_name  \\\n",
       "27158   disease                            osteogenesis imperfecta   \n",
       "27159   disease          autosomal recessive nonsyndromic deafness   \n",
       "27160   disease  congenital stationary night blindness autosoma...   \n",
       "27161   disease           autosomal dominant nonsyndromic deafness   \n",
       "27162   disease                      deafness, autosomal recessive   \n",
       "27163   disease                  keratoderma hereditarium mutilans   \n",
       "27164   disease  immunodeficiency-centromeric instability-facia...   \n",
       "27167   disease                                 GM1 gangliosidosis   \n",
       "27170   disease                                 agammaglobulinemia   \n",
       "27173   disease               Fanconi anemia complementation group   \n",
       "\n",
       "         node_source  \n",
       "27158  MONDO_grouped  \n",
       "27159  MONDO_grouped  \n",
       "27160  MONDO_grouped  \n",
       "27161  MONDO_grouped  \n",
       "27162  MONDO_grouped  \n",
       "27163  MONDO_grouped  \n",
       "27164  MONDO_grouped  \n",
       "27167  MONDO_grouped  \n",
       "27170  MONDO_grouped  \n",
       "27173  MONDO_grouped  "
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_MONDO_grouped = nodes_df[nodes_df['node_source'] == 'MONDO_grouped']\n",
    "\n",
    "nodes_MONDO_grouped.head(10)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "bc92c727-121a-4736-ab8e-83cda50ce782",
   "metadata": {},
   "source": [
    "* **Questions 2:** How do or should we want to integrate those additional feature information of drug & disease into our KG?\n",
    "* Not using those features for now!!!\n",
    "* Sample information can be obtained from disease features\n",
    "    - mondo_id\tmondo_name\tgroup_id_bert\tgroup_name_bert\tmondo_definition\tumls_description\torphanet_definition\t**orphanet_prevalence**\torphanet_epidemiology\torphanet_clinical_description\t**orphanet_management_and_treatment**\tmayo_symptoms\tmayo_causes\tmayo_risk_factors\tmayo_complications\tmayo_prevention\tmayo_see_doc\n",
    "    - 12345\tacral peeling skin syndrome\t\t\tAcral peeling skin syndrome (PSS) is a form of PSS characterized by superficial peeling of the skin predominantly affecting the dorsa of the hands and feet.\t\tA rare peeling skin syndrome characterized by superficial peeling of the skin predominantly affecting the dorsa of the hands and feet.\t**<1/1000000**\tAcral PSS is rare, with approximately 40 cases described in the literature to date.\tThe disease manifests shortly after birth or in early childhood with superficial peeling on the palmar, plantar and dorsal surfaces of the hands and feet, that leaves residual painless erythema. Manual skin removal is also possible. Seasonal variations are generally observed. Heat, humidity, exposure to water and friction or minor trauma can induce exfoliation. The lesions are not painful and heal without scarring.\t**There is no effective treatment.** Emollients are often used to reduce skin peeling. Patients must avoid immersion in water and are recommended to use absorbing powders or aluminum antiperspirants.\t\t\t\t\t\t\n",
    "* Sample information can be obtained from drug features\n",
    "    - description\t**half_life**\tindication\t**mechanism_of_action**\tprotein_binding\tpharmacodynamics\tstate\tatc_1\tatc_2\tatc_3\tatc_4\tcategory\tgroup\tpathway\tmolecular_weight\ttpsa\t**clogp**\n",
    "    - Budesonide is a glucocorticoid that is a mix of the 22R and 22S epimer used to treat inflammatory conditions of the lungs and intestines such as asthma, COPD, Crohn's disease, and ulcerative colitis.\t**Budesonide has a plasma elimination half life of 2-3.6h. The terminal elimination half life in asthmatic children 4-6 years old is 2.3h.**\tBudesonide extended release capsules are indicated for the treatment and maintenance of mild to moderate Crohnâ€™s disease. Various inhaled budesonide products are indicated for prophylactic therapy in asthma and reducing exacerbations of COPD. A budesonide nasal spray is available over the counter for symptoms of hay fever and upper respiratory allergies. Extended release capsules are indicated to induce remission of mild to moderate ulcerative colitis and a rectal foam is used for mild to moderate distal ulcerative colitis.\t**The short term effects of corticosteroids are decreased vasodilation and permeability of capillaries, as well as decreased leukocyte migration to sites of inflammation.** Corticosteroids binding to the glucocorticoid receptor mediates changes in gene expression that lead to multiple downstream effects over hours to days.\tCorticosteroids are generally bound to corticosteroid binding globulin and serum albumin in plasma. Budesonide is 85-90% protein bound in plasma.\tBudesonide is a glucocorticoid used to treat respiratory and digestive conditions by reducing inflammation. It has a wide therapeutic index, as dosing varies highly from patient to patient. Patients should be counselled regarding the risk of hypercorticism and adrenal axis suppression.\tBudesonide is a solid.\tBudesonide is anatomically related to dermatologicals and respiratory system and respiratory system and alimentary tract and metabolism and respiratory system.\tBudesonide is in the therapeutic group of corticosteroids, dermatological preparations and nasal preparations and drugs for obstructive airway diseases and antidiarrheals, intestinal antiinflammatory/antiinfective agents and drugs for obstructive airway diseases.\tBudesonide is pharmacologically related to corticosteroids, plain and decongestants and other nasal preparations for topical use and adrenergics, inhalants and intestinal antiinflammatory agents and other drugs for obstructive airway diseases, inhalants.\tThe chemical and functional group of  is corticosteroids, potent (group iii) and corticosteroids and adrenergics in combination with corticosteroids or other drugs, excl. anticholinergics and corticosteroids acting locally and glucocorticoids.\tBudesonide is part of Adrenal Cortex Hormones ; Adrenals ; Agents to Treat Airway Disease ; Alimentary Tract and Metabolism ; Anti-Asthmatic Agents ; Anti-Inflammatory Agents ; Antidiarrheals, Intestinal Antiinflammatory/antiinfective Agents ; Autonomic Agents ; Bronchodilator Agents ; BSEP/ABCB11 Substrates ; Corticosteroid Hormone Receptor Agonists ; Corticosteroids ; Corticosteroids Acting Locally ; Corticosteroids for Systemic Use ; Corticosteroids, Dermatological Preparations ; Corticosteroids, Potent (Group III) ; Cytochrome P-450 CYP2A6 Inducers ; Cytochrome P-450 CYP2B6 Inducers ; Cytochrome P-450 CYP2B6 Inducers (strength unknown) ; Cytochrome P-450 CYP2C19 Inducers ; Cytochrome P-450 CYP2C19 Inducers (strength unknown) ; Cytochrome P-450 CYP2C8 Inducers ; Cytochrome P-450 CYP2C8 Inducers (strength unknown) ; Cytochrome P-450 CYP2C9 Inducers ; Cytochrome P-450 CYP2C9 Inducers (strength unknown) ; Cytochrome P-450 CYP3A Inducers ; Cytochrome P-450 CYP3A Substrates ; Cytochrome P-450 CYP3A4 Inducers ; Cytochrome P-450 CYP3A4 Inducers (strength unknown) ; Cytochrome P-450 CYP3A4 Substrates ; Cytochrome P-450 CYP3A5 Inducers ; Cytochrome P-450 CYP3A5 Inducers (moderate) ; Cytochrome P-450 Enzyme Inducers ; Cytochrome P-450 Substrates ; Dermatologicals ; Drugs for Obstructive Airway Diseases ; Drugs that are Mainly Renally Excreted ; Fused-Ring Compounds ; Hormones ; Hormones, Hormone Substitutes, and Hormone Antagonists ; Immunosuppressive Agents ; Intestinal Antiinflammatory Agents ; Nasal Preparations ; OAT3/SLC22A8 Substrates ; P-glycoprotein substrates ; Peripheral Nervous System Agents ; Pregnanes ; Pregnenediones ; Pregnenes ; Respiratory System Agents ; Steroids.\tBudesonide is approved.\t\tThe molecular weight is 430.54.\tBudesonide has a topological polar surface area of 93.06.\t**The log p value of  is 2.9.**\n",
    "    - clogp: a predicted value for the partition coefficient (LogP) of a molecule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "ebf19e8c-cadf-44d7-8d5b-cf860e1ddfd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## Read each individual csv files\n",
    "# disease_features_df = pd.read_csv(raw_files_path + 'disease_features.csv')\n",
    "\n",
    "# disease_features_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "e9eb6976-b5a2-4f66-90b3-fc7f604dfe3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## Read each individual csv files\n",
    "# drug_features_df = pd.read_csv(raw_files_path + 'drug_features.csv')\n",
    "\n",
    "# drug_features_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "1cac2a26-d39f-4699-a59c-862fe69905eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "68019\n"
     ]
    }
   ],
   "source": [
    "## Select only the 4 must have node types\n",
    "must_have_list = ['gene/protein', 'drug', 'disease', 'effect/phenotype']\n",
    "nodes_select_df = nodes_df[nodes_df['node_type'].isin(must_have_list)]\n",
    "\n",
    "## count rows\n",
    "unique_node_count = nodes_select_df['node_index'].nunique()\n",
    "print(unique_node_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "8a3404c9-7dac-4fb9-a601-82c21901ec81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_index</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>node_name</th>\n",
       "      <th>node_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3772</th>\n",
       "      <td>3772</td>\n",
       "      <td>5156</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36722</th>\n",
       "      <td>36722</td>\n",
       "      <td>5156</td>\n",
       "      <td>disease</td>\n",
       "      <td>encephalomyelitis</td>\n",
       "      <td>MONDO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86127</th>\n",
       "      <td>86127</td>\n",
       "      <td>5156</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Hypoplastic left atrium</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       node_index node_id         node_type                node_name  \\\n",
       "3772         3772    5156      gene/protein                   PDGFRA   \n",
       "36722       36722    5156           disease        encephalomyelitis   \n",
       "86127       86127    5156  effect/phenotype  Hypoplastic left atrium   \n",
       "\n",
       "      node_source  \n",
       "3772         NCBI  \n",
       "36722       MONDO  \n",
       "86127         HPO  "
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sanity_check_df = nodes_select_df[nodes_select_df['node_id'] == '5156']\n",
    "\n",
    "sanity_check_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "e413ea01-a044-48b4-b020-b894f0f3fe96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "66752\n"
     ]
    }
   ],
   "source": [
    "## let's exclude those 'MONDO_grouped' rows for now\n",
    "nodes_select_no_MONDO_grouped_df = nodes_select_df[nodes_select_df['node_source'] != 'MONDO_grouped']\n",
    "\n",
    "## count rows\n",
    "unique_node_count = nodes_select_no_MONDO_grouped_df['node_index'].nunique()\n",
    "print(unique_node_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "2889e19c-a5a8-424a-b215-9f186f560bd9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_17326/2807197227.py:7: DtypeWarning: Columns (3,8) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  edges_df = pd.read_csv(raw_files_path + 'kg.csv')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8100498\n",
      "4009644\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>0</td>\n",
       "      <td>9796</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PHYHIP</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>8889</td>\n",
       "      <td>56992</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>KIF15</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>1</td>\n",
       "      <td>7918</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GPANK1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>2798</td>\n",
       "      <td>9240</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PNMA1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>2</td>\n",
       "      <td>8233</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>5646</td>\n",
       "      <td>23548</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TTC33</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>3</td>\n",
       "      <td>4899</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NRF1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>11592</td>\n",
       "      <td>11253</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MAN1B1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>4</td>\n",
       "      <td>5297</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PI4KA</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>2122</td>\n",
       "      <td>8601</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>RGS20</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          relation display_relation  x_index  x_id        x_type  x_name  \\\n",
       "0  protein_protein              ppi        0  9796  gene/protein  PHYHIP   \n",
       "1  protein_protein              ppi        1  7918  gene/protein  GPANK1   \n",
       "2  protein_protein              ppi        2  8233  gene/protein   ZRSR2   \n",
       "3  protein_protein              ppi        3  4899  gene/protein    NRF1   \n",
       "4  protein_protein              ppi        4  5297  gene/protein   PI4KA   \n",
       "\n",
       "  x_source  y_index   y_id        y_type  y_name y_source  \n",
       "0     NCBI     8889  56992  gene/protein   KIF15     NCBI  \n",
       "1     NCBI     2798   9240  gene/protein   PNMA1     NCBI  \n",
       "2     NCBI     5646  23548  gene/protein   TTC33     NCBI  \n",
       "3     NCBI    11592  11253  gene/protein  MAN1B1     NCBI  \n",
       "4     NCBI     2122   8601  gene/protein   RGS20     NCBI  "
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get the list of unique values in column node_id of nodes_select_no_MONDO_grouped_df\n",
    "## since node_id neededs to combine with node_source as prefix to obtain a unique identifier\n",
    "## see above example of same \"node_id\" in multiple vocabularies\n",
    "unique_node_ids = nodes_select_no_MONDO_grouped_df['node_index'].unique().tolist()\n",
    "\n",
    "## Read each individual csv files\n",
    "edges_df = pd.read_csv(raw_files_path + 'kg.csv')\n",
    "\n",
    "## count rows\n",
    "print(len(edges_df))\n",
    "\n",
    "## Filter to only include edges that are in the selected node_id list\n",
    "edges_filtered_df = edges_df[edges_df['x_index'].isin(unique_node_ids)]\n",
    "edges_filtered2_df = edges_filtered_df[edges_filtered_df['y_index'].isin(unique_node_ids)]\n",
    "\n",
    "## count rows\n",
    "print(len(edges_filtered2_df))\n",
    "\n",
    "edges_filtered2_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "7521afb6-4256-4656-814d-da22573393c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All possible relation are here:  relation\n",
      "drug_drug                     2672628\n",
      "protein_protein                642150\n",
      "disease_phenotype_positive     222000\n",
      "drug_effect                    129568\n",
      "disease_protein                127408\n",
      "drug_protein                    51306\n",
      "contraindication                49924\n",
      "disease_disease                 49452\n",
      "phenotype_phenotype             37472\n",
      "indication                      15318\n",
      "phenotype_protein                6660\n",
      "off-label use                    4372\n",
      "disease_phenotype_negative       1386\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check unique relation\n",
    "unique_relation_values = edges_filtered2_df['relation'].value_counts()\n",
    "print(\"All possible relation are here: \" , unique_relation_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "14347f18-1073-4cee-8574-9e51e976f19c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All possible display relation are here:  display_relation\n",
      "synergistic interaction    2672628\n",
      "ppi                         642150\n",
      "phenotype present           222000\n",
      "associated with             134068\n",
      "side effect                 129568\n",
      "parent-child                 86924\n",
      "contraindication             49924\n",
      "target                       32760\n",
      "indication                   15318\n",
      "enzyme                       10634\n",
      "transporter                   6184\n",
      "off-label use                 4372\n",
      "carrier                       1728\n",
      "phenotype absent              1386\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check unique relation\n",
    "unique_relation_values = edges_filtered2_df['display_relation'].value_counts()\n",
    "print(\"All possible display relation are here: \" , unique_relation_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "e5d7f3c4-2609-4543-80a7-5116ba0a906a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All possible node types are here:  {'gene/protein', 'drug', 'disease', 'effect/phenotype'}\n"
     ]
    }
   ],
   "source": [
    "## check unique relation\n",
    "unique_type_values = set(list(edges_filtered2_df['x_type'].unique()) + list(edges_filtered2_df['y_type'].unique()))\n",
    "print(\"All possible node types are here: \" , unique_type_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "cef0fb11-86ee-4f76-8401-123356332d12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All possible node sources are here:  {'HPO', 'MONDO', 'NCBI', 'DrugBank'}\n"
     ]
    }
   ],
   "source": [
    "## check unique relation\n",
    "unique_type_values = set(list(edges_filtered2_df['x_source'].unique()) + list(edges_filtered2_df['y_source'].unique()))\n",
    "print(\"All possible node sources are here: \" , unique_type_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "ee91a835-4310-4e55-a8ff-910617b11849",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22\n",
      "                   x_type            y_type                    relation\n",
      "0            gene/protein      gene/protein             protein_protein\n",
      "321075               drug      gene/protein                drug_protein\n",
      "346728               drug           disease            contraindication\n",
      "346730               drug           disease                  indication\n",
      "347016               drug           disease               off-label use\n",
      "389359               drug              drug                   drug_drug\n",
      "3061987      gene/protein  effect/phenotype           phenotype_protein\n",
      "3065317  effect/phenotype  effect/phenotype         phenotype_phenotype\n",
      "3084066           disease  effect/phenotype  disease_phenotype_negative\n",
      "3085282           disease  effect/phenotype  disease_phenotype_positive\n",
      "3234966  effect/phenotype           disease  disease_phenotype_positive\n",
      "3236466      gene/protein           disease             disease_protein\n",
      "3315993           disease           disease             disease_disease\n",
      "3348335              drug  effect/phenotype                 drug_effect\n",
      "5707875      gene/protein              drug                drug_protein\n",
      "5733528           disease              drug            contraindication\n",
      "5733530           disease              drug                  indication\n",
      "5733816           disease              drug               off-label use\n",
      "5776159  effect/phenotype      gene/protein           phenotype_protein\n",
      "5798238  effect/phenotype           disease  disease_phenotype_negative\n",
      "5950600           disease      gene/protein             disease_protein\n",
      "6062173  effect/phenotype              drug                 drug_effect\n"
     ]
    }
   ],
   "source": [
    "## get all unqiue combinations of subject, object, and relation\n",
    "## check all unique combinations between subject & object\n",
    "from itertools import combinations\n",
    "\n",
    "# Get all combinations of 3 column names\n",
    "col_combos = ['x_type', 'y_type', 'relation']\n",
    "\n",
    "unique_combinations = edges_filtered2_df[col_combos].drop_duplicates()\n",
    "\n",
    "print(len(unique_combinations))\n",
    "print(unique_combinations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "fc73d4b7-b291-4317-a9c4-1eef16f105c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n",
      "                   x_type            y_type         display_relation\n",
      "0            gene/protein      gene/protein                      ppi\n",
      "321075               drug      gene/protein                  carrier\n",
      "321939               drug      gene/protein                   enzyme\n",
      "327256               drug      gene/protein                   target\n",
      "343636               drug      gene/protein              transporter\n",
      "346728               drug           disease         contraindication\n",
      "346730               drug           disease               indication\n",
      "347016               drug           disease            off-label use\n",
      "389359               drug              drug  synergistic interaction\n",
      "3061987      gene/protein  effect/phenotype          associated with\n",
      "3065317  effect/phenotype  effect/phenotype             parent-child\n",
      "3084066           disease  effect/phenotype         phenotype absent\n",
      "3085282           disease  effect/phenotype        phenotype present\n",
      "3234966  effect/phenotype           disease        phenotype present\n",
      "3236466      gene/protein           disease          associated with\n",
      "3315993           disease           disease             parent-child\n",
      "3348335              drug  effect/phenotype              side effect\n",
      "5707875      gene/protein              drug                  carrier\n",
      "5708739      gene/protein              drug                   enzyme\n",
      "5714056      gene/protein              drug                   target\n",
      "5730436      gene/protein              drug              transporter\n",
      "5733528           disease              drug         contraindication\n",
      "5733530           disease              drug               indication\n",
      "5733816           disease              drug            off-label use\n",
      "5776159  effect/phenotype      gene/protein          associated with\n",
      "5798238  effect/phenotype           disease         phenotype absent\n",
      "5950600           disease      gene/protein          associated with\n",
      "6062173  effect/phenotype              drug              side effect\n"
     ]
    }
   ],
   "source": [
    "## get all unqiue combinations of subject, object, and relation\n",
    "## check all unique combinations between subject & object\n",
    "from itertools import combinations\n",
    "\n",
    "# Get all combinations of 3 column names\n",
    "col_combos = ['x_type', 'y_type', 'display_relation']\n",
    "\n",
    "unique_combinations = edges_filtered2_df[col_combos].drop_duplicates()\n",
    "\n",
    "print(len(unique_combinations))\n",
    "print(unique_combinations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "7ed86f01-179d-41cd-a299-96e1aa7c23f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## uncertain mapping\n",
    "##  drug      gene/protein                   enzyme\n",
    "## gene/protein              drug                   enzyme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "29e4857b-ad80-4f31-835c-fb9bf0bb3632",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [relation, display_relation, x_index, x_id, x_type, x_name, x_source, y_index, y_id, y_type, y_name, y_source]\n",
       "Index: []"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Sanity check anatomy_anatomy relations\n",
    "## should return empty df\n",
    "sanity_check_df = edges_filtered2_df[edges_filtered2_df['relation'] == 'anatomy_anatomy']\n",
    "sanity_check_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4edaaee2-a012-4291-91ff-63d9e4946481",
   "metadata": {},
   "source": [
    "### Investigate those reverse order pairs of subject & objects\n",
    "* seem like in raw PrimeKG file, they kept a reverse copy of all those edges\n",
    "* Strategy is to throw away those pairs that \"contradict\" to what defined in the biolink associations & predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "6e073027-3227-44d6-8470-00ce66b188b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3092\n"
     ]
    }
   ],
   "source": [
    "sanity_check_df = edges_filtered2_df[ (edges_filtered2_df['display_relation'] == 'transporter') & (edges_filtered2_df['x_type'] == 'drug')]\n",
    "\n",
    "print(len(sanity_check_df))\n",
    "# sanity_check_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "093a5c08-471d-46a7-9dbf-66edeb44a1a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3092\n"
     ]
    }
   ],
   "source": [
    "sanity_check_df = edges_filtered2_df[ (edges_filtered2_df['display_relation'] == 'transporter') & (edges_filtered2_df['x_type'] == 'gene/protein')]\n",
    "\n",
    "print(len(sanity_check_df))\n",
    "# sanity_check_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "e98dc05a-cdf0-4acf-8df1-f3c791a6a6db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16380\n"
     ]
    }
   ],
   "source": [
    "sanity_check_df = edges_filtered2_df[ (edges_filtered2_df['display_relation'] == 'target') & (edges_filtered2_df['x_type'] == 'drug')]\n",
    "\n",
    "print(len(sanity_check_df))\n",
    "# sanity_check_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "30a0efbe-7586-480e-be88-b2a1f1673766",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16380\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5714056</th>\n",
       "      <td>drug_protein</td>\n",
       "      <td>target</td>\n",
       "      <td>13174</td>\n",
       "      <td>3067</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>HDC</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>15906</td>\n",
       "      <td>DB00114</td>\n",
       "      <td>drug</td>\n",
       "      <td>Pyridoxal phosphate</td>\n",
       "      <td>DrugBank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5714057</th>\n",
       "      <td>drug_protein</td>\n",
       "      <td>target</td>\n",
       "      <td>13174</td>\n",
       "      <td>3067</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>HDC</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>15907</td>\n",
       "      <td>DB00117</td>\n",
       "      <td>drug</td>\n",
       "      <td>Histidine</td>\n",
       "      <td>DrugBank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5714058</th>\n",
       "      <td>drug_protein</td>\n",
       "      <td>target</td>\n",
       "      <td>13982</td>\n",
       "      <td>27165</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GLS2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>15908</td>\n",
       "      <td>DB00142</td>\n",
       "      <td>drug</td>\n",
       "      <td>Glutamic acid</td>\n",
       "      <td>DrugBank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5714059</th>\n",
       "      <td>drug_protein</td>\n",
       "      <td>target</td>\n",
       "      <td>476</td>\n",
       "      <td>2162</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>F13A1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>15909</td>\n",
       "      <td>DB02340</td>\n",
       "      <td>drug</td>\n",
       "      <td>N-Acetyl-Serine</td>\n",
       "      <td>DrugBank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5714060</th>\n",
       "      <td>drug_protein</td>\n",
       "      <td>target</td>\n",
       "      <td>476</td>\n",
       "      <td>2162</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>F13A1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>15910</td>\n",
       "      <td>DB11300</td>\n",
       "      <td>drug</td>\n",
       "      <td>Thrombin</td>\n",
       "      <td>DrugBank</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             relation display_relation  x_index   x_id        x_type x_name  \\\n",
       "5714056  drug_protein           target    13174   3067  gene/protein    HDC   \n",
       "5714057  drug_protein           target    13174   3067  gene/protein    HDC   \n",
       "5714058  drug_protein           target    13982  27165  gene/protein   GLS2   \n",
       "5714059  drug_protein           target      476   2162  gene/protein  F13A1   \n",
       "5714060  drug_protein           target      476   2162  gene/protein  F13A1   \n",
       "\n",
       "        x_source  y_index     y_id y_type               y_name  y_source  \n",
       "5714056     NCBI    15906  DB00114   drug  Pyridoxal phosphate  DrugBank  \n",
       "5714057     NCBI    15907  DB00117   drug            Histidine  DrugBank  \n",
       "5714058     NCBI    15908  DB00142   drug        Glutamic acid  DrugBank  \n",
       "5714059     NCBI    15909  DB02340   drug      N-Acetyl-Serine  DrugBank  \n",
       "5714060     NCBI    15910  DB11300   drug             Thrombin  DrugBank  "
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sanity_check_df = edges_filtered2_df[ (edges_filtered2_df['display_relation'] == 'target') & (edges_filtered2_df['x_type'] == 'gene/protein')]\n",
    "\n",
    "print(len(sanity_check_df))\n",
    "sanity_check_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "d3fe1bb5-766a-4cac-a87d-bc2f80f2e752",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2186\n"
     ]
    }
   ],
   "source": [
    "sanity_check_df = edges_filtered2_df[ (edges_filtered2_df['display_relation'] == 'off-label use') & (edges_filtered2_df['x_type'] == 'drug')]\n",
    "\n",
    "print(len(sanity_check_df))\n",
    "# sanity_check_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "bea37fa0-2259-4b31-897c-d06b7dd261de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2186\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5733816</th>\n",
       "      <td>off-label use</td>\n",
       "      <td>off-label use</td>\n",
       "      <td>33577</td>\n",
       "      <td>5044</td>\n",
       "      <td>disease</td>\n",
       "      <td>hypertensive disorder</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>14257</td>\n",
       "      <td>DB00903</td>\n",
       "      <td>drug</td>\n",
       "      <td>Etacrynic acid</td>\n",
       "      <td>DrugBank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5734096</th>\n",
       "      <td>off-label use</td>\n",
       "      <td>off-label use</td>\n",
       "      <td>33577</td>\n",
       "      <td>5044</td>\n",
       "      <td>disease</td>\n",
       "      <td>hypertensive disorder</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>14668</td>\n",
       "      <td>DB00887</td>\n",
       "      <td>drug</td>\n",
       "      <td>Bumetanide</td>\n",
       "      <td>DrugBank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5734351</th>\n",
       "      <td>off-label use</td>\n",
       "      <td>off-label use</td>\n",
       "      <td>38121</td>\n",
       "      <td>5391</td>\n",
       "      <td>disease</td>\n",
       "      <td>restless legs syndrome</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>14321</td>\n",
       "      <td>DB01235</td>\n",
       "      <td>drug</td>\n",
       "      <td>Levodopa</td>\n",
       "      <td>DrugBank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5734353</th>\n",
       "      <td>off-label use</td>\n",
       "      <td>off-label use</td>\n",
       "      <td>38121</td>\n",
       "      <td>5391</td>\n",
       "      <td>disease</td>\n",
       "      <td>restless legs syndrome</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>19207</td>\n",
       "      <td>DB00190</td>\n",
       "      <td>drug</td>\n",
       "      <td>Carbidopa</td>\n",
       "      <td>DrugBank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5734665</th>\n",
       "      <td>off-label use</td>\n",
       "      <td>off-label use</td>\n",
       "      <td>28396</td>\n",
       "      <td>7803</td>\n",
       "      <td>disease</td>\n",
       "      <td>multiple system atrophy</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>15003</td>\n",
       "      <td>DB01380</td>\n",
       "      <td>drug</td>\n",
       "      <td>Cortisone acetate</td>\n",
       "      <td>DrugBank</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              relation display_relation  x_index  x_id   x_type  \\\n",
       "5733816  off-label use    off-label use    33577  5044  disease   \n",
       "5734096  off-label use    off-label use    33577  5044  disease   \n",
       "5734351  off-label use    off-label use    38121  5391  disease   \n",
       "5734353  off-label use    off-label use    38121  5391  disease   \n",
       "5734665  off-label use    off-label use    28396  7803  disease   \n",
       "\n",
       "                          x_name x_source  y_index     y_id y_type  \\\n",
       "5733816    hypertensive disorder    MONDO    14257  DB00903   drug   \n",
       "5734096    hypertensive disorder    MONDO    14668  DB00887   drug   \n",
       "5734351   restless legs syndrome    MONDO    14321  DB01235   drug   \n",
       "5734353   restless legs syndrome    MONDO    19207  DB00190   drug   \n",
       "5734665  multiple system atrophy    MONDO    15003  DB01380   drug   \n",
       "\n",
       "                    y_name  y_source  \n",
       "5733816     Etacrynic acid  DrugBank  \n",
       "5734096         Bumetanide  DrugBank  \n",
       "5734351           Levodopa  DrugBank  \n",
       "5734353          Carbidopa  DrugBank  \n",
       "5734665  Cortisone acetate  DrugBank  "
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sanity_check_df = edges_filtered2_df[ (edges_filtered2_df['display_relation'] == 'off-label use') & (edges_filtered2_df['x_type'] == 'disease')]\n",
    "\n",
    "print(len(sanity_check_df))\n",
    "sanity_check_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93f2f05a-6dbc-43dd-af1e-585635f3d79f",
   "metadata": {},
   "source": [
    "### Now handles the different type of nodes differently\n",
    "* following example inputs here for node normalizer service: https://github.com/TranslatorSRI/NodeNormalization/blob/master/documentation/NodeNormalization.ipynb\n",
    "* Drug: DRUGBANK:DB09145\n",
    "* Disease: MONDO:0004976\n",
    "* Gene / Protein: NCBIGene:9496\n",
    "* Effect / Phenotype: HP:0007354\n",
    "* Need to find a way to add zeros padding for MONDO and HP codes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "177030d3-2975-4bd3-86c9-2b017022b51d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Start with check MONDO or HPO in either x_source or y_source\n",
    "## this is just for test run, when the code is done, directly run on full df\n",
    "mondo_or_hpo = ['MONDO', 'HPO']\n",
    "edges_mondo_hpo_df = edges_filtered2_df[ (edges_filtered2_df['x_source'].isin(mondo_or_hpo)) | (edges_filtered2_df['y_source'].isin(mondo_or_hpo))]\n",
    "\n",
    "# Randomly select one row for each unique combination\n",
    "sampled_df = edges_mondo_hpo_df.groupby(['x_source', 'y_source']).sample(n=1, random_state=151)  # Set random_state for reproducibility\n",
    "# sampled_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "e6357368-ee74-4446-8f75-02daeaf8c23e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "      <th>subject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1362595</th>\n",
       "      <td>drug_drug</td>\n",
       "      <td>synergistic interaction</td>\n",
       "      <td>21014</td>\n",
       "      <td>DB13648</td>\n",
       "      <td>drug</td>\n",
       "      <td>Alcuronium</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>14614</td>\n",
       "      <td>DB06738</td>\n",
       "      <td>drug</td>\n",
       "      <td>Ketobemidone</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>DRUGBANK:DB13648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5891379</th>\n",
       "      <td>disease_phenotype_positive</td>\n",
       "      <td>phenotype present</td>\n",
       "      <td>88949</td>\n",
       "      <td>11800</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Midface retrusion</td>\n",
       "      <td>HPO</td>\n",
       "      <td>31464</td>\n",
       "      <td>9841</td>\n",
       "      <td>disease</td>\n",
       "      <td>PEHO syndrome</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>HP:0011800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3153485</th>\n",
       "      <td>disease_phenotype_positive</td>\n",
       "      <td>phenotype present</td>\n",
       "      <td>31611</td>\n",
       "      <td>14892</td>\n",
       "      <td>disease</td>\n",
       "      <td>micrognathia-recurrent infections-behavioral a...</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>22715</td>\n",
       "      <td>750</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Delayed speech and language development</td>\n",
       "      <td>HPO</td>\n",
       "      <td>MONDO:0014892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116373</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>2071</td>\n",
       "      <td>27074</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>LAMP3</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>10788</td>\n",
       "      <td>347733</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TUBB2B</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBIGene:27074</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           relation         display_relation  x_index  \\\n",
       "1362595                   drug_drug  synergistic interaction    21014   \n",
       "5891379  disease_phenotype_positive        phenotype present    88949   \n",
       "3153485  disease_phenotype_positive        phenotype present    31611   \n",
       "116373              protein_protein                      ppi     2071   \n",
       "\n",
       "            x_id            x_type  \\\n",
       "1362595  DB13648              drug   \n",
       "5891379    11800  effect/phenotype   \n",
       "3153485    14892           disease   \n",
       "116373     27074      gene/protein   \n",
       "\n",
       "                                                    x_name  x_source  y_index  \\\n",
       "1362595                                         Alcuronium  DrugBank    14614   \n",
       "5891379                                  Midface retrusion       HPO    31464   \n",
       "3153485  micrognathia-recurrent infections-behavioral a...     MONDO    22715   \n",
       "116373                                               LAMP3      NCBI    10788   \n",
       "\n",
       "            y_id            y_type                                   y_name  \\\n",
       "1362595  DB06738              drug                             Ketobemidone   \n",
       "5891379     9841           disease                            PEHO syndrome   \n",
       "3153485      750  effect/phenotype  Delayed speech and language development   \n",
       "116373    347733      gene/protein                                   TUBB2B   \n",
       "\n",
       "         y_source           subject  \n",
       "1362595  DrugBank  DRUGBANK:DB13648  \n",
       "5891379     MONDO        HP:0011800  \n",
       "3153485       HPO     MONDO:0014892  \n",
       "116373       NCBI    NCBIGene:27074  "
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## work directly on the full dataframe\n",
    "## first loop x_id \n",
    "## to add prefix MONDO or HP for rows with sources \"MONDO\" or \"HPO\" and padding to 7 digits as required by example\n",
    "\n",
    "## deep copy a df\n",
    "edges_deep_copy_df = edges_filtered2_df.copy(deep=True)\n",
    "\n",
    "## Prefix mapping for selected sources\n",
    "prefix_map = {\n",
    "    'HPO': 'HP:',\n",
    "    'MONDO': 'MONDO:',\n",
    "    'DrugBank': 'DRUGBANK:', \n",
    "    'NCBI': 'NCBIGene:'\n",
    "}\n",
    "\n",
    "## Initialize new column with default values (original id as string)\n",
    "edges_deep_copy_df['subject'] = edges_deep_copy_df['x_id'].astype(str)\n",
    "\n",
    "## Mask for HPO and MONDO â†’ prefix + zero-padded id\n",
    "mask_hpo_mondo = edges_deep_copy_df['x_source'].isin(['HPO', 'MONDO'])\n",
    "edges_deep_copy_df.loc[mask_hpo_mondo, 'subject'] = (\n",
    "    edges_deep_copy_df.loc[mask_hpo_mondo, 'x_source'].map(prefix_map) +\n",
    "    edges_deep_copy_df.loc[mask_hpo_mondo, 'x_id'].astype(str).str.zfill(7)\n",
    ")\n",
    "\n",
    "## Mask for DrugBank â†’ prefix only, no padding\n",
    "mask_drugbank_ncbi = edges_deep_copy_df['x_source'].isin(['DrugBank', 'NCBI'])\n",
    "edges_deep_copy_df.loc[mask_drugbank_ncbi, 'subject'] = (\n",
    "    edges_deep_copy_df.loc[mask_drugbank_ncbi, 'x_source'].map(prefix_map) +\n",
    "    edges_deep_copy_df.loc[mask_drugbank_ncbi, 'x_id'].astype(str)\n",
    ")\n",
    "\n",
    "sampled_df = edges_deep_copy_df.groupby(['x_source']).sample(n=1, random_state=151)  # Set random_state for reproducibility\n",
    "sampled_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "e8134742-1498-4878-8436-bcf6431ca013",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1423055</th>\n",
       "      <td>drug_drug</td>\n",
       "      <td>synergistic interaction</td>\n",
       "      <td>15103</td>\n",
       "      <td>DB00496</td>\n",
       "      <td>drug</td>\n",
       "      <td>Darifenacin</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>15441</td>\n",
       "      <td>DB00402</td>\n",
       "      <td>drug</td>\n",
       "      <td>Eszopiclone</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>DRUGBANK:DB00496</td>\n",
       "      <td>DRUGBANK:DB00402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3200343</th>\n",
       "      <td>disease_phenotype_positive</td>\n",
       "      <td>phenotype present</td>\n",
       "      <td>28319</td>\n",
       "      <td>8900</td>\n",
       "      <td>disease</td>\n",
       "      <td>camptodactyly with fibrous tissue hyperplasia ...</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>22650</td>\n",
       "      <td>100490</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Camptodactyly of finger</td>\n",
       "      <td>HPO</td>\n",
       "      <td>MONDO:0008900</td>\n",
       "      <td>HP:0100490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3249032</th>\n",
       "      <td>disease_protein</td>\n",
       "      <td>associated with</td>\n",
       "      <td>1749</td>\n",
       "      <td>7150</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>84244</td>\n",
       "      <td>16267</td>\n",
       "      <td>disease</td>\n",
       "      <td>undifferentiated carcinoma of the corpus uteri</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>NCBIGene:7150</td>\n",
       "      <td>MONDO:0016267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116373</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>2071</td>\n",
       "      <td>27074</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>LAMP3</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>10788</td>\n",
       "      <td>347733</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TUBB2B</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBIGene:27074</td>\n",
       "      <td>NCBIGene:347733</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           relation         display_relation  x_index  \\\n",
       "1423055                   drug_drug  synergistic interaction    15103   \n",
       "3200343  disease_phenotype_positive        phenotype present    28319   \n",
       "3249032             disease_protein          associated with     1749   \n",
       "116373              protein_protein                      ppi     2071   \n",
       "\n",
       "            x_id        x_type  \\\n",
       "1423055  DB00496          drug   \n",
       "3200343     8900       disease   \n",
       "3249032     7150  gene/protein   \n",
       "116373     27074  gene/protein   \n",
       "\n",
       "                                                    x_name  x_source  y_index  \\\n",
       "1423055                                        Darifenacin  DrugBank    15441   \n",
       "3200343  camptodactyly with fibrous tissue hyperplasia ...     MONDO    22650   \n",
       "3249032                                               TOP1      NCBI    84244   \n",
       "116373                                               LAMP3      NCBI    10788   \n",
       "\n",
       "            y_id            y_type  \\\n",
       "1423055  DB00402              drug   \n",
       "3200343   100490  effect/phenotype   \n",
       "3249032    16267           disease   \n",
       "116373    347733      gene/protein   \n",
       "\n",
       "                                                 y_name  y_source  \\\n",
       "1423055                                     Eszopiclone  DrugBank   \n",
       "3200343                         Camptodactyly of finger       HPO   \n",
       "3249032  undifferentiated carcinoma of the corpus uteri     MONDO   \n",
       "116373                                           TUBB2B      NCBI   \n",
       "\n",
       "                  subject            object  \n",
       "1423055  DRUGBANK:DB00496  DRUGBANK:DB00402  \n",
       "3200343     MONDO:0008900        HP:0100490  \n",
       "3249032     NCBIGene:7150     MONDO:0016267  \n",
       "116373     NCBIGene:27074   NCBIGene:347733  "
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## similarly change for y_id\n",
    "## Initialize new column with default values (original id as string)\n",
    "edges_deep_copy_df['object'] = edges_deep_copy_df['y_id'].astype(str)\n",
    "\n",
    "## Mask for HPO and MONDO â†’ prefix + zero-padded id\n",
    "mask_hpo_mondo = edges_deep_copy_df['y_source'].isin(['HPO', 'MONDO'])\n",
    "edges_deep_copy_df.loc[mask_hpo_mondo, 'object'] = (\n",
    "    edges_deep_copy_df.loc[mask_hpo_mondo, 'y_source'].map(prefix_map) +\n",
    "    edges_deep_copy_df.loc[mask_hpo_mondo, 'y_id'].astype(str).str.zfill(7)\n",
    ")\n",
    "\n",
    "## Mask for DrugBank â†’ prefix only, no padding\n",
    "mask_drugbank_ncbi = edges_deep_copy_df['y_source'].isin(['DrugBank', 'NCBI'])\n",
    "edges_deep_copy_df.loc[mask_drugbank_ncbi, 'object'] = (\n",
    "    edges_deep_copy_df.loc[mask_drugbank_ncbi, 'y_source'].map(prefix_map) +\n",
    "    edges_deep_copy_df.loc[mask_drugbank_ncbi, 'y_id'].astype(str)\n",
    ")\n",
    "\n",
    "sampled_df = edges_deep_copy_df.groupby(['y_source']).sample(n=1, random_state=151)  # Set random_state for reproducibility\n",
    "sampled_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01704dc6-00d7-4d83-ac52-e4075929692a",
   "metadata": {},
   "source": [
    "* **Questions 3:** for effect/phenotype nodes which biolink node type to map to\n",
    "* following are detail on candidates\n",
    "* DiseaseOrPhenotypicFeature\n",
    "  - Either one of a disease or an individual phenotypic feature. Some knowledge resources such as Monarch treat these as distinct, others such as MESH conflate. Please see definitions of phenotypic feature and disease in this model for their independent descriptions. This class is helpful to enforce domains and ranges that may involve either a disease or a phenotypic feature.\n",
    "* DiseaseOrPhenotypicFeatureExposure\n",
    "  - A disease or phenotypic feature state, when viewed as an exposure, represents an precondition, leading to or influencing an outcome, e.g. HIV predisposing an individual to infections; a relative deficiency of skin pigmentation predisposing an individual to skin cancer.\n",
    "* DiseaseOrPhenotypicFeatureOutcome\n",
    "  - Physiological outcomes resulting from an exposure event which is the manifestation of a disease or other characteristic phenotype."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21ce5994-e044-44a5-9885-6088c1a8a398",
   "metadata": {},
   "source": [
    "### Load the Biolink category and predicate dictionary for mapping subject, object, and predicate types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "34456c0d-5b86-4796-bc47-56df11c41ca7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Date of last update:  2025-09-04\n",
      "Order is to always process Node/category map first, since the Edeg/predicate map depends on biolink-complainat node values\n",
      "-----------------------------------------------------------------------------------------------------------------------------\n",
      "Dictionary: category_map, Key template: Subject_category or Object_category\n",
      "------------------------------------------------------------------------------------------\n",
      "Dictionary: predicate_map, Key template: (Subject_category, Object_category, Predicate)\n"
     ]
    }
   ],
   "source": [
    "## Load the Biolink category and predicate dictionary for mapping subject, object, and predicate types\n",
    "%run ./Biolink_category_and_predication_dictionary.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "c542d9b7-2fac-4d26-8877-bae52fac512c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3157330</th>\n",
       "      <td>disease_phenotype_positive</td>\n",
       "      <td>phenotype present</td>\n",
       "      <td>31809</td>\n",
       "      <td>32864</td>\n",
       "      <td>disease</td>\n",
       "      <td>intellectual developmental disorder with speec...</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>22759</td>\n",
       "      <td>6</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Autosomal dominant inheritance</td>\n",
       "      <td>HPO</td>\n",
       "      <td>MONDO:0032864</td>\n",
       "      <td>HP:0000006</td>\n",
       "      <td>biolink:Disease</td>\n",
       "      <td>biolink:PhenotypicFeature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1162436</th>\n",
       "      <td>drug_drug</td>\n",
       "      <td>synergistic interaction</td>\n",
       "      <td>20349</td>\n",
       "      <td>DB09543</td>\n",
       "      <td>drug</td>\n",
       "      <td>Methyl salicylate</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>14432</td>\n",
       "      <td>DB09383</td>\n",
       "      <td>drug</td>\n",
       "      <td>Meprednisone</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>DRUGBANK:DB09543</td>\n",
       "      <td>DRUGBANK:DB09383</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:Drug</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6112214</th>\n",
       "      <td>drug_effect</td>\n",
       "      <td>side effect</td>\n",
       "      <td>22939</td>\n",
       "      <td>83</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Renal insufficiency</td>\n",
       "      <td>HPO</td>\n",
       "      <td>17730</td>\n",
       "      <td>DB00710</td>\n",
       "      <td>drug</td>\n",
       "      <td>Ibandronate</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>HP:0000083</td>\n",
       "      <td>DRUGBANK:DB00710</td>\n",
       "      <td>biolink:PhenotypicFeature</td>\n",
       "      <td>biolink:Drug</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5473379</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>637</td>\n",
       "      <td>5725</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PTBP1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>866</td>\n",
       "      <td>5595</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MAPK3</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NCBIGene:5725</td>\n",
       "      <td>NCBIGene:5595</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           relation         display_relation  x_index  \\\n",
       "3157330  disease_phenotype_positive        phenotype present    31809   \n",
       "1162436                   drug_drug  synergistic interaction    20349   \n",
       "6112214                 drug_effect              side effect    22939   \n",
       "5473379             protein_protein                      ppi      637   \n",
       "\n",
       "            x_id            x_type  \\\n",
       "3157330    32864           disease   \n",
       "1162436  DB09543              drug   \n",
       "6112214       83  effect/phenotype   \n",
       "5473379     5725      gene/protein   \n",
       "\n",
       "                                                    x_name  x_source  y_index  \\\n",
       "3157330  intellectual developmental disorder with speec...     MONDO    22759   \n",
       "1162436                                  Methyl salicylate  DrugBank    14432   \n",
       "6112214                                Renal insufficiency       HPO    17730   \n",
       "5473379                                              PTBP1      NCBI      866   \n",
       "\n",
       "            y_id            y_type                          y_name  y_source  \\\n",
       "3157330        6  effect/phenotype  Autosomal dominant inheritance       HPO   \n",
       "1162436  DB09383              drug                    Meprednisone  DrugBank   \n",
       "6112214  DB00710              drug                     Ibandronate  DrugBank   \n",
       "5473379     5595      gene/protein                           MAPK3      NCBI   \n",
       "\n",
       "                  subject            object           subject_category  \\\n",
       "3157330     MONDO:0032864        HP:0000006            biolink:Disease   \n",
       "1162436  DRUGBANK:DB09543  DRUGBANK:DB09383               biolink:Drug   \n",
       "6112214        HP:0000083  DRUGBANK:DB00710  biolink:PhenotypicFeature   \n",
       "5473379     NCBIGene:5725     NCBIGene:5595               biolink:Gene   \n",
       "\n",
       "                   object_category  \n",
       "3157330  biolink:PhenotypicFeature  \n",
       "1162436               biolink:Drug  \n",
       "6112214               biolink:Drug  \n",
       "5473379               biolink:Gene  "
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## deep copy a df\n",
    "edges_deep_copy_df2 = edges_deep_copy_df.copy(deep=True)\n",
    "\n",
    "## assign biolink node category using dictionary directly\n",
    "edges_deep_copy_df2['subject_category'] = (\n",
    "    edges_deep_copy_df2['x_type'].map(category_map)\n",
    ")\n",
    "\n",
    "edges_deep_copy_df2['object_category'] = (\n",
    "    edges_deep_copy_df2['y_type'].map(category_map)\n",
    ")\n",
    "\n",
    "sampled_df = edges_deep_copy_df2.groupby(['x_type']).sample(n=1, random_state=151)  # Set random_state for reproducibility\n",
    "sampled_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "d7d81104-73de-471a-a3a6-fa24e1eb487a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "129568\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3348335</th>\n",
       "      <td>drug_effect</td>\n",
       "      <td>side effect</td>\n",
       "      <td>16322</td>\n",
       "      <td>DB00583</td>\n",
       "      <td>drug</td>\n",
       "      <td>Levocarnitine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>23158</td>\n",
       "      <td>2027</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Abdominal pain</td>\n",
       "      <td>HPO</td>\n",
       "      <td>DRUGBANK:DB00583</td>\n",
       "      <td>HP:0002027</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:PhenotypicFeature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3348336</th>\n",
       "      <td>drug_effect</td>\n",
       "      <td>side effect</td>\n",
       "      <td>16322</td>\n",
       "      <td>DB00583</td>\n",
       "      <td>drug</td>\n",
       "      <td>Levocarnitine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>85849</td>\n",
       "      <td>4396</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Poor appetite</td>\n",
       "      <td>HPO</td>\n",
       "      <td>DRUGBANK:DB00583</td>\n",
       "      <td>HP:0004396</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:PhenotypicFeature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3348337</th>\n",
       "      <td>drug_effect</td>\n",
       "      <td>side effect</td>\n",
       "      <td>16322</td>\n",
       "      <td>DB00583</td>\n",
       "      <td>drug</td>\n",
       "      <td>Levocarnitine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>22447</td>\n",
       "      <td>739</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Anxiety</td>\n",
       "      <td>HPO</td>\n",
       "      <td>DRUGBANK:DB00583</td>\n",
       "      <td>HP:0000739</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:PhenotypicFeature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3348338</th>\n",
       "      <td>drug_effect</td>\n",
       "      <td>side effect</td>\n",
       "      <td>16322</td>\n",
       "      <td>DB00583</td>\n",
       "      <td>drug</td>\n",
       "      <td>Levocarnitine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>22831</td>\n",
       "      <td>11675</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Arrhythmia</td>\n",
       "      <td>HPO</td>\n",
       "      <td>DRUGBANK:DB00583</td>\n",
       "      <td>HP:0011675</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:PhenotypicFeature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3348339</th>\n",
       "      <td>drug_effect</td>\n",
       "      <td>side effect</td>\n",
       "      <td>16322</td>\n",
       "      <td>DB00583</td>\n",
       "      <td>drug</td>\n",
       "      <td>Levocarnitine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>23469</td>\n",
       "      <td>3418</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Back pain</td>\n",
       "      <td>HPO</td>\n",
       "      <td>DRUGBANK:DB00583</td>\n",
       "      <td>HP:0003418</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:PhenotypicFeature</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            relation display_relation  x_index     x_id x_type         x_name  \\\n",
       "3348335  drug_effect      side effect    16322  DB00583   drug  Levocarnitine   \n",
       "3348336  drug_effect      side effect    16322  DB00583   drug  Levocarnitine   \n",
       "3348337  drug_effect      side effect    16322  DB00583   drug  Levocarnitine   \n",
       "3348338  drug_effect      side effect    16322  DB00583   drug  Levocarnitine   \n",
       "3348339  drug_effect      side effect    16322  DB00583   drug  Levocarnitine   \n",
       "\n",
       "         x_source  y_index   y_id            y_type          y_name y_source  \\\n",
       "3348335  DrugBank    23158   2027  effect/phenotype  Abdominal pain      HPO   \n",
       "3348336  DrugBank    85849   4396  effect/phenotype   Poor appetite      HPO   \n",
       "3348337  DrugBank    22447    739  effect/phenotype         Anxiety      HPO   \n",
       "3348338  DrugBank    22831  11675  effect/phenotype      Arrhythmia      HPO   \n",
       "3348339  DrugBank    23469   3418  effect/phenotype       Back pain      HPO   \n",
       "\n",
       "                  subject      object subject_category  \\\n",
       "3348335  DRUGBANK:DB00583  HP:0002027     biolink:Drug   \n",
       "3348336  DRUGBANK:DB00583  HP:0004396     biolink:Drug   \n",
       "3348337  DRUGBANK:DB00583  HP:0000739     biolink:Drug   \n",
       "3348338  DRUGBANK:DB00583  HP:0011675     biolink:Drug   \n",
       "3348339  DRUGBANK:DB00583  HP:0003418     biolink:Drug   \n",
       "\n",
       "                   object_category  \n",
       "3348335  biolink:PhenotypicFeature  \n",
       "3348336  biolink:PhenotypicFeature  \n",
       "3348337  biolink:PhenotypicFeature  \n",
       "3348338  biolink:PhenotypicFeature  \n",
       "3348339  biolink:PhenotypicFeature  "
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Sanity check anatomy_anatomy relations\n",
    "## should return empty df after mapping to dictionary\n",
    "sanity_check_df = edges_deep_copy_df2[edges_deep_copy_df2['display_relation'] == 'side effect']\n",
    "print(len(sanity_check_df))\n",
    "sanity_check_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "7ace4ff8-593c-4234-8a2b-dbb8db4c2417",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>321939</th>\n",
       "      <td>drug_protein</td>\n",
       "      <td>enzyme</td>\n",
       "      <td>14584</td>\n",
       "      <td>DB00130</td>\n",
       "      <td>drug</td>\n",
       "      <td>L-Glutamine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>13982</td>\n",
       "      <td>27165</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GLS2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>DRUGBANK:DB00130</td>\n",
       "      <td>NCBIGene:27165</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321940</th>\n",
       "      <td>drug_protein</td>\n",
       "      <td>enzyme</td>\n",
       "      <td>14585</td>\n",
       "      <td>DB11118</td>\n",
       "      <td>drug</td>\n",
       "      <td>Ammonia</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>13982</td>\n",
       "      <td>27165</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GLS2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>DRUGBANK:DB11118</td>\n",
       "      <td>NCBIGene:27165</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321941</th>\n",
       "      <td>drug_protein</td>\n",
       "      <td>enzyme</td>\n",
       "      <td>14584</td>\n",
       "      <td>DB00130</td>\n",
       "      <td>drug</td>\n",
       "      <td>L-Glutamine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>476</td>\n",
       "      <td>2162</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>F13A1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>DRUGBANK:DB00130</td>\n",
       "      <td>NCBIGene:2162</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321942</th>\n",
       "      <td>drug_protein</td>\n",
       "      <td>enzyme</td>\n",
       "      <td>14275</td>\n",
       "      <td>DB00997</td>\n",
       "      <td>drug</td>\n",
       "      <td>Doxorubicin</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>373</td>\n",
       "      <td>4843</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NOS2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>DRUGBANK:DB00997</td>\n",
       "      <td>NCBIGene:4843</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321943</th>\n",
       "      <td>drug_protein</td>\n",
       "      <td>enzyme</td>\n",
       "      <td>14423</td>\n",
       "      <td>DB09237</td>\n",
       "      <td>drug</td>\n",
       "      <td>Levamlodipine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>373</td>\n",
       "      <td>4843</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NOS2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>DRUGBANK:DB09237</td>\n",
       "      <td>NCBIGene:4843</td>\n",
       "      <td>biolink:Drug</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            relation display_relation  x_index     x_id x_type         x_name  \\\n",
       "321939  drug_protein           enzyme    14584  DB00130   drug    L-Glutamine   \n",
       "321940  drug_protein           enzyme    14585  DB11118   drug        Ammonia   \n",
       "321941  drug_protein           enzyme    14584  DB00130   drug    L-Glutamine   \n",
       "321942  drug_protein           enzyme    14275  DB00997   drug    Doxorubicin   \n",
       "321943  drug_protein           enzyme    14423  DB09237   drug  Levamlodipine   \n",
       "\n",
       "        x_source  y_index   y_id        y_type y_name y_source  \\\n",
       "321939  DrugBank    13982  27165  gene/protein   GLS2     NCBI   \n",
       "321940  DrugBank    13982  27165  gene/protein   GLS2     NCBI   \n",
       "321941  DrugBank      476   2162  gene/protein  F13A1     NCBI   \n",
       "321942  DrugBank      373   4843  gene/protein   NOS2     NCBI   \n",
       "321943  DrugBank      373   4843  gene/protein   NOS2     NCBI   \n",
       "\n",
       "                 subject          object subject_category object_category  \n",
       "321939  DRUGBANK:DB00130  NCBIGene:27165     biolink:Drug    biolink:Gene  \n",
       "321940  DRUGBANK:DB11118  NCBIGene:27165     biolink:Drug    biolink:Gene  \n",
       "321941  DRUGBANK:DB00130   NCBIGene:2162     biolink:Drug    biolink:Gene  \n",
       "321942  DRUGBANK:DB00997   NCBIGene:4843     biolink:Drug    biolink:Gene  \n",
       "321943  DRUGBANK:DB09237   NCBIGene:4843     biolink:Drug    biolink:Gene  "
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Sanity check anatomy_anatomy relations\n",
    "## should return empty df after mapping to dictionary\n",
    "## enzyme: for drug metabolism\n",
    "sanity_check_df = edges_deep_copy_df2[edges_deep_copy_df2['display_relation'] == 'enzyme']\n",
    "sanity_check_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d64eae4d-fc8b-4ad8-bd28-af26a7dd3744",
   "metadata": {},
   "source": [
    "* **Question 4:** general mapping for predicate\n",
    "* protein-protein interacts, use physical_interacts_with\n",
    "* carrier, use 'biolink:can_be_carried_out_by'?\n",
    "* enzyme?\n",
    "* indication?\n",
    "* synergistic interaction?\n",
    "* parent-child, between two phenotypes nodes, should I use biolink:broad_match?\n",
    "* phenotype absent, opposite of has_phenotype? or just ignore?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "9afe40c9-b7e7-4866-ae92-f749635358d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Predicate: associated with\n",
      "  (biolink:Gene, biolink:PhenotypicFeature)\n",
      "  (biolink:Gene, biolink:Disease)\n",
      "  (biolink:PhenotypicFeature, biolink:Gene)\n",
      "  (biolink:Disease, biolink:Gene)\n",
      "\n",
      "Predicate: carrier\n",
      "  (biolink:Drug, biolink:Gene)\n",
      "  (biolink:Gene, biolink:Drug)\n",
      "\n",
      "Predicate: contraindication\n",
      "  (biolink:Drug, biolink:Disease)\n",
      "  (biolink:Disease, biolink:Drug)\n",
      "\n",
      "Predicate: enzyme\n",
      "  (biolink:Drug, biolink:Gene)\n",
      "  (biolink:Gene, biolink:Drug)\n",
      "\n",
      "Predicate: indication\n",
      "  (biolink:Drug, biolink:Disease)\n",
      "  (biolink:Disease, biolink:Drug)\n",
      "\n",
      "Predicate: off-label use\n",
      "  (biolink:Drug, biolink:Disease)\n",
      "  (biolink:Disease, biolink:Drug)\n",
      "\n",
      "Predicate: parent-child\n",
      "  (biolink:PhenotypicFeature, biolink:PhenotypicFeature)\n",
      "  (biolink:Disease, biolink:Disease)\n",
      "\n",
      "Predicate: phenotype absent\n",
      "  (biolink:Disease, biolink:PhenotypicFeature)\n",
      "  (biolink:PhenotypicFeature, biolink:Disease)\n",
      "\n",
      "Predicate: phenotype present\n",
      "  (biolink:Disease, biolink:PhenotypicFeature)\n",
      "  (biolink:PhenotypicFeature, biolink:Disease)\n",
      "\n",
      "Predicate: ppi\n",
      "  (biolink:Gene, biolink:Gene)\n",
      "\n",
      "Predicate: side effect\n",
      "  (biolink:Drug, biolink:PhenotypicFeature)\n",
      "  (biolink:PhenotypicFeature, biolink:Drug)\n",
      "\n",
      "Predicate: synergistic interaction\n",
      "  (biolink:Drug, biolink:Drug)\n",
      "\n",
      "Predicate: target\n",
      "  (biolink:Drug, biolink:Gene)\n",
      "  (biolink:Gene, biolink:Drug)\n",
      "\n",
      "Predicate: transporter\n",
      "  (biolink:Drug, biolink:Gene)\n",
      "  (biolink:Gene, biolink:Drug)\n"
     ]
    }
   ],
   "source": [
    "## Group by predicate\n",
    "grouped = edges_deep_copy_df2.groupby('display_relation')\n",
    "\n",
    "## For each predicate, output unique (subject_category, object_category) pairs\n",
    "for predicate, group in grouped:\n",
    "    print(f\"\\nPredicate: {predicate}\")\n",
    "    pairs = group[['subject_category', 'object_category']].drop_duplicates()\n",
    "    for _, row in pairs.iterrows():\n",
    "        print(f\"  ({row['subject_category']}, {row['object_category']})\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "0e77e54b-d0a7-46be-835b-496967bb5772",
   "metadata": {},
   "outputs": [],
   "source": [
    "## start convert to biolink compliant predicate types based on display_relation column\n",
    "## check supplementary table of primeKG, descriptions on relations\n",
    "## translator discussion point - data ingestion group meeting\n",
    "## separate out off label drug usage, clinical trial, and FDA approved usage\n",
    "## for 'synergistic interaction', combine with knowledge from LLM clinical trial attempts and find drug combination and which disease \n",
    "## they targeted\n",
    "## currently exclude associated with\n",
    "## not need to include 'phenotype absent'\n",
    "\n",
    "## match only combinations allowed in the pair dictionary\n",
    "## Apply the mapping and return a Series with 3 columns\n",
    "edges_deep_copy_df2[['predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier']] = edges_deep_copy_df2.apply(\n",
    "    lambda row: pd.Series(\n",
    "        predicate_map.get(\n",
    "            (row['subject_category'], row['object_category'], row['display_relation']),\n",
    "            [None, None, None]  # Default if not found\n",
    "        )\n",
    "    ),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9973a249-d2b2-42e9-8b9f-00637e100019",
   "metadata": {},
   "source": [
    "## add individual data source (original instead of PrimeKG) for each edge in the PrimeKG\n",
    "* The mapping can be found in the build_kg script: https://github.com/mims-harvard/PrimeKG/blob/main/knowledge_graph/build_graph.ipynb\n",
    "* Then we create our own dictionary to map the individual data souce to the ingested KG\n",
    "* use a unique combon of following columns:\n",
    "* x_source, y_source, relation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "8306fffd-a8b0-4dc5-85ee-4b629ce26f82",
   "metadata": {},
   "outputs": [],
   "source": [
    "## temp: use the whole PrimeKG as knowledge source for now\n",
    "# data_source_map_dict = {\n",
    "#     ('NCBI', 'NCBI', 'protein_protein'): 'NCBI',\n",
    "#     ('NCBI', 'MONDO', 'disease_protein'): 'DisGenNet',\n",
    "#     ('NCBI', 'HPO', 'phenotype_protein'): 'DisGenNet',\n",
    "#     ('NCBI', 'GO', 'molfunc_protein'): 'Gene2GO',\n",
    "#     ('NCBI', 'GO', 'cellcomp_protein'): 'Gene2GO',\n",
    "#     ('NCBI', 'GO', 'bioprocess_protein'): 'Gene2GO',\n",
    "#     ('NCBI', 'UBERON', 'anatomy_protein_present'): 'BGEE',\n",
    "#     ('NCBI', 'UBERON', 'anatomy_protein_absent'): 'BGEE',\n",
    "#     ('NCBI', 'REACTOME', 'pathway_protein'): 'Reactome',\n",
    "#     #('DrugBank', 'MONDO', 'contraindication'): 'DrugCentral', ## Note: need to double check if other relation types are for this data source\n",
    "#     #('DrugBank', 'MONDO', 'indication'): 'DrugCentral', ## note: there is a typo? of DrugCentral into DiseaseCentral in their comments\n",
    "#     ('DrugBank', 'MONDO', 'off-label use'): 'DrugCentral',\n",
    "#     ('DrugBank', 'NCBI', 'drug_protein'): 'DrugBank',\n",
    "#     ('DrugBank', 'DrugBank', 'drug_drug'): 'DrugBank',\n",
    "#     ('DrugBank', 'HPO', 'drug_effect'): 'SIDER',\n",
    "#     ('MONDO', 'MONDO', 'disease_disease'): 'MONDO',\n",
    "#     ('MONDO', 'NCBI', 'disease_protein'): 'MONDO',\n",
    "#     ('MONDO', 'HPO', 'disease_phenotype_positive'): 'HPO-A',\n",
    "#     ('MONDO', 'HPO', 'disease_phenotype_negative'): 'HPO-A',\n",
    "#     ('HPO', 'HPO', 'phenotype_phenotype'): 'HPO',\n",
    "#     ('GO', 'GO', 'bioprocess_bioprocess'): 'GO',\n",
    "#     ('GO', 'GO', 'molfunc_molfunc'): 'GO',\n",
    "#     ('GO', 'GO', 'cellcomp_cellcomp'): 'GO',\n",
    "#     ('CTD', 'NCBI', 'exposure_protein'): 'CTD',\n",
    "#     ('CTD', 'MONDO', 'exposure_disease'): 'CTD',\n",
    "#     ('CTD', 'CTD', 'exposure_exposure'): 'CTD',\n",
    "#     ('CTD', 'GO', 'exposure_bioprocess'): 'CTD',\n",
    "#     ('CTD', 'GO', 'exposure_molfunc'): 'CTD',\n",
    "#     ('CTD', 'GO', 'exposure_cellcomp'): 'CTD',\n",
    "#     ('UBERON', 'UBERON', 'anatomy_anatomy'): 'UBERON',\n",
    "#     ('REACTOME', 'REACTOME', 'pathway_pathway'): 'Reactome',\n",
    "# }\n",
    "\n",
    "## Map the combination to a new column 'knowledge_source'\n",
    "## if not finding a mapped original data source then label as the edge is from PrimeKG itself\n",
    "# edges_deep_copy_df['knowledge_source'] = edges_deep_copy_df.apply(\n",
    "#     lambda row: data_source_map_dict.get((row['x_source'], row['y_source'], row['relation']), 'PrimeKG'), axis=1)\n",
    "\n",
    "edges_deep_copy_df2['knowledge_source'] = 'PrimeKG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "0e609a0b-9d04-4b8a-89ab-feb3437bdace",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "knowledge_source\n",
      "PrimeKG    4009644\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Count occurrences of each unique value in 'knowledge_source'\n",
    "counts = edges_deep_copy_df2['knowledge_source'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce95bea9-317e-422d-97b6-0f97aa5154aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Sanity check those rows with PrimeKG as datasource\n",
    "## and randomly select 1 row out of each relation type\n",
    "\n",
    "# sampled_df = edges_deep_copy_df2[edges_deep_copy_df2['knowledge_source'] == 'PrimeKG'].groupby(['relation', 'x_source', 'y_source']).sample(n=1, random_state=151)  # Set random_state for reproducibility\n",
    "# sampled_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "330f71af-b101-42cb-9ee9-d3592e269baf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4009644\n",
      "4009644\n"
     ]
    }
   ],
   "source": [
    "## add a new knowledge_souce column and set value to be \"PrimeKG\"\n",
    "edges_deep_copy_df2['knowledge_souce_total'] = 'PrimeKG'\n",
    "## add a new knowledge_level column and set value to be 'knowledge_assertion'\n",
    "edges_deep_copy_df2['knowledge_level'] = 'knowledge_assertion'\n",
    "## add a new agent_type column and set value to be 'manual_agent'\n",
    "edges_deep_copy_df2['agent_type'] = 'automated_agent'\n",
    "\n",
    "## copy to a final df\n",
    "edge_df = edges_deep_copy_df2.copy(deep = True)\n",
    "\n",
    "print(edge_df.shape[0])\n",
    "## Remove rows where subject or object is empty\n",
    "# Remove rows where 'Subject' OR 'Object' have NaN values\n",
    "edge_df = edge_df.dropna(subset=['subject', 'object'])\n",
    "\n",
    "print(edge_df.shape[0])\n",
    "\n",
    "edge_df['deploy_date'] = deployment_date\n",
    "# edge_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "042ff148-2391-4d87-9b90-df67cccc97af",
   "metadata": {},
   "outputs": [],
   "source": [
    "## rename those columns into desired format\n",
    "edge_df.rename(columns={'x_name': 'subject_name', 'y_name': 'object_name'}, inplace=True)\n",
    "\n",
    "## create a publication column and fill na\n",
    "## if all of them are empty then fill na\n",
    "edge_df['publications'] = np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "6f582fd3-e661-4493-8b8d-4282054293c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## create a context_qualifier column and fill na\n",
    "## if all of them are empty then fill na\n",
    "edge_df['context_qualifier'] = np.nan\n",
    "# edge_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2213ea6e-7983-4131-b98f-57bf901cc291",
   "metadata": {},
   "source": [
    "### Get all helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "5c1184f1-bcf1-45c5-960d-2192290ef126",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load all helper functions\n",
    "%run /Users/Weiqi0/ISB_working/Hadlock_lab/QI_ISB_Git_repo/TranslatorPharcogenomicsKG/Parser_helper_functions.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "4665996c-6e27-410e-893c-982a43644d58",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add resources_id column, checking whether edge is already\n",
    "column_list = ['subject', 'predicate', 'object', 'context_qualifier', 'deploy_date']\n",
    "# Apply the function to each row to generate UUIDs\n",
    "edge_df['id'] = edge_df[column_list].apply(generate_uuid, axis=1)\n",
    "\n",
    "# edge_df['id'] = generate_uuid_from_columns(edge_df, column_list)\n",
    "# edge_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9fcad02-03f0-42a6-b333-1abf46a92772",
   "metadata": {},
   "source": [
    "### Now create the corresponding node file\n",
    "* only need three columns: id, name, category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "2b520220-06e9-4c94-8737-70341470ad94",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_17326/1611443582.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  node_subject_df.rename(columns={'subject': 'id', 'subject_name': 'name', 'subject_category': 'category'}, inplace=True)\n",
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_17326/1611443582.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  node_object_df.rename(columns={'object': 'id', 'object_name': 'name', 'object_category': 'category'}, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:9796</td>\n",
       "      <td>PHYHIP</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBIGene:7918</td>\n",
       "      <td>GPANK1</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBIGene:8233</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBIGene:4899</td>\n",
       "      <td>NRF1</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBIGene:5297</td>\n",
       "      <td>PI4KA</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              id    name      category\n",
       "0  NCBIGene:9796  PHYHIP  biolink:Gene\n",
       "1  NCBIGene:7918  GPANK1  biolink:Gene\n",
       "2  NCBIGene:8233   ZRSR2  biolink:Gene\n",
       "3  NCBIGene:4899    NRF1  biolink:Gene\n",
       "4  NCBIGene:5297   PI4KA  biolink:Gene"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_subject_df = edge_df[['subject', 'subject_name', 'subject_category']]\n",
    "node_object_df = edge_df[['object', 'object_name', 'object_category']]\n",
    "\n",
    "## rename those columns into desired format\n",
    "node_subject_df.rename(columns={'subject': 'id', 'subject_name': 'name', 'subject_category': 'category'}, inplace=True)\n",
    "node_object_df.rename(columns={'object': 'id', 'object_name': 'name', 'object_category': 'category'}, inplace=True)\n",
    "\n",
    "concat_node_df = pd.concat([node_subject_df, node_object_df]).drop_duplicates(keep='first')\n",
    "\n",
    "concat_node_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "9d02424f-cb0b-4767-8424-671272ed865f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['relation', 'display_relation', 'x_index', 'x_id', 'x_type', 'subject_name', 'x_source', 'y_index', 'y_id', 'y_type', 'object_name', 'y_source', 'subject', 'object', 'subject_category', 'object_category', 'predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier', 'knowledge_source', 'knowledge_souce_total', 'knowledge_level', 'agent_type', 'deploy_date', 'publications', 'context_qualifier', 'id']\n"
     ]
    }
   ],
   "source": [
    "## drop no longer needed columns from edge df\n",
    "print(edge_df.columns.tolist())\n",
    "drop_cols = ['relation', 'display_relation', 'x_index', 'x_id', 'x_type',  'x_source', 'y_index', 'y_id', 'y_type', 'y_source']\n",
    "\n",
    "edge_output_df = edge_df.drop(drop_cols, axis=1)\n",
    "\n",
    "# edge_output_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "56f2da1e-5ad6-408d-be27-e068d8eed87d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## exclude rows with Null and empty string values in the predicate column\n",
    "# edge_output_df = edge_output_df.dropna(subset=['predicate'])\n",
    "\n",
    "## Drop rows where 'name' is NaN, None, or empty string\n",
    "edge_output_df = edge_output_df[~edge_output_df['predicate'].isna() & (edge_output_df['predicate'].str.strip() != '')]\n",
    "\n",
    "## throw away those rows which either subject_category, object_category, or predicate is not start with \"biolink:\" prefix\n",
    "## since they cannot be biolink-compliant converted\n",
    "## Keep only rows where all three columns start with 'biolink:'\n",
    "# edge_output_df = edge_output_df[\n",
    "#     edge_output_df['subject_category'].str.startswith('biolink:') &\n",
    "#     edge_output_df['object_category'].str.startswith('biolink:') &\n",
    "#     edge_output_df['predicate'].str.startswith('biolink:')\n",
    "# ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "bd927a19-77ab-4b49-b285-7112a5469ccd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "knowledge_source\n",
      "PrimeKG    954592\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## Check the knowledge_source column again\n",
    "## Count occurrences of each unique value in 'knowledge_source'\n",
    "counts = edge_output_df['knowledge_source'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "9586402b-cc5b-4247-83a2-ad9c4d670dbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "954592\n"
     ]
    }
   ],
   "source": [
    "print(len(edge_output_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "f8344044-2e14-4361-8339-d888d6cf627b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['subject_name', 'object_name', 'subject', 'object', 'subject_category', 'object_category', 'predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier', 'knowledge_source', 'knowledge_souce_total', 'knowledge_level', 'agent_type', 'deploy_date', 'publications', 'context_qualifier', 'id']\n"
     ]
    }
   ],
   "source": [
    "## sanity on the column names of the parsed KG\n",
    "## the desired order should be: \n",
    "## 'subject', 'object', 'subject_category', 'object_category', 'predicate', 'knowledge_source', 'knowledge_souce_total', \n",
    "## 'knowledge_level', 'agent_type', 'deploy_date', 'context_qualifier', 'id'\n",
    "\n",
    "## Print the order of column names\n",
    "print(list(edge_output_df.columns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "7d00d0970fff9045",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Predicate: biolink:GeneAffectsChemicalAssociation\n",
      "  (biolink:Gene, biolink:Drug)\n",
      "\n",
      "Predicate: biolink:broad_match\n",
      "  (biolink:PhenotypicFeature, biolink:PhenotypicFeature)\n",
      "  (biolink:Disease, biolink:Disease)\n",
      "\n",
      "Predicate: biolink:can_be_carried_out_by\n",
      "  (biolink:Gene, biolink:Drug)\n",
      "\n",
      "Predicate: biolink:genetically_associated_with\n",
      "  (biolink:Gene, biolink:PhenotypicFeature)\n",
      "  (biolink:Gene, biolink:Disease)\n",
      "\n",
      "Predicate: biolink:has_contraindication\n",
      "  (biolink:Drug, biolink:Disease)\n",
      "\n",
      "Predicate: biolink:has_phenotype\n",
      "  (biolink:Disease, biolink:PhenotypicFeature)\n",
      "\n",
      "Predicate: biolink:physical_interacts_with\n",
      "  (biolink:Gene, biolink:Gene)\n",
      "\n",
      "Predicate: biolink:target_for\n",
      "  (biolink:Gene, biolink:Drug)\n",
      "\n",
      "Predicate: biolink:treats\n",
      "  (biolink:Drug, biolink:Disease)\n"
     ]
    }
   ],
   "source": [
    "## Group by predicate\n",
    "grouped = edge_output_df.groupby('predicate')\n",
    "\n",
    "## For each predicate, output unique (subject_category, object_category) pairs\n",
    "for predicate, group in grouped:\n",
    "    print(f\"\\nPredicate: {predicate}\")\n",
    "    pairs = group[['subject_category', 'object_category']].drop_duplicates()\n",
    "    for _, row in pairs.iterrows():\n",
    "        print(f\"  ({row['subject_category']}, {row['object_category']})\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "24cb30e8-43fe-4172-a386-d2116918de8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_name</th>\n",
       "      <th>object_name</th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_category</th>\n",
       "      <th>predicate</th>\n",
       "      <th>subject_direction_qualifier</th>\n",
       "      <th>object_direction_qualifier</th>\n",
       "      <th>object_aspect_qualifier</th>\n",
       "      <th>knowledge_source</th>\n",
       "      <th>knowledge_souce_total</th>\n",
       "      <th>knowledge_level</th>\n",
       "      <th>agent_type</th>\n",
       "      <th>deploy_date</th>\n",
       "      <th>publications</th>\n",
       "      <th>context_qualifier</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PHYHIP</td>\n",
       "      <td>KIF15</td>\n",
       "      <td>NCBIGene:9796</td>\n",
       "      <td>NCBIGene:56992</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:physical_interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PrimeKG</td>\n",
       "      <td>PrimeKG</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2026-01-06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>82779b17-5c64-5e78-b12a-0b0490d1e404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GPANK1</td>\n",
       "      <td>PNMA1</td>\n",
       "      <td>NCBIGene:7918</td>\n",
       "      <td>NCBIGene:9240</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:physical_interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PrimeKG</td>\n",
       "      <td>PrimeKG</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2026-01-06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0210c181-665f-549e-a8e0-6216eb316312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>TTC33</td>\n",
       "      <td>NCBIGene:8233</td>\n",
       "      <td>NCBIGene:23548</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:physical_interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PrimeKG</td>\n",
       "      <td>PrimeKG</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2026-01-06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>e8538283-d8d4-5c52-934e-ac94f66cca6e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NRF1</td>\n",
       "      <td>MAN1B1</td>\n",
       "      <td>NCBIGene:4899</td>\n",
       "      <td>NCBIGene:11253</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:physical_interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PrimeKG</td>\n",
       "      <td>PrimeKG</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2026-01-06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dabd6522-e108-5b35-9eba-dcf322c8eddb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PI4KA</td>\n",
       "      <td>RGS20</td>\n",
       "      <td>NCBIGene:5297</td>\n",
       "      <td>NCBIGene:8601</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:physical_interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PrimeKG</td>\n",
       "      <td>PrimeKG</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2026-01-06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>90fc7fc7-3a73-5945-a074-8bc47634d23b</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_name object_name        subject          object subject_category  \\\n",
       "0       PHYHIP       KIF15  NCBIGene:9796  NCBIGene:56992     biolink:Gene   \n",
       "1       GPANK1       PNMA1  NCBIGene:7918   NCBIGene:9240     biolink:Gene   \n",
       "2        ZRSR2       TTC33  NCBIGene:8233  NCBIGene:23548     biolink:Gene   \n",
       "3         NRF1      MAN1B1  NCBIGene:4899  NCBIGene:11253     biolink:Gene   \n",
       "4        PI4KA       RGS20  NCBIGene:5297   NCBIGene:8601     biolink:Gene   \n",
       "\n",
       "  object_category                        predicate  \\\n",
       "0    biolink:Gene  biolink:physical_interacts_with   \n",
       "1    biolink:Gene  biolink:physical_interacts_with   \n",
       "2    biolink:Gene  biolink:physical_interacts_with   \n",
       "3    biolink:Gene  biolink:physical_interacts_with   \n",
       "4    biolink:Gene  biolink:physical_interacts_with   \n",
       "\n",
       "  subject_direction_qualifier object_direction_qualifier  \\\n",
       "0                        None                       None   \n",
       "1                        None                       None   \n",
       "2                        None                       None   \n",
       "3                        None                       None   \n",
       "4                        None                       None   \n",
       "\n",
       "   object_aspect_qualifier knowledge_source knowledge_souce_total  \\\n",
       "0                      NaN          PrimeKG               PrimeKG   \n",
       "1                      NaN          PrimeKG               PrimeKG   \n",
       "2                      NaN          PrimeKG               PrimeKG   \n",
       "3                      NaN          PrimeKG               PrimeKG   \n",
       "4                      NaN          PrimeKG               PrimeKG   \n",
       "\n",
       "       knowledge_level       agent_type deploy_date  publications  \\\n",
       "0  knowledge_assertion  automated_agent  2026-01-06           NaN   \n",
       "1  knowledge_assertion  automated_agent  2026-01-06           NaN   \n",
       "2  knowledge_assertion  automated_agent  2026-01-06           NaN   \n",
       "3  knowledge_assertion  automated_agent  2026-01-06           NaN   \n",
       "4  knowledge_assertion  automated_agent  2026-01-06           NaN   \n",
       "\n",
       "   context_qualifier                                    id  \n",
       "0                NaN  82779b17-5c64-5e78-b12a-0b0490d1e404  \n",
       "1                NaN  0210c181-665f-549e-a8e0-6216eb316312  \n",
       "2                NaN  e8538283-d8d4-5c52-934e-ac94f66cca6e  \n",
       "3                NaN  dabd6522-e108-5b35-9eba-dcf322c8eddb  \n",
       "4                NaN  90fc7fc7-3a73-5945-a074-8bc47634d23b  "
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_output_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "744ef596-58aa-460c-abc2-34486191ed1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['subject', 'subject_name', 'subject_category', 'object', 'object_name', 'object_category', 'predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier', 'knowledge_source', 'publications', 'knowledge_level', 'context_qualifier', 'agent_type', 'deploy_date', 'id']\n"
     ]
    }
   ],
   "source": [
    "## reorder the dataframe\n",
    "desired_order = [\n",
    "    'subject', 'subject_name', 'subject_category', 'object', 'object_name', 'object_category', \n",
    "    'predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier',\n",
    "    'knowledge_source', 'publications', 'knowledge_level', 'context_qualifier', 'agent_type', \n",
    "    'deploy_date', 'id'\n",
    "]\n",
    "\n",
    "edge_output_df = edge_output_df[desired_order]\n",
    "\n",
    "print(edge_output_df.columns.tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dae84b0b-e8c4-41f8-bf46-df652439656e",
   "metadata": {},
   "source": [
    "### Now quality control of the parsed PrimeKG dataframe\n",
    "* Original edge counts: 954592\n",
    "* QA:\n",
    "  - dropped duplicated edges in reverse direction\n",
    "  - dropped enzyme and node types not easily mapped in biolink\n",
    "  - removed edges with invalid predicate, subject, and object relationship"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "6fcf6924-4fc1-433e-8e1e-be8a5f1c49a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes 55770\n",
      "Number of edges 589950\n",
      "Average degree 21.15653577192039\n"
     ]
    }
   ],
   "source": [
    "## Create a graph from the DataFrame\n",
    "graph = nx.from_pandas_edgelist(edge_output_df, 'subject', 'object', edge_attr='predicate')\n",
    "\n",
    "## Print graph information\n",
    "print('Number of nodes', len(set(graph.nodes)))\n",
    "print('Number of edges', len(set(graph.edges)))\n",
    "print('Average degree', sum(dict(graph.degree).values()) / len(graph.nodes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "9a370747-af67-46cc-a67b-c397bbf597ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "category\n",
      "biolink:Gene                 18995\n",
      "biolink:Disease              15343\n",
      "biolink:PhenotypicFeature    15271\n",
      "biolink:Drug                  7944\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = concat_node_df['category'].value_counts()\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "07171492-30b9-4f2f-b137-ec90f2447c2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "predicate                      3055052\n",
      "subject_direction_qualifier    4009644\n",
      "object_direction_qualifier     4009644\n",
      "object_aspect_qualifier        4009644\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check missingness\n",
    "print(edge_df[['predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier']].isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "id": "50efbe2a-d3a9-4250-8f7b-5e2fa611a6e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                predicate subject_direction_qualifier  \\\n",
      "0                                    None                        None   \n",
      "1         biolink:physical_interacts_with                        None   \n",
      "2                   biolink:has_phenotype                        None   \n",
      "3                     biolink:broad_match                        None   \n",
      "4     biolink:genetically_associated_with                        None   \n",
      "5            biolink:has_contraindication                        None   \n",
      "6                      biolink:target_for                        None   \n",
      "7  biolink:GeneAffectsChemicalAssociation                        None   \n",
      "8                          biolink:treats                        None   \n",
      "9           biolink:can_be_carried_out_by                        None   \n",
      "\n",
      "  object_direction_qualifier object_aspect_qualifier    count  \n",
      "0                       None                    None  3055052  \n",
      "1                       None                    None   642150  \n",
      "2                       None                    None   111000  \n",
      "3                       None                    None    86924  \n",
      "4                       None                    None    67034  \n",
      "5                       None                    None    24962  \n",
      "6                       None                    None    16380  \n",
      "7                       None                    None     3092  \n",
      "8                       None                    None     2186  \n",
      "9                       None                    None      864  \n"
     ]
    }
   ],
   "source": [
    "## fill na temporily to get the combo number\n",
    "filled = edge_df[['predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier']].fillna('None')\n",
    "counts = filled.value_counts().reset_index(name='count')\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7908c29f-40b9-498c-8b1c-f0f4657a9621",
   "metadata": {},
   "source": [
    "## Now download the concatenated node & edge files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "001d25d2-52e8-4331-ac10-86ba358f7ba9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the output path for node & edge files after formatting\n",
    "download_path_node_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/wQualityControl/PrimeKG_parsed_node_{version_number}.tsv'\n",
    "download_path_edge_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/wQualityControl/PrimeKG_parsed_edge_{version_number}.tsv'\n",
    "\n",
    "## download both node and edge files\n",
    "## Download the result df\n",
    "concat_node_df.to_csv(download_path_node_file, sep ='\\t', index=False)\n",
    "edge_output_df.to_csv(download_path_edge_file, sep ='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "c4b94c89-00da-461c-ba8e-ad8f34d96271",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The formatted node file will be saved in this path:  /Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/wQualityControl/PrimeKG_parsed_node_01_06_2026.tsv\n",
      "The formatted edge file will be saved in this path:  /Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/wQualityControl/PrimeKG_parsed_edge_01_06_2026.tsv\n"
     ]
    }
   ],
   "source": [
    "print(\"The formatted node file will be saved in this path: \", download_path_node_file)\n",
    "print(\"The formatted edge file will be saved in this path: \", download_path_edge_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "29e53e4d-193e-4830-8826-03874958a7dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## throw away those rows which either subject_category, object_category, or predicate is not start with \"biolink:\" prefix\n",
    "# ## since they cannot be biolink-compliant converted\n",
    "# ## Keep only rows where all three columns start with 'biolink:'\n",
    "# cellmarker_signor_edge_df_filtered = cellmarker_signor_edge_df_filtered[\n",
    "#     cellmarker_signor_edge_df_filtered['subject_category'].str.startswith('biolink:') &\n",
    "#     cellmarker_signor_edge_df_filtered['object_category'].str.startswith('biolink:') &\n",
    "#     cellmarker_signor_edge_df_filtered['predicate'].str.startswith('biolink:')\n",
    "# ]\n",
    "\n",
    "# cellmarker_signor_node_df = cellmarker_signor_node_df[\n",
    "#     cellmarker_signor_node_df['category'].str.startswith('biolink:')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "828def99-611a-4d39-b430-2c11ba179f29",
   "metadata": {},
   "source": [
    "## Now checking how many duplicates are there in the BigGIM edge file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "a4716c58-dd37-4973-8ef7-aec81f91a5ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the version number\n",
    "BigGIM_version_number = \"03_04_2025\"\n",
    "BigGIM_path_edge_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/BigGIM_plover_deployment/files/BigGIM_DrugResponse/biolink_compliant/BigGIM.interacts_with_associated_with.edges.biolink.compliant.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "2993c7ab-5a30-4937-858c-e9b59283a8bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_71220/4120548939.py:1: DtypeWarning: Columns (5,6,12,13,14,16,17,19) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  BigGIM_edge_df = pd.read_csv(BigGIM_path_edge_file, sep = '\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "      <th>agent_type</th>\n",
       "      <th>knowledge_level</th>\n",
       "      <th>anatomical_context_qualifier</th>\n",
       "      <th>knowledge_source</th>\n",
       "      <th>object_category</th>\n",
       "      <th>publications</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>...</th>\n",
       "      <th>subject_aspect_qualifier</th>\n",
       "      <th>supporting_study_cohort</th>\n",
       "      <th>Data_set</th>\n",
       "      <th>P_value</th>\n",
       "      <th>context_qualifier</th>\n",
       "      <th>statistics_method</th>\n",
       "      <th>supporting_study_size</th>\n",
       "      <th>object_aspect_qualifier</th>\n",
       "      <th>deploy_date</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:10461</td>\n",
       "      <td>biolink:expressed_in</td>\n",
       "      <td>CL:0000235</td>\n",
       "      <td>manual_agent</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>UBERON_0000916</td>\n",
       "      <td>CellMarker</td>\n",
       "      <td>biolink:Cell</td>\n",
       "      <td>PMID:31982413</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-03-01</td>\n",
       "      <td>b041e47f-c4f5-5166-b4fe-557207fcb593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBIGene:2215</td>\n",
       "      <td>biolink:expressed_in</td>\n",
       "      <td>CL:0000235</td>\n",
       "      <td>manual_agent</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>UBERON_0000916</td>\n",
       "      <td>CellMarker</td>\n",
       "      <td>biolink:Cell</td>\n",
       "      <td>PMID:31982413</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-03-01</td>\n",
       "      <td>99f71540-4efa-55bf-87f1-48cdceb65579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBIGene:4360</td>\n",
       "      <td>biolink:expressed_in</td>\n",
       "      <td>CL:0000235</td>\n",
       "      <td>manual_agent</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>UBERON_0000916</td>\n",
       "      <td>CellMarker</td>\n",
       "      <td>biolink:Cell</td>\n",
       "      <td>PMID:31982413</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-03-01</td>\n",
       "      <td>9683d6f3-ab01-5331-8a8d-0b1dcd291432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBIGene:11326</td>\n",
       "      <td>biolink:expressed_in</td>\n",
       "      <td>CL:0000235</td>\n",
       "      <td>manual_agent</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>UBERON_0000916</td>\n",
       "      <td>CellMarker</td>\n",
       "      <td>biolink:Cell</td>\n",
       "      <td>PMID:31982413</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-03-01</td>\n",
       "      <td>b035b572-65d6-5602-ad4e-bbe2c36cf9e9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBIGene:9332</td>\n",
       "      <td>biolink:expressed_in</td>\n",
       "      <td>CL:0000235</td>\n",
       "      <td>manual_agent</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>UBERON_0000916</td>\n",
       "      <td>CellMarker</td>\n",
       "      <td>biolink:Cell</td>\n",
       "      <td>PMID:31982413</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-03-01</td>\n",
       "      <td>20203999-528e-5335-906e-5441d94ee18c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          subject             predicate      object    agent_type  \\\n",
       "0  NCBIGene:10461  biolink:expressed_in  CL:0000235  manual_agent   \n",
       "1   NCBIGene:2215  biolink:expressed_in  CL:0000235  manual_agent   \n",
       "2   NCBIGene:4360  biolink:expressed_in  CL:0000235  manual_agent   \n",
       "3  NCBIGene:11326  biolink:expressed_in  CL:0000235  manual_agent   \n",
       "4   NCBIGene:9332  biolink:expressed_in  CL:0000235  manual_agent   \n",
       "\n",
       "       knowledge_level anatomical_context_qualifier knowledge_source  \\\n",
       "0  knowledge_assertion               UBERON_0000916       CellMarker   \n",
       "1  knowledge_assertion               UBERON_0000916       CellMarker   \n",
       "2  knowledge_assertion               UBERON_0000916       CellMarker   \n",
       "3  knowledge_assertion               UBERON_0000916       CellMarker   \n",
       "4  knowledge_assertion               UBERON_0000916       CellMarker   \n",
       "\n",
       "  object_category   publications subject_category  ...  \\\n",
       "0    biolink:Cell  PMID:31982413     biolink:Gene  ...   \n",
       "1    biolink:Cell  PMID:31982413     biolink:Gene  ...   \n",
       "2    biolink:Cell  PMID:31982413     biolink:Gene  ...   \n",
       "3    biolink:Cell  PMID:31982413     biolink:Gene  ...   \n",
       "4    biolink:Cell  PMID:31982413     biolink:Gene  ...   \n",
       "\n",
       "   subject_aspect_qualifier  supporting_study_cohort Data_set P_value  \\\n",
       "0                       NaN                      NaN      NaN     NaN   \n",
       "1                       NaN                      NaN      NaN     NaN   \n",
       "2                       NaN                      NaN      NaN     NaN   \n",
       "3                       NaN                      NaN      NaN     NaN   \n",
       "4                       NaN                      NaN      NaN     NaN   \n",
       "\n",
       "  context_qualifier  statistics_method supporting_study_size  \\\n",
       "0               NaN                NaN                   NaN   \n",
       "1               NaN                NaN                   NaN   \n",
       "2               NaN                NaN                   NaN   \n",
       "3               NaN                NaN                   NaN   \n",
       "4               NaN                NaN                   NaN   \n",
       "\n",
       "  object_aspect_qualifier  deploy_date                                    id  \n",
       "0                     NaN   2025-03-01  b041e47f-c4f5-5166-b4fe-557207fcb593  \n",
       "1                     NaN   2025-03-01  99f71540-4efa-55bf-87f1-48cdceb65579  \n",
       "2                     NaN   2025-03-01  9683d6f3-ab01-5331-8a8d-0b1dcd291432  \n",
       "3                     NaN   2025-03-01  b035b572-65d6-5602-ad4e-bbe2c36cf9e9  \n",
       "4                     NaN   2025-03-01  20203999-528e-5335-906e-5441d94ee18c  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BigGIM_edge_df = pd.read_csv(BigGIM_path_edge_file, sep = '\\t')\n",
    "BigGIM_edge_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "a3c3118d-77e1-4ca8-a89a-244756c8165d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicated rows: 148185\n"
     ]
    }
   ],
   "source": [
    "# Specify the columns to compare\n",
    "cols = ['subject', 'object', 'predicate']\n",
    "\n",
    "# Use merge to find common rows based on the specified columns\n",
    "common = pd.merge(edge_final_df[cols], BigGIM_edge_df[cols], on=cols)\n",
    "\n",
    "# Get the number of duplicated rows\n",
    "num_duplicates = len(common)\n",
    "\n",
    "print(f\"Number of duplicated rows: {num_duplicates}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf2efd46-5831-4696-851b-02c6a15d7495",
   "metadata": {},
   "source": [
    "## Now checking how many duplicates are there in the PharmGKB edge file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "ea98e53a-9ec8-45ab-9bf6-a52d47aa426a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>publications</th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>subject_category</th>\n",
       "      <th>object_category</th>\n",
       "      <th>predicate</th>\n",
       "      <th>knowledge_source</th>\n",
       "      <th>knowledge_level</th>\n",
       "      <th>agent_type</th>\n",
       "      <th>deploy_date</th>\n",
       "      <th>context_qualifier</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>25695618</td>\n",
       "      <td>NCBIGene:162282</td>\n",
       "      <td>MeSH:D000075222</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Disease</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3f4622cd-cc81-5c60-9fce-4d25214b9b86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25695618</td>\n",
       "      <td>NCBIGene:162282</td>\n",
       "      <td>PUBCHEM.COMPOUND:3639</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2ebd74fd-1cb6-519c-9051-4e0e5880a57d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18511948;25545243</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>PUBCHEM.COMPOUND:3385</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3e9e9eee-e33e-5903-b6c0-fd6f960ab809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>24924344;25545243</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>PUBCHEM.COMPOUND:6857599</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5dfcff5b-dea4-53f7-b9ff-1db8d23309df</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30237583</td>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>MeSH:D046152</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>biolink:Disease</td>\n",
       "      <td>biolink:associated_with</td>\n",
       "      <td>PharmGKB</td>\n",
       "      <td>knowledge_assertion</td>\n",
       "      <td>automated_agent</td>\n",
       "      <td>2025-07-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6149843b-5b1d-562b-b2e9-82391a552db3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        publications          subject                    object  \\\n",
       "0           25695618  NCBIGene:162282           MeSH:D000075222   \n",
       "1           25695618  NCBIGene:162282     PUBCHEM.COMPOUND:3639   \n",
       "2  18511948;25545243    NCBIGene:1728     PUBCHEM.COMPOUND:3385   \n",
       "3  24924344;25545243    NCBIGene:1728  PUBCHEM.COMPOUND:6857599   \n",
       "4           30237583    NCBIGene:1728              MeSH:D046152   \n",
       "\n",
       "  subject_category         object_category                predicate  \\\n",
       "0     biolink:Gene         biolink:Disease  biolink:associated_with   \n",
       "1     biolink:Gene  biolink:ChemicalEntity  biolink:associated_with   \n",
       "2     biolink:Gene  biolink:ChemicalEntity  biolink:associated_with   \n",
       "3     biolink:Gene  biolink:ChemicalEntity  biolink:associated_with   \n",
       "4     biolink:Gene         biolink:Disease  biolink:associated_with   \n",
       "\n",
       "  knowledge_source      knowledge_level       agent_type deploy_date  \\\n",
       "0         PharmGKB  knowledge_assertion  automated_agent  2025-07-08   \n",
       "1         PharmGKB  knowledge_assertion  automated_agent  2025-07-08   \n",
       "2         PharmGKB  knowledge_assertion  automated_agent  2025-07-08   \n",
       "3         PharmGKB  knowledge_assertion  automated_agent  2025-07-08   \n",
       "4         PharmGKB  knowledge_assertion  automated_agent  2025-07-08   \n",
       "\n",
       "   context_qualifier                                    id  \n",
       "0                NaN  3f4622cd-cc81-5c60-9fce-4d25214b9b86  \n",
       "1                NaN  2ebd74fd-1cb6-519c-9051-4e0e5880a57d  \n",
       "2                NaN  3e9e9eee-e33e-5903-b6c0-fd6f960ab809  \n",
       "3                NaN  5dfcff5b-dea4-53f7-b9ff-1db8d23309df  \n",
       "4                NaN  6149843b-5b1d-562b-b2e9-82391a552db3  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Define the version number\n",
    "PharmGKB_version_number = \"07_08_2025\"\n",
    "PharmGKB_path_edge_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/PharmGKB_parsed_edge_{PharmGKB_version_number}.tsv'\n",
    "\n",
    "PharmGKB_edge_df = pd.read_csv(PharmGKB_path_edge_file, sep = '\\t')\n",
    "PharmGKB_edge_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "f8c3e922-0cb6-4e64-b3c0-0df18aa9ed0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "subject_category\n",
      "biolink:Gene              11702\n",
      "biolink:ChemicalEntity     5907\n",
      "biolink:Disease            3349\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = PharmGKB_edge_df['subject_category'].value_counts()\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "1e1fa2df-5da5-4aa2-8604-8954a88b1659",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "subject_category\n",
      "biolink:Gene              11702\n",
      "biolink:ChemicalEntity     5907\n",
      "biolink:Disease            3349\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = PharmGKB_edge_df['subject_category'].value_counts()\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "28bd323a-8450-459e-9601-7367e6b84a3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "predicate\n",
      "biolink:associated_with    20958\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = PharmGKB_edge_df['predicate'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "b1b25de2-4b3c-4fa2-a539-e1500f97bf1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicated rows: 0\n"
     ]
    }
   ],
   "source": [
    "# Specify the columns to compare\n",
    "cols = ['subject', 'object', 'predicate']\n",
    "\n",
    "# Use merge to find common rows based on the specified columns\n",
    "common = pd.merge(edge_final_df[cols], PharmGKB_edge_df[cols], on=cols)\n",
    "\n",
    "# Get the number of duplicated rows\n",
    "num_duplicates = len(common)\n",
    "\n",
    "print(f\"Number of duplicated rows: {num_duplicates}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "3b469c05-99a2-4f89-b56a-190f8241e0b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:162282</td>\n",
       "      <td>ANKFN1</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBIGene:1728</td>\n",
       "      <td>NQO1</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBIGene:6813</td>\n",
       "      <td>STXBP2</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PUBCHEM.COMPOUND:33</td>\n",
       "      <td>chloroacetaldehyde</td>\n",
       "      <td>biolink:ChemicalEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBIGene:8000</td>\n",
       "      <td>PSCA</td>\n",
       "      <td>biolink:Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id                name                category\n",
       "0      NCBIGene:162282              ANKFN1            biolink:Gene\n",
       "1        NCBIGene:1728                NQO1            biolink:Gene\n",
       "2        NCBIGene:6813              STXBP2            biolink:Gene\n",
       "3  PUBCHEM.COMPOUND:33  chloroacetaldehyde  biolink:ChemicalEntity\n",
       "4        NCBIGene:8000                PSCA            biolink:Gene"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PharmGKB_path_node_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/PharmGKB_parsed_node_{version_number}.tsv'\n",
    "\n",
    "PharmGKB_node_df = pd.read_csv(PharmGKB_path_node_file, sep = '\\t')\n",
    "PharmGKB_node_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "8eb23046-341e-41cd-94a0-b5595c938d1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Concatenate to obtain overall node & edge dataframes\n",
    "\n",
    "list_of_input_node_files = [node_final_df, PharmGKB_node_df]\n",
    "\n",
    "list_of_input_edge_files = [edge_final_df, PharmGKB_edge_df]\n",
    "\n",
    "node_final_df = pd.concat(list_of_input_node_files).drop_duplicates()\n",
    "edge_final_df = pd.concat(list_of_input_edge_files).drop_duplicates()\n",
    "\n",
    "# edge_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "c0682b6b-915a-4aa7-a7c2-79161b60ce2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## remove NA or empty strings rows from both subject and object columns\n",
    "# Drop rows where 'name' is NaN, None, or empty string\n",
    "\n",
    "edge_final_df = edge_final_df[~edge_final_df['subject'].isna() & (edge_final_df['subject'].str.strip() != '')]\n",
    "edge_final_df = edge_final_df[~edge_final_df['object'].isna() & (edge_final_df['object'].str.strip() != '')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "e167f410-41f0-4a51-8e3a-03f33f21f349",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes 52602\n",
      "Number of edges 599515\n",
      "Average degree 22.794380441808297\n"
     ]
    }
   ],
   "source": [
    "## Create a graph from the DataFrame\n",
    "graph = nx.from_pandas_edgelist(edge_final_df, 'subject', 'object', edge_attr='predicate')\n",
    "\n",
    "## Print graph information\n",
    "print('Number of nodes', len(set(graph.nodes)))\n",
    "print('Number of edges', len(set(graph.edges)))\n",
    "print('Average degree', sum(dict(graph.degree).values()) / len(graph.nodes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "c1f8c720-6b05-4d14-b855-9b4a9269f7b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the output path for node & edge files after formatting\n",
    "download_path_node_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/Pharmagenomics_KG_parsed_node_{version_number}.tsv'\n",
    "download_path_edge_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/Pharmagenomics_KG_parsed_edge_{version_number}.tsv'\n",
    "\n",
    "## download both node and edge files\n",
    "## Download the result df\n",
    "## disable download for testing\n",
    "node_final_df.to_csv(download_path_node_file, sep ='\\t', index=False)\n",
    "edge_final_df.to_csv(download_path_edge_file, sep ='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "ffeae20a-73d5-46e2-9a7f-baf50e599183",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The formatted node file will be saved in this path:  /Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/Pharmagenomics_KG_parsed_node_07_08_2025.tsv\n",
      "The formatted edge file will be saved in this path:  /Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/Pharmagenomics_KG_parsed_edge_07_08_2025.tsv\n"
     ]
    }
   ],
   "source": [
    "print(\"The formatted node file will be saved in this path: \", download_path_node_file)\n",
    "print(\"The formatted edge file will be saved in this path: \", download_path_edge_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "d64ede71-6cc4-4ab6-a5b0-5c719775437d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['subject', 'object', 'subject_category', 'object_category', 'predicate',\n",
       "       'knowledge_source', 'knowledge_level', 'agent_type', 'deploy_date',\n",
       "       'context_qualifier', 'id', 'predicate_source_name',\n",
       "       'anatomical_context_qualifier', 'publications'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_final_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "967ece2c-61ca-4905-a2ec-a2162c582efd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "knowledge_source\n",
      "PrimeKG          914870\n",
      "PharmGKB          20427\n",
      "CellMarker         6599\n",
      "SIGNOR-250870         1\n",
      "SIGNOR-58928          1\n",
      "                  ...  \n",
      "SIGNOR-249677         1\n",
      "SIGNOR-277555         1\n",
      "SIGNOR-256127         1\n",
      "SIGNOR-261145         1\n",
      "SIGNOR-261887         1\n",
      "Name: count, Length: 33227, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## Check the knowledge_source column again\n",
    "## Count occurrences of each unique value in 'knowledge_source'\n",
    "counts = edge_final_df['knowledge_source'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "274f9d23-40d3-4955-8ccc-62f64124bcc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(0)"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_final_df['knowledge_source'].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "4ebf3711-0e33-4df4-83bc-9457782366e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(0)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_final_df['id'].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "22946e9d-c24f-4518-b7d6-86192ea32def",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(0)"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_final_df['predicate'].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "5a71ffc1-6342-41fe-8c9d-eec88078e9a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(0)"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_final_df['object'].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "32cfd59c-ed50-4548-b5e6-4a270aed65e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(0)"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_final_df['subject'].isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9270ee0d-419f-4580-93b8-63aa8d1dc462",
   "metadata": {},
   "source": [
    "## following codes are used for quality control and sanity check\n",
    "* check and confirm all subject & object types are correctly formatted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "9d9a0010-4ca3-49ad-9f13-ae0daa84a68f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "subject_category\n",
      "biolink:Gene                     705625\n",
      "biolink:Disease                  163798\n",
      "biolink:Drug                      91961\n",
      "biolink:ChemicalEntity             8166\n",
      "biolink:MacromolecularComplex      2632\n",
      "biolink:ProteinFamily              1341\n",
      "biolink:SmallMolecule              1085\n",
      "biolink:Protein                     357\n",
      "biolink:EnvironmentalProcess        109\n",
      "biolink:MicroRNA                     43\n",
      "biolink:Noncoding_RNAProduct          3\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = edge_final_df['subject_category'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "3a53e4e7-e608-4a2d-8013-414ef07a7b93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "object_category\n",
      "biolink:Gene                          683140\n",
      "biolink:DiseaseOrPhenotypicFeature    175784\n",
      "biolink:Disease                        79946\n",
      "biolink:Drug                           20336\n",
      "biolink:Cell                            6599\n",
      "biolink:ChemicalEntity                  5644\n",
      "biolink:PhenotypicFeature               1573\n",
      "biolink:MacromolecularComplex           1218\n",
      "biolink:SmallMolecule                    426\n",
      "biolink:ProteinFamily                    229\n",
      "biolink:Protein                          225\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = edge_final_df['object_category'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "638c86dd-e26e-43dc-a0cb-16d88abc52f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "predicate\n",
      "biolink:physical_interacts_with           642150\n",
      "biolink:has_phenotype                     111000\n",
      "biolink:has_side_effect                    64784\n",
      "biolink:broad_match                        49452\n",
      "biolink:regulates                          33096\n",
      "biolink:has_contraindication               24962\n",
      "biolink:associated_with                    20427\n",
      "biolink:target_for                         16380\n",
      "biolink:expressed_in                        6599\n",
      "biolink:GeneAffectsChemicalAssociation      3092\n",
      "biolink:treats                              2186\n",
      "biolink:can_be_carried_out_by                864\n",
      "biolink:in_complex_with                      128\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## check all unique predicate values\n",
    "counts = edge_final_df['predicate'].value_counts()\n",
    "\n",
    "print(counts)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (pytorch_env)",
   "language": "python",
   "name": "pytorch_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
